{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "dbacb7ca",
   "metadata": {},
   "source": [
    "CellChat inference and analysis of spatially proximal cell-cell communication from spatially resolved transcriptomics (multiple datasets)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "f532b324",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: dplyr\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "Attaching package: ‘dplyr’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    filter, lag\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    intersect, setdiff, setequal, union\n",
      "\n",
      "\n",
      "Loading required package: igraph\n",
      "\n",
      "\n",
      "Attaching package: ‘igraph’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:dplyr’:\n",
      "\n",
      "    as_data_frame, groups, union\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    decompose, spectrum\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:base’:\n",
      "\n",
      "    union\n",
      "\n",
      "\n",
      "Loading required package: ggplot2\n",
      "\n",
      "Attaching SeuratObject\n",
      "\n"
     ]
    }
   ],
   "source": [
    "ptm = Sys.time()\n",
    "\n",
    "library(CellChat)\n",
    "library(Seurat)\n",
    "library(patchwork)\n",
    "options(stringsAsFactors = FALSE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "f51bd276",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "results_folder ='/run/user/1804238067/gvfs/sftp:host=clust1-sub-1,user=lythgo02/mnt/nas-data/fmlab/group_folders/lythgo02/OV_visium/emily/cell2location/cellTrek/cellchat'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "f43705a7",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "#cellTrekObs <- readRDS(paste0(ov_visium, \"ov_cellTrek_obj.rds\"))\n",
    "#cell2loc = '/run/user/1804238067/gvfs/sftp:host=clust1-sub-1,user=lythgo02/mnt/scratchc/fmlab/lythgo02/OV_visium/emily/cell2location/'\n",
    "#adata_vis_match <- zellkonverter::readH5AD(file.path(cell2loc, \"final_adata_vis_match.h5ad\"))\n",
    "\n",
    "cell2loc = '/run/user/1804238067/gvfs/sftp:host=clust1-sub-1,user=lythgo02/mnt/nas-data/fmlab/group_folders/lythgo02/OV_visium/emily/cell2location/'\n",
    "#ovTrainAll <- readRDS(paste0(cell2loc, \"cellTrek/celltrek_results_ntree1000_p500_celltrek.rds\"))\n",
    "ovCellTrekList <- readRDS(paste0(cell2loc, \"cellTrek/celltrek_results_updated.rds\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "98181c69",
   "metadata": {},
   "source": [
    "For one sample at a time"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a14e9f31",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "ovCellTrek <- adata_vis_match[[1]]\n",
    "# 1. Extract expression and meta\n",
    "data.input <- GetAssayData(ovCellTrek, slot = \"data\")  # or \"counts\"\n",
    "meta <- ovCellTrek@meta.data\n",
    "# 2. Create CellChat object\n",
    "cellchat <- CellChat::createCellChat(object = data.input, meta = meta, group.by = \"cell_type\")\n",
    "CellChatDB <- CellChatDB.mouse  # or .human\n",
    "cellchat@DB <- CellChatDB\n",
    "cellchat <- subsetData(cellchat)\n",
    "cellchat <- identifyOverExpressedGenes(cellchat)\n",
    "cellchat <- identifyOverExpressedInteractions(cellchat)\n",
    "cellchat <- computeCommunProb(cellchat)\n",
    "cellchat <- computeCommunProbPathway(cellchat)\n",
    "cellchat <- aggregateNet(cellchat)\n",
    "# all predicted interactions (data.frame)\n",
    "comm <- subsetCommunication(cellchat)\n",
    "head(comm)\n",
    "# save\n",
    "write.csv(comm, file = file.path(results_folder, \"ov1_cellchat_all_interactions.csv\"), row.names = FALSE)\n",
    "# list pathways and group sizes\n",
    "signaling_paths <- names(cellchat@netP)\n",
    "print(signaling_paths)\n",
    "groupSize <- as.numeric(table(cellchat@idents))\n",
    "groupSize\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "043b5ebc",
   "metadata": {},
   "source": [
    "For all in a loop"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "d2c5cfef",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Creating CellChat object for OV_1\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Create a CellChat object from a data matrix\"\n",
      "Set cell identities for the new CellChat object \n",
      "The cell groups used for CellChat analysis are  B naive activated, B naive resting, CD4, CD8, Cycling Plasma Cells, Dendritic Cells, Endothelial, Macrophage, Mesenchymal, Monocytes, Neutrophils, NK_cell, Plasma Cells, Th1, Th17, Treg, Tumour \n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Creating CellChat object for OV_2\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Create a CellChat object from a data matrix\"\n",
      "Set cell identities for the new CellChat object \n",
      "The cell groups used for CellChat analysis are  B naive activated, B naive resting, CD4, CD8, Cycling Plasma Cells, Dendritic Cells, Endothelial, Macrophage, Mesenchymal, Monocytes, Neutrophils, NK_cell, Plasma Cells, Th1, Th17, Treg, Tumour \n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Creating CellChat object for OV_3\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Create a CellChat object from a data matrix\"\n",
      "Set cell identities for the new CellChat object \n",
      "The cell groups used for CellChat analysis are  B naive activated, B naive resting, CD4, CD8, Dendritic Cells, Endothelial, Macrophage, Mesenchymal, Monocytes, Neutrophils, NK_cell, Plasma Cells, Th1, Th17, Treg, Tumour \n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Creating CellChat object for OV_4\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Create a CellChat object from a data matrix\"\n",
      "Set cell identities for the new CellChat object \n",
      "The cell groups used for CellChat analysis are  B naive activated, B naive resting, CD4, CD8, Cycling Plasma Cells, Dendritic Cells, Endothelial, Macrophage, Mesenchymal, Monocytes, Neutrophils, NK_cell, Plasma Cells, Th1, Th17, Treg, Tumour \n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Creating CellChat object for OV_5\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Create a CellChat object from a data matrix\"\n",
      "Set cell identities for the new CellChat object \n",
      "The cell groups used for CellChat analysis are  B naive activated, B naive resting, CD4, CD8, Dendritic Cells, Endothelial, Macrophage, Mesenchymal, Monocytes, Neutrophils, NK_cell, Plasma Cells, Th1, Th17, Treg, Tumour \n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Creating CellChat object for OV_6\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Create a CellChat object from a data matrix\"\n",
      "Set cell identities for the new CellChat object \n",
      "The cell groups used for CellChat analysis are  B naive activated, B naive resting, CD4, CD8, Cycling Plasma Cells, Dendritic Cells, Endothelial, Macrophage, Mesenchymal, Monocytes, Neutrophils, NK_cell, Plasma Cells, Th1, Th17, Treg, Tumour \n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "$OV_1\n",
       "An object of class CellChat created from a single dataset \n",
       " 14542 genes.\n",
       " 9276 cells. \n",
       "CellChat analysis of single cell RNA-seq data! \n",
       "\n",
       "$OV_2\n",
       "An object of class CellChat created from a single dataset \n",
       " 14542 genes.\n",
       " 13696 cells. \n",
       "CellChat analysis of single cell RNA-seq data! \n",
       "\n",
       "$OV_3\n",
       "An object of class CellChat created from a single dataset \n",
       " 14542 genes.\n",
       " 10291 cells. \n",
       "CellChat analysis of single cell RNA-seq data! \n",
       "\n",
       "$OV_4\n",
       "An object of class CellChat created from a single dataset \n",
       " 14542 genes.\n",
       " 5964 cells. \n",
       "CellChat analysis of single cell RNA-seq data! \n",
       "\n",
       "$OV_5\n",
       "An object of class CellChat created from a single dataset \n",
       " 14542 genes.\n",
       " 9080 cells. \n",
       "CellChat analysis of single cell RNA-seq data! \n",
       "\n",
       "$OV_6\n",
       "An object of class CellChat created from a single dataset \n",
       " 14542 genes.\n",
       " 5391 cells. \n",
       "CellChat analysis of single cell RNA-seq data! \n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "\n",
    "cellchat_list <- list()\n",
    "\n",
    "for (i in seq_along(ovCellTrekList)) {\n",
    "  \n",
    "  sample_name <- if (!is.null(names(ovCellTrekList)) && names(ovCellTrekList)[i] != \"\") {\n",
    "    names(ovCellTrekList)[i]\n",
    "  } else {\n",
    "    paste0(\"sample_\", i)\n",
    "  }\n",
    "  message(\"Creating CellChat object for \", sample_name)\n",
    "  \n",
    "  ovCellTrek <- ovCellTrekList[[i]]\n",
    "  \n",
    "  # 1. Extract expression & metadata\n",
    "  data.input <- GetAssayData(ovCellTrek, slot = \"data\")\n",
    "  meta <- ovCellTrek@meta.data\n",
    "  \n",
    "  # Ensure you have a grouping column (rename if needed)\n",
    "  # e.g. your column might be \"cell_type_collapsed\"\n",
    "  if (!\"cell_type\" %in% colnames(meta) && \"cell_type_collapsed\" %in% colnames(meta)) {\n",
    "    meta$cell_type <- meta$cell_type_collapsed\n",
    "  }\n",
    "  \n",
    "  # 2. Create CellChat object\n",
    "  cellchat <- CellChat::createCellChat(\n",
    "    object = data.input,\n",
    "    meta = meta,\n",
    "    group.by = \"cell_type\"\n",
    "  )\n",
    "  \n",
    "  cellchat_list[[sample_name]] <- cellchat\n",
    "}\n",
    "\n",
    "# --- Optional: preview what was created ---\n",
    "cellchat_list\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "9676385c",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The number of highly variable ligand-receptor pairs used for signaling inference is 1255 \n",
      "triMean is used for calculating the average gene expression per cell group. \n",
      "[1] \">>> Run CellChat on sc/snRNA-seq data <<< [2025-10-27 16:47:57.542365]\"\n",
      "[1] \">>> CellChat inference is done. Parameter values are stored in `object@options$parameter` <<< [2025-10-27 16:52:59.831249]\"\n",
      "The number of highly variable ligand-receptor pairs used for signaling inference is 1286 \n",
      "triMean is used for calculating the average gene expression per cell group. \n",
      "[1] \">>> Run CellChat on sc/snRNA-seq data <<< [2025-10-27 16:53:05.010571]\"\n",
      "[1] \">>> CellChat inference is done. Parameter values are stored in `object@options$parameter` <<< [2025-10-27 16:58:42.816095]\"\n",
      "The number of highly variable ligand-receptor pairs used for signaling inference is 1256 \n",
      "triMean is used for calculating the average gene expression per cell group. \n",
      "[1] \">>> Run CellChat on sc/snRNA-seq data <<< [2025-10-27 16:58:47.366644]\"\n",
      "[1] \">>> CellChat inference is done. Parameter values are stored in `object@options$parameter` <<< [2025-10-27 17:03:52.626945]\"\n",
      "The number of highly variable ligand-receptor pairs used for signaling inference is 1276 \n",
      "triMean is used for calculating the average gene expression per cell group. \n",
      "[1] \">>> Run CellChat on sc/snRNA-seq data <<< [2025-10-27 17:03:56.353556]\"\n",
      "[1] \">>> CellChat inference is done. Parameter values are stored in `object@options$parameter` <<< [2025-10-27 17:08:40.708992]\"\n",
      "The number of highly variable ligand-receptor pairs used for signaling inference is 1286 \n",
      "triMean is used for calculating the average gene expression per cell group. \n",
      "[1] \">>> Run CellChat on sc/snRNA-seq data <<< [2025-10-27 17:08:45.060249]\"\n",
      "[1] \">>> CellChat inference is done. Parameter values are stored in `object@options$parameter` <<< [2025-10-27 17:13:44.421377]\"\n",
      "The number of highly variable ligand-receptor pairs used for signaling inference is 1294 \n",
      "triMean is used for calculating the average gene expression per cell group. \n",
      "[1] \">>> Run CellChat on sc/snRNA-seq data <<< [2025-10-27 17:13:48.047112]\"\n",
      "[1] \">>> CellChat inference is done. Parameter values are stored in `object@options$parameter` <<< [2025-10-27 17:18:23.598066]\"\n"
     ]
    }
   ],
   "source": [
    "# Load mouse ligand-receptor database\n",
    "CellChatDB <- CellChatDB.mouse\n",
    "for (i in names(cellchat_list)) {\n",
    "  cellchat_list[[i]]@DB <- CellChatDB\n",
    "  cellchat_list[[i]] <- subsetData(cellchat_list[[i]])           # subset relevant genes\n",
    "  cellchat_list[[i]] <- identifyOverExpressedGenes(cellchat_list[[i]])\n",
    "  cellchat_list[[i]] <- identifyOverExpressedInteractions(cellchat_list[[i]])\n",
    "  cellchat_list[[i]] <- computeCommunProb(cellchat_list[[i]])\n",
    "  cellchat_list[[i]] <- computeCommunProbPathway(cellchat_list[[i]])\n",
    "  cellchat_list[[i]] <- aggregateNet(cellchat_list[[i]])\n",
    "}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "d965a1c5",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 6 × 12</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>sample</th><th scope=col>source</th><th scope=col>target</th><th scope=col>ligand</th><th scope=col>receptor</th><th scope=col>prob</th><th scope=col>pval</th><th scope=col>interaction_name</th><th scope=col>interaction_name_2</th><th scope=col>pathway_name</th><th scope=col>annotation</th><th scope=col>evidence</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>OV_1</td><td>B naive activated</td><td>B naive activated</td><td>Tgfb1</td><td>TGFbR1_R2</td><td>0.02177701</td><td>0</td><td>TGFB1_TGFBR1_TGFBR2</td><td>Tgfb1 - (Tgfbr1+Tgfbr2)</td><td>TGFb</td><td>Secreted Signaling</td><td>KEGG: mmu04350</td></tr>\n",
       "\t<tr><th scope=row>2</th><td>OV_1</td><td>B naive resting  </td><td>B naive activated</td><td>Tgfb1</td><td>TGFbR1_R2</td><td>0.02292956</td><td>0</td><td>TGFB1_TGFBR1_TGFBR2</td><td>Tgfb1 - (Tgfbr1+Tgfbr2)</td><td>TGFb</td><td>Secreted Signaling</td><td>KEGG: mmu04350</td></tr>\n",
       "\t<tr><th scope=row>3</th><td>OV_1</td><td>CD4              </td><td>B naive activated</td><td>Tgfb1</td><td>TGFbR1_R2</td><td>0.01804501</td><td>0</td><td>TGFB1_TGFBR1_TGFBR2</td><td>Tgfb1 - (Tgfbr1+Tgfbr2)</td><td>TGFb</td><td>Secreted Signaling</td><td>KEGG: mmu04350</td></tr>\n",
       "\t<tr><th scope=row>4</th><td>OV_1</td><td>CD8              </td><td>B naive activated</td><td>Tgfb1</td><td>TGFbR1_R2</td><td>0.02235130</td><td>0</td><td>TGFB1_TGFBR1_TGFBR2</td><td>Tgfb1 - (Tgfbr1+Tgfbr2)</td><td>TGFb</td><td>Secreted Signaling</td><td>KEGG: mmu04350</td></tr>\n",
       "\t<tr><th scope=row>5</th><td>OV_1</td><td>Dendritic Cells  </td><td>B naive activated</td><td>Tgfb1</td><td>TGFbR1_R2</td><td>0.01184832</td><td>0</td><td>TGFB1_TGFBR1_TGFBR2</td><td>Tgfb1 - (Tgfbr1+Tgfbr2)</td><td>TGFb</td><td>Secreted Signaling</td><td>KEGG: mmu04350</td></tr>\n",
       "\t<tr><th scope=row>6</th><td>OV_1</td><td>Endothelial      </td><td>B naive activated</td><td>Tgfb1</td><td>TGFbR1_R2</td><td>0.01188053</td><td>0</td><td>TGFB1_TGFBR1_TGFBR2</td><td>Tgfb1 - (Tgfbr1+Tgfbr2)</td><td>TGFb</td><td>Secreted Signaling</td><td>KEGG: mmu04350</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6 × 12\n",
       "\\begin{tabular}{r|llllllllllll}\n",
       "  & sample & source & target & ligand & receptor & prob & pval & interaction\\_name & interaction\\_name\\_2 & pathway\\_name & annotation & evidence\\\\\n",
       "  & <chr> & <fct> & <fct> & <chr> & <chr> & <dbl> & <dbl> & <fct> & <chr> & <chr> & <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\t1 & OV\\_1 & B naive activated & B naive activated & Tgfb1 & TGFbR1\\_R2 & 0.02177701 & 0 & TGFB1\\_TGFBR1\\_TGFBR2 & Tgfb1 - (Tgfbr1+Tgfbr2) & TGFb & Secreted Signaling & KEGG: mmu04350\\\\\n",
       "\t2 & OV\\_1 & B naive resting   & B naive activated & Tgfb1 & TGFbR1\\_R2 & 0.02292956 & 0 & TGFB1\\_TGFBR1\\_TGFBR2 & Tgfb1 - (Tgfbr1+Tgfbr2) & TGFb & Secreted Signaling & KEGG: mmu04350\\\\\n",
       "\t3 & OV\\_1 & CD4               & B naive activated & Tgfb1 & TGFbR1\\_R2 & 0.01804501 & 0 & TGFB1\\_TGFBR1\\_TGFBR2 & Tgfb1 - (Tgfbr1+Tgfbr2) & TGFb & Secreted Signaling & KEGG: mmu04350\\\\\n",
       "\t4 & OV\\_1 & CD8               & B naive activated & Tgfb1 & TGFbR1\\_R2 & 0.02235130 & 0 & TGFB1\\_TGFBR1\\_TGFBR2 & Tgfb1 - (Tgfbr1+Tgfbr2) & TGFb & Secreted Signaling & KEGG: mmu04350\\\\\n",
       "\t5 & OV\\_1 & Dendritic Cells   & B naive activated & Tgfb1 & TGFbR1\\_R2 & 0.01184832 & 0 & TGFB1\\_TGFBR1\\_TGFBR2 & Tgfb1 - (Tgfbr1+Tgfbr2) & TGFb & Secreted Signaling & KEGG: mmu04350\\\\\n",
       "\t6 & OV\\_1 & Endothelial       & B naive activated & Tgfb1 & TGFbR1\\_R2 & 0.01188053 & 0 & TGFB1\\_TGFBR1\\_TGFBR2 & Tgfb1 - (Tgfbr1+Tgfbr2) & TGFb & Secreted Signaling & KEGG: mmu04350\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6 × 12\n",
       "\n",
       "| <!--/--> | sample &lt;chr&gt; | source &lt;fct&gt; | target &lt;fct&gt; | ligand &lt;chr&gt; | receptor &lt;chr&gt; | prob &lt;dbl&gt; | pval &lt;dbl&gt; | interaction_name &lt;fct&gt; | interaction_name_2 &lt;chr&gt; | pathway_name &lt;chr&gt; | annotation &lt;chr&gt; | evidence &lt;chr&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| 1 | OV_1 | B naive activated | B naive activated | Tgfb1 | TGFbR1_R2 | 0.02177701 | 0 | TGFB1_TGFBR1_TGFBR2 | Tgfb1 - (Tgfbr1+Tgfbr2) | TGFb | Secreted Signaling | KEGG: mmu04350 |\n",
       "| 2 | OV_1 | B naive resting   | B naive activated | Tgfb1 | TGFbR1_R2 | 0.02292956 | 0 | TGFB1_TGFBR1_TGFBR2 | Tgfb1 - (Tgfbr1+Tgfbr2) | TGFb | Secreted Signaling | KEGG: mmu04350 |\n",
       "| 3 | OV_1 | CD4               | B naive activated | Tgfb1 | TGFbR1_R2 | 0.01804501 | 0 | TGFB1_TGFBR1_TGFBR2 | Tgfb1 - (Tgfbr1+Tgfbr2) | TGFb | Secreted Signaling | KEGG: mmu04350 |\n",
       "| 4 | OV_1 | CD8               | B naive activated | Tgfb1 | TGFbR1_R2 | 0.02235130 | 0 | TGFB1_TGFBR1_TGFBR2 | Tgfb1 - (Tgfbr1+Tgfbr2) | TGFb | Secreted Signaling | KEGG: mmu04350 |\n",
       "| 5 | OV_1 | Dendritic Cells   | B naive activated | Tgfb1 | TGFbR1_R2 | 0.01184832 | 0 | TGFB1_TGFBR1_TGFBR2 | Tgfb1 - (Tgfbr1+Tgfbr2) | TGFb | Secreted Signaling | KEGG: mmu04350 |\n",
       "| 6 | OV_1 | Endothelial       | B naive activated | Tgfb1 | TGFbR1_R2 | 0.01188053 | 0 | TGFB1_TGFBR1_TGFBR2 | Tgfb1 - (Tgfbr1+Tgfbr2) | TGFb | Secreted Signaling | KEGG: mmu04350 |\n",
       "\n"
      ],
      "text/plain": [
       "  sample source            target            ligand receptor  prob       pval\n",
       "1 OV_1   B naive activated B naive activated Tgfb1  TGFbR1_R2 0.02177701 0   \n",
       "2 OV_1   B naive resting   B naive activated Tgfb1  TGFbR1_R2 0.02292956 0   \n",
       "3 OV_1   CD4               B naive activated Tgfb1  TGFbR1_R2 0.01804501 0   \n",
       "4 OV_1   CD8               B naive activated Tgfb1  TGFbR1_R2 0.02235130 0   \n",
       "5 OV_1   Dendritic Cells   B naive activated Tgfb1  TGFbR1_R2 0.01184832 0   \n",
       "6 OV_1   Endothelial       B naive activated Tgfb1  TGFbR1_R2 0.01188053 0   \n",
       "  interaction_name    interaction_name_2      pathway_name annotation        \n",
       "1 TGFB1_TGFBR1_TGFBR2 Tgfb1 - (Tgfbr1+Tgfbr2) TGFb         Secreted Signaling\n",
       "2 TGFB1_TGFBR1_TGFBR2 Tgfb1 - (Tgfbr1+Tgfbr2) TGFb         Secreted Signaling\n",
       "3 TGFB1_TGFBR1_TGFBR2 Tgfb1 - (Tgfbr1+Tgfbr2) TGFb         Secreted Signaling\n",
       "4 TGFB1_TGFBR1_TGFBR2 Tgfb1 - (Tgfbr1+Tgfbr2) TGFb         Secreted Signaling\n",
       "5 TGFB1_TGFBR1_TGFBR2 Tgfb1 - (Tgfbr1+Tgfbr2) TGFb         Secreted Signaling\n",
       "6 TGFB1_TGFBR1_TGFBR2 Tgfb1 - (Tgfbr1+Tgfbr2) TGFb         Secreted Signaling\n",
       "  evidence      \n",
       "1 KEGG: mmu04350\n",
       "2 KEGG: mmu04350\n",
       "3 KEGG: mmu04350\n",
       "4 KEGG: mmu04350\n",
       "5 KEGG: mmu04350\n",
       "6 KEGG: mmu04350"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "\n",
    "comm_list=list()\n",
    "\n",
    "for (sample_name in names(cellchat_list)) {\n",
    "  cellchat <- cellchat_list[[sample_name]]\n",
    "  \n",
    "  # Extract all significant interactions\n",
    "  comm <- subsetCommunication(cellchat)\n",
    "  \n",
    "  # Save to CSV\n",
    "  write.csv(\n",
    "    comm,\n",
    "    file = file.path(results_folder, paste0(sample_name, \"_cellchat_all_interactions.csv\")),\n",
    "    row.names = FALSE\n",
    "  )\n",
    "  \n",
    "  # Store in a list for later combination\n",
    "  comm_list[[sample_name]] <- comm\n",
    "}\n",
    "\n",
    "# --- Optional: combine all into one dataframe ---\n",
    "comm_all <- bind_rows(comm_list, .id = \"sample\")\n",
    "\n",
    "# Quick preview\n",
    "head(comm_all)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "abe60b70",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: viridisLite\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<strong>agg_record_559351520:</strong> 2"
      ],
      "text/latex": [
       "\\textbf{agg\\textbackslash{}\\_record\\textbackslash{}\\_559351520:} 2"
      ],
      "text/markdown": [
       "**agg_record_559351520:** 2"
      ],
      "text/plain": [
       "agg_record_559351520 \n",
       "                   2 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<strong>agg_record_559351520:</strong> 2"
      ],
      "text/latex": [
       "\\textbf{agg\\textbackslash{}\\_record\\textbackslash{}\\_559351520:} 2"
      ],
      "text/markdown": [
       "**agg_record_559351520:** 2"
      ],
      "text/plain": [
       "agg_record_559351520 \n",
       "                   2 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "gene not found:ITGA4_ITGB1\n",
      "\n",
      "gene not found:ITGAV_ITGB1\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# ...existing code...\n",
    "\n",
    "# pick a CellChat/Seurat object that has expression + metadata + coords\n",
    "cc_obj <- if (exists(\"cellchat\")) cellchat else if (exists(\"cellchat_list\")) cellchat_list[[1]] else stop(\"no cellchat object found\")\n",
    "expr_mat <- as.matrix(cc_obj@data)            # genes x cells\n",
    "meta_df   <- cc_obj@meta                       # metadata; rows = cell names\n",
    "\n",
    "# detect coordinate columns\n",
    "coords_cols <- intersect(c(\"coord_x\",\"coord_y\",\"x\",\"y\"), colnames(meta_df))\n",
    "if (length(coords_cols) < 2) stop(\"no coord columns found in metadata (coord_x/coord_y or x/y)\")\n",
    "\n",
    "# helper to highlight tumour and CD8 cells\n",
    "is_tumour <- function(x) grepl(\"Tumou?r|Tumor|Cancer\", x, ignore.case = TRUE)\n",
    "is_cd8    <- function(x) grepl(\"CD8\", x, ignore.case = TRUE)\n",
    "\n",
    "meta_df <- meta_df %>%\n",
    "  mutate(cell_type_label = if (\"cell_type\" %in% colnames(meta_df)) as.character(cell_type) else as.character(ident),\n",
    "         is_tumour = is_tumour(cell_type_label),\n",
    "         is_cd8   = is_cd8(cell_type_label))\n",
    "\n",
    "# spatial plot for one gene (inline + save)\n",
    "plot_spatial_gene_pair <- function(gene, prefix = \"spatial\", save_to = NULL) {\n",
    "  if (! gene %in% rownames(expr_mat)) {\n",
    "    message(\"gene not found: \", gene); return(NULL)\n",
    "  }\n",
    "  tmp <- meta_df\n",
    "  tmp$expr <- as.numeric(expr_mat[gene, rownames(tmp)])\n",
    "  # expression heatmap\n",
    "  p_expr <- ggplot(tmp, aes_string(x = coords_cols[1], y = coords_cols[2], color = \"expr\")) +\n",
    "    geom_point(size = 0.7) + scale_color_viridis_c(option = \"viridis\") +\n",
    "    coord_fixed() + ggtitle(paste0(gene, \" expression\")) + theme_minimal() +\n",
    "    theme(axis.title = element_blank())\n",
    "  # highlight tumour vs CD8\n",
    "  tmp$label <- case_when(tmp$is_tumour ~ \"Tumour\",\n",
    "                         tmp$is_cd8   ~ \"CD8\",\n",
    "                         TRUE          ~ \"Other\")\n",
    "  p_label <- ggplot(tmp, aes_string(x = coords_cols[1], y = coords_cols[2])) +\n",
    "    geom_point(aes(color = label), size = 0.7) +\n",
    "    scale_color_manual(values = c(\"Tumour\" = \"red\", \"CD8\" = \"blue\", \"Other\" = \"grey70\")) +\n",
    "    coord_fixed() + ggtitle(\"Tumour / CD8 locations\") + theme_minimal() +\n",
    "    theme(axis.title = element_blank(), legend.position = \"right\")\n",
    "  combo <- p_expr + p_label + plot_layout(ncol = 2, widths = c(1,0.8))\n",
    "  if (!is.null(save_to)) ggsave(filename = save_to, plot = combo, width = 10, height = 5, dpi = 150)\n",
    "  return(combo)\n",
    "}\n",
    "\n",
    "# genes of interest (from your interaction list)\n",
    "genes <- c(\"Pvr\",\"Cd226\",\"Raet1e\",\"Ulbp1\",\"Nectin2\",\"Cd8a\",\"Cd8b1\",\"H2-D1\",\"Cxcl16\",\"Cxcr6\",\"Clec2d\",\"Klrb1b\",\"Klrk1\",\"Hcst\")\n",
    "\n",
    "# quick check which are present\n",
    "present <- genes[genes %in% rownames(expr_mat)]\n",
    "missing <- setdiff(genes, present)\n",
    "message(\"present: \", paste(present, collapse = \", \"))\n",
    "if (length(missing)) message(\"missing (not found in expr): \", paste(missing, collapse = \", \"))\n",
    "\n",
    "# plot and save each (files saved to results_folder)\n",
    "for (g in present) {\n",
    "  out_file <- file.path(results_folder, paste0(\"spatial_\", g, \".png\"))\n",
    "  p <- plot_spatial_gene_pair(g, save_to = out_file)\n",
    "  print(p)   # show inline in notebook\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1e2efb05",
   "metadata": {},
   "source": [
    "previous attempts below"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "57641b9b",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "adata_vis_match[[1]]@assays$RNA@data[1:5,1:5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c929785c",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "distMats <- lapply(cellTrekObs, function(ctOb){\n",
    "    obCoord <- as.matrix(ctOb@meta.data[,c(\"coord_x\", \"coord_y\")])\n",
    "    obDist <- CellChat::computeRegionDistance(obCoord,\n",
    "                                              meta = ctOb@meta.data,\n",
    "                                              contact.knn.k=10)\n",
    "    return(obDist)\n",
    "})"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "20d0c634",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "source": [
    "The above is seeing if it will work if I load Ollie's version unprocessed, still no result"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fa463a17",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "#local\n",
    "cell2loc = '/run/user/1804238067/gvfs/sftp:host=clust1-sub-1,user=lythgo02/mnt/scratchc/fmlab/lythgo02/OV_visium/emily/cell2location/cellTrek/'\n",
    "ov_visium = '/run/user/1804238067/gvfs/sftp:host=clust1-sub-1,user=lythgo02/mnt/scratchc/fmlab/lythgo02/OV_visium/emily/cell2location/cellTrek/'\n",
    "\n",
    "#### Load Visium Data\n",
    "ovVis <- readRDS(paste0(ov_visium, \"celltrek_results.rds\"))  #version of Ollie's filtered against mine\n",
    "\n",
    "#### Load scRNA-seq\n",
    "ovSc <- readRDS(\"/home/lythgo02/Documents/OV_visium/upk10_sc_400_newlyAnnot.rds\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9853df96",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "#convert updated annotation to factor with levels organised in alphabetical order for plotting purposes \n",
    "ovVis <- lapply(ovVis, function(i){\n",
    "    i$cell_type <- ifelse(grepl(\"Tumour\", i$updated_annotation), \"Tumour\",  #collapse tumour 1,2,3 subgroups into one \n",
    "                                    i$updated_annotation)  \n",
    "    i$cell_type <- factor(i$cell_type,\n",
    "                                    levels=sort(unique(i$cell_type)))                                                        \n",
    "    return(i)\n",
    "})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e3ce424f",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "\n",
    "ovVis <- lapply(ovVis, function(ovCellTrek) {\n",
    "    cell_type <- as.character(ovCellTrek@meta.data$cell_type)\n",
    "\n",
    "    #merge categories with consistently low counts across samples\n",
    "    cell_type[grepl(\"Th\", cell_type)] <- \"T-helper\"\n",
    "    cell_type[grepl(\"Cycling\", cell_type)] <- \"Plasma Cells\"\n",
    "    # Add to metadata\n",
    "    ovCellTrek@meta.data$cell_type_collapsed <- factor(cell_type)\n",
    "\n",
    "    return(ovCellTrek)\n",
    "})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ed93c0f3",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'orig.ident'</li><li>'nCount_RNA'</li><li>'nFeature_RNA'</li><li>'nCount_originalexp'</li><li>'nFeature_originalexp'</li><li>'percentMt'</li><li>'nCount_SCT'</li><li>'nFeature_SCT'</li><li>'SCT_snn_res.1'</li><li>'seurat_clusters'</li><li>'pANN_0.25_0.19_161'</li><li>'DF.classifications_0.25_0.19_161'</li><li>'doublet_classification'</li><li>'CellAssign'</li><li>'Treatment'</li><li>'Cell_Type'</li><li>'newAnnot'</li><li>'newAnnot_1'</li><li>'inferCNV_Score'</li><li>'inferCNV_Class'</li><li>'Final_Annotation'</li><li>'CopyKat'</li><li>'newAnnot_2'</li><li>'pANN_0.25_0.05_197'</li><li>'DF.classifications_0.25_0.05_197'</li><li>'pANN_0.25_0.19_227'</li><li>'DF.classifications_0.25_0.19_227'</li><li>'pANN_0.25_0.06_117'</li><li>'DF.classifications_0.25_0.06_117'</li><li>'pANN_0.25_0.28_314'</li><li>'DF.classifications_0.25_0.28_314'</li><li>'SCT_snn_res.0.5'</li><li>'ident'</li><li>'updated_annotation'</li><li>'X_indices'</li><li>'X_scvi_batch'</li><li>'X_scvi_labels'</li><li>'id_raw'</li><li>'id_new'</li><li>'coord_x'</li><li>'coord_y'</li><li>'cell_type'</li><li>'cell_type_collapsed'</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 'orig.ident'\n",
       "\\item 'nCount\\_RNA'\n",
       "\\item 'nFeature\\_RNA'\n",
       "\\item 'nCount\\_originalexp'\n",
       "\\item 'nFeature\\_originalexp'\n",
       "\\item 'percentMt'\n",
       "\\item 'nCount\\_SCT'\n",
       "\\item 'nFeature\\_SCT'\n",
       "\\item 'SCT\\_snn\\_res.1'\n",
       "\\item 'seurat\\_clusters'\n",
       "\\item 'pANN\\_0.25\\_0.19\\_161'\n",
       "\\item 'DF.classifications\\_0.25\\_0.19\\_161'\n",
       "\\item 'doublet\\_classification'\n",
       "\\item 'CellAssign'\n",
       "\\item 'Treatment'\n",
       "\\item 'Cell\\_Type'\n",
       "\\item 'newAnnot'\n",
       "\\item 'newAnnot\\_1'\n",
       "\\item 'inferCNV\\_Score'\n",
       "\\item 'inferCNV\\_Class'\n",
       "\\item 'Final\\_Annotation'\n",
       "\\item 'CopyKat'\n",
       "\\item 'newAnnot\\_2'\n",
       "\\item 'pANN\\_0.25\\_0.05\\_197'\n",
       "\\item 'DF.classifications\\_0.25\\_0.05\\_197'\n",
       "\\item 'pANN\\_0.25\\_0.19\\_227'\n",
       "\\item 'DF.classifications\\_0.25\\_0.19\\_227'\n",
       "\\item 'pANN\\_0.25\\_0.06\\_117'\n",
       "\\item 'DF.classifications\\_0.25\\_0.06\\_117'\n",
       "\\item 'pANN\\_0.25\\_0.28\\_314'\n",
       "\\item 'DF.classifications\\_0.25\\_0.28\\_314'\n",
       "\\item 'SCT\\_snn\\_res.0.5'\n",
       "\\item 'ident'\n",
       "\\item 'updated\\_annotation'\n",
       "\\item 'X\\_indices'\n",
       "\\item 'X\\_scvi\\_batch'\n",
       "\\item 'X\\_scvi\\_labels'\n",
       "\\item 'id\\_raw'\n",
       "\\item 'id\\_new'\n",
       "\\item 'coord\\_x'\n",
       "\\item 'coord\\_y'\n",
       "\\item 'cell\\_type'\n",
       "\\item 'cell\\_type\\_collapsed'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 'orig.ident'\n",
       "2. 'nCount_RNA'\n",
       "3. 'nFeature_RNA'\n",
       "4. 'nCount_originalexp'\n",
       "5. 'nFeature_originalexp'\n",
       "6. 'percentMt'\n",
       "7. 'nCount_SCT'\n",
       "8. 'nFeature_SCT'\n",
       "9. 'SCT_snn_res.1'\n",
       "10. 'seurat_clusters'\n",
       "11. 'pANN_0.25_0.19_161'\n",
       "12. 'DF.classifications_0.25_0.19_161'\n",
       "13. 'doublet_classification'\n",
       "14. 'CellAssign'\n",
       "15. 'Treatment'\n",
       "16. 'Cell_Type'\n",
       "17. 'newAnnot'\n",
       "18. 'newAnnot_1'\n",
       "19. 'inferCNV_Score'\n",
       "20. 'inferCNV_Class'\n",
       "21. 'Final_Annotation'\n",
       "22. 'CopyKat'\n",
       "23. 'newAnnot_2'\n",
       "24. 'pANN_0.25_0.05_197'\n",
       "25. 'DF.classifications_0.25_0.05_197'\n",
       "26. 'pANN_0.25_0.19_227'\n",
       "27. 'DF.classifications_0.25_0.19_227'\n",
       "28. 'pANN_0.25_0.06_117'\n",
       "29. 'DF.classifications_0.25_0.06_117'\n",
       "30. 'pANN_0.25_0.28_314'\n",
       "31. 'DF.classifications_0.25_0.28_314'\n",
       "32. 'SCT_snn_res.0.5'\n",
       "33. 'ident'\n",
       "34. 'updated_annotation'\n",
       "35. 'X_indices'\n",
       "36. 'X_scvi_batch'\n",
       "37. 'X_scvi_labels'\n",
       "38. 'id_raw'\n",
       "39. 'id_new'\n",
       "40. 'coord_x'\n",
       "41. 'coord_y'\n",
       "42. 'cell_type'\n",
       "43. 'cell_type_collapsed'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       " [1] \"orig.ident\"                       \"nCount_RNA\"                      \n",
       " [3] \"nFeature_RNA\"                     \"nCount_originalexp\"              \n",
       " [5] \"nFeature_originalexp\"             \"percentMt\"                       \n",
       " [7] \"nCount_SCT\"                       \"nFeature_SCT\"                    \n",
       " [9] \"SCT_snn_res.1\"                    \"seurat_clusters\"                 \n",
       "[11] \"pANN_0.25_0.19_161\"               \"DF.classifications_0.25_0.19_161\"\n",
       "[13] \"doublet_classification\"           \"CellAssign\"                      \n",
       "[15] \"Treatment\"                        \"Cell_Type\"                       \n",
       "[17] \"newAnnot\"                         \"newAnnot_1\"                      \n",
       "[19] \"inferCNV_Score\"                   \"inferCNV_Class\"                  \n",
       "[21] \"Final_Annotation\"                 \"CopyKat\"                         \n",
       "[23] \"newAnnot_2\"                       \"pANN_0.25_0.05_197\"              \n",
       "[25] \"DF.classifications_0.25_0.05_197\" \"pANN_0.25_0.19_227\"              \n",
       "[27] \"DF.classifications_0.25_0.19_227\" \"pANN_0.25_0.06_117\"              \n",
       "[29] \"DF.classifications_0.25_0.06_117\" \"pANN_0.25_0.28_314\"              \n",
       "[31] \"DF.classifications_0.25_0.28_314\" \"SCT_snn_res.0.5\"                 \n",
       "[33] \"ident\"                            \"updated_annotation\"              \n",
       "[35] \"X_indices\"                        \"X_scvi_batch\"                    \n",
       "[37] \"X_scvi_labels\"                    \"id_raw\"                          \n",
       "[39] \"id_new\"                           \"coord_x\"                         \n",
       "[41] \"coord_y\"                          \"cell_type\"                       \n",
       "[43] \"cell_type_collapsed\"             "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>Tumour</li><li>Macrophage</li><li>Monocytes</li><li>Th17</li><li>CD4</li><li>CD8</li><li>Mesenchymal</li><li>Dendritic Cells</li><li>Plasma Cells</li><li>B naive activated</li><li>NK_cell</li><li>Endothelial</li><li>Treg</li><li>Cycling Plasma Cells</li><li>B naive resting</li><li>Th1</li><li>Neutrophils</li></ol>\n",
       "\n",
       "<details>\n",
       "\t<summary style=display:list-item;cursor:pointer>\n",
       "\t\t<strong>Levels</strong>:\n",
       "\t</summary>\n",
       "\t<style>\n",
       "\t.list-inline {list-style: none; margin:0; padding: 0}\n",
       "\t.list-inline>li {display: inline-block}\n",
       "\t.list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "\t</style>\n",
       "\t<ol class=list-inline><li>'B naive activated'</li><li>'B naive resting'</li><li>'CD4'</li><li>'CD8'</li><li>'Cycling Plasma Cells'</li><li>'Dendritic Cells'</li><li>'Endothelial'</li><li>'Macrophage'</li><li>'Mesenchymal'</li><li>'Monocytes'</li><li>'Neutrophils'</li><li>'NK_cell'</li><li>'Plasma Cells'</li><li>'Th1'</li><li>'Th17'</li><li>'Treg'</li><li>'Tumour'</li></ol>\n",
       "</details>"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item Tumour\n",
       "\\item Macrophage\n",
       "\\item Monocytes\n",
       "\\item Th17\n",
       "\\item CD4\n",
       "\\item CD8\n",
       "\\item Mesenchymal\n",
       "\\item Dendritic Cells\n",
       "\\item Plasma Cells\n",
       "\\item B naive activated\n",
       "\\item NK\\_cell\n",
       "\\item Endothelial\n",
       "\\item Treg\n",
       "\\item Cycling Plasma Cells\n",
       "\\item B naive resting\n",
       "\\item Th1\n",
       "\\item Neutrophils\n",
       "\\end{enumerate*}\n",
       "\n",
       "\\emph{Levels}: \\begin{enumerate*}\n",
       "\\item 'B naive activated'\n",
       "\\item 'B naive resting'\n",
       "\\item 'CD4'\n",
       "\\item 'CD8'\n",
       "\\item 'Cycling Plasma Cells'\n",
       "\\item 'Dendritic Cells'\n",
       "\\item 'Endothelial'\n",
       "\\item 'Macrophage'\n",
       "\\item 'Mesenchymal'\n",
       "\\item 'Monocytes'\n",
       "\\item 'Neutrophils'\n",
       "\\item 'NK\\_cell'\n",
       "\\item 'Plasma Cells'\n",
       "\\item 'Th1'\n",
       "\\item 'Th17'\n",
       "\\item 'Treg'\n",
       "\\item 'Tumour'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. Tumour\n",
       "2. Macrophage\n",
       "3. Monocytes\n",
       "4. Th17\n",
       "5. CD4\n",
       "6. CD8\n",
       "7. Mesenchymal\n",
       "8. Dendritic Cells\n",
       "9. Plasma Cells\n",
       "10. B naive activated\n",
       "11. NK_cell\n",
       "12. Endothelial\n",
       "13. Treg\n",
       "14. Cycling Plasma Cells\n",
       "15. B naive resting\n",
       "16. Th1\n",
       "17. Neutrophils\n",
       "\n",
       "\n",
       "\n",
       "**Levels**: 1. 'B naive activated'\n",
       "2. 'B naive resting'\n",
       "3. 'CD4'\n",
       "4. 'CD8'\n",
       "5. 'Cycling Plasma Cells'\n",
       "6. 'Dendritic Cells'\n",
       "7. 'Endothelial'\n",
       "8. 'Macrophage'\n",
       "9. 'Mesenchymal'\n",
       "10. 'Monocytes'\n",
       "11. 'Neutrophils'\n",
       "12. 'NK_cell'\n",
       "13. 'Plasma Cells'\n",
       "14. 'Th1'\n",
       "15. 'Th17'\n",
       "16. 'Treg'\n",
       "17. 'Tumour'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       " [1] Tumour               Macrophage           Monocytes           \n",
       " [4] Th17                 CD4                  CD8                 \n",
       " [7] Mesenchymal          Dendritic Cells      Plasma Cells        \n",
       "[10] B naive activated    NK_cell              Endothelial         \n",
       "[13] Treg                 Cycling Plasma Cells B naive resting     \n",
       "[16] Th1                  Neutrophils         \n",
       "17 Levels: B naive activated B naive resting CD4 CD8 ... Tumour"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "   Min. 1st Qu.  Median    Mean 3rd Qu.    Max. \n",
       "   8682   11999   14394   14429   16742   20581 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "$spot\n",
       "[1] 0.06443299\n",
       "\n",
       "$fiducial\n",
       "[1] 415.707\n",
       "\n",
       "$hires\n",
       "[1] 0.06443299\n",
       "\n",
       "$lowres\n",
       "[1] 0.0193299\n",
       "\n",
       "$spot_dis\n",
       "[1] 397\n",
       "\n",
       "$spot_dis_intp\n",
       "[1] 396.0051\n",
       "\n",
       "attr(,\"class\")\n",
       "[1] \"scalefactors\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "0.01339262179412"
      ],
      "text/latex": [
       "0.01339262179412"
      ],
      "text/markdown": [
       "0.01339262179412"
      ],
      "text/plain": [
       "[1] 0.01339262"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'counts'</li><li>'data'</li><li>'scale.data'</li><li>'key'</li><li>'assay.orig'</li><li>'var.features'</li><li>'meta.features'</li><li>'misc'</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 'counts'\n",
       "\\item 'data'\n",
       "\\item 'scale.data'\n",
       "\\item 'key'\n",
       "\\item 'assay.orig'\n",
       "\\item 'var.features'\n",
       "\\item 'meta.features'\n",
       "\\item 'misc'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 'counts'\n",
       "2. 'data'\n",
       "3. 'scale.data'\n",
       "4. 'key'\n",
       "5. 'assay.orig'\n",
       "6. 'var.features'\n",
       "7. 'meta.features'\n",
       "8. 'misc'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] \"counts\"        \"data\"          \"scale.data\"    \"key\"          \n",
       "[5] \"assay.orig\"    \"var.features\"  \"meta.features\" \"misc\"         "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "ov1 <- ovVis[[1]]\n",
    "colnames(ov1@meta.data)\n",
    "unique(ov1@meta.data$cell_type) #from celltrek\n",
    "summary(ov1@meta.data$coord_x) #dimensions in pixels so will need converting \n",
    "ov1@images$slice1@scale.factors\n",
    "ov1@images$slice1@spot.radius\n",
    "slotNames(ov1[[\"RNA\"]])\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c46387c1",
   "metadata": {},
   "source": [
    "Cellchat requires normalised data\n",
    " - Ollie appears to have previously used SCTransform which is from the same authors as Cellchat (Seurat lot)\n",
    " - In their workflow they appear to use SCT data but this is the residuals after regression (like Z-scaled), given they say to use log normalised data, not scaled, will use NormaliseData function \n",
    " - I am using the output from celltrek as input for cellchat so the normalised data is in obj[[\"RNA]]@data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "24e80765",
   "metadata": {},
   "source": [
    "Prepare inputs\n",
    "CHECK CONVERSION OF PIXEL TO UM - CHCK THE LOGIC\n",
    " - extract normalised expression values \n",
    " - create metadataframe with labels and treatments \n",
    " - convert spot pixel coordinates to um for spatial.locs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5541ff3d",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Minimum NN distance (µm): 16.1544 \n",
      "Minimum NN distance (µm): 16.61242 \n",
      "Minimum NN distance (µm): 16.32836 \n",
      "Minimum NN distance (µm): 14.77146 \n",
      "Minimum NN distance (µm): 13.82296 \n",
      "Minimum NN distance (µm): 16.49941 \n"
     ]
    }
   ],
   "source": [
    "ovVis_input <- lapply(ovVis, function(x) {\n",
    "  # Extract normalized expression matrix\n",
    "  data.input <- GetAssayData(x, assay = \"RNA\", slot = \"data\")\n",
    "  \n",
    "  # Use full metadata from Seurat object directly for CellChat\n",
    "  meta.input <- x@meta.data\n",
    "  \n",
    "  # Convert pixel coordinates to microns\n",
    "  spot.size <- 65  # theoretical spot size in µm\n",
    "  pixel.diam <- x@images$slice1@scale.factors$spot_dis  # spot diameter in pixels\n",
    "  conversion.factor <- spot.size / pixel.diam           # µm per pixel\n",
    "  \n",
    "  spatial.locs <- as.matrix(meta.input[, c(\"coord_x\", \"coord_y\")]) * conversion.factor\n",
    "  colnames(spatial.locs) <- c(\"x\", \"y\")\n",
    "  \n",
    "  scale.factors <- list(\n",
    "  spot.diameter = 65,\n",
    "  spot = x@images$slice1@scale.factors$spot)  # 'spot' is the pixel diameter at full-res\n",
    "  \n",
    "  # QC: compute nearest-neighbor distances\n",
    "  d.spatial <- computeCellDistance(coordinates = spatial.locs,\n",
    "                                   ratio = 1, tol = spot.size/2)\n",
    "  cat(\"Minimum NN distance (µm):\", min(d.spatial[d.spatial != 0]), \"\\n\")\n",
    "  \n",
    "  # Return list with expression, metadata, spatial coordinates\n",
    "  list(data = data.input, meta = meta.input, spatial = spatial.locs, scale.factors=scale.factors)\n",
    "})\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dfae9b2b",
   "metadata": {},
   "source": [
    "Visium spots:\n",
    " - The center-to-center distance between spots is ~100 µm (this comes from the array design: ~65 µm spot diameter, ~100 µm spacing).\n",
    " - If you were plotting spot coordinates directly, the nearest-neighbor distances between spots should be ~100 µm.\n",
    " - CellTrek interpolates cells within each spot. Multiple cells can occupy a single spot or nearby positions so the NN distances between cells are smaller — ~10–20 µm; cell-to-cell rather than spot-to-spot \n",
    " - ~100 µm = nearest spots in Visium\n",
    " - ~14 µm = nearest mapped cells after CellTrek"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0338815a",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "head(ovVis_input$OV_1$meta)\n",
    "head(ovVis[[1]]@meta.data[, c(\"coord_x\", \"coord_y\")])  # raw pixel coords\n",
    "head(ovVis_input$OV_1$spatial)  \n",
    "head(ovVis_input$OV_1$data)\n",
    "ovVis_input$OV_1$scale.factors"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "13a59304",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "ov1 <- ovVis_input[[1]]\n",
    "colnames(ov1$meta)\n",
    "all(rownames(meta) == rownames(coordinates))\n",
    "table(meta$group)\n",
    "nrow(ov1$spatial)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "80f94a74",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAA0gAAANICAIAAAByhViMAAAACXBIWXMAABJ0AAASdAHeZh94AAAgAElEQVR4nOzdd1xV9R/H8c8FZIkgKG5Exb1X7oEjzT3Sn+aszCxLrSxXWpZamTNtl5WVabkqUytz770nIu6FKCAiyLi/PxC4smTccc73vp6P3+Pxg8M5555L9+1538+592IwGo0CAAAA/XOw9QEAAADAPCh2AAAAiqDYAQAAKIJiBwAAoAiKHQAAgCIodgAAAIqg2AEAACiCYgcAAKAIih0AAIAiKHYAAACKoNgBAAAogmIHAACgCIodAACAIih2AAAAiqDYAQAAKIJiBwAAoAiKHQAAgCIodgAAAIqg2AEAACiCYgcAAKAIih0AAIAiKHYAAACKoNgBAAAogmIHAACgCIodAACAIih2AAAAiqDYAQAAKIJiBwAAoAiKHQAAgCIodgAAAIqg2AEAACiCYgcAAKAIih0AAIAiKHYAAACKoNgBAAAogmIHAACgCIodAACAIih2AAAAiqDYAQAAKIJiBwAAoAiKHQAAgCIodgAAAIqg2AEAACiCYgcAAKAIih0AAIAiKHYAAACKoNgBAAAogmIHAACgCIodAACAIih2AAAAiqDYAQAAKIJiBwAAoAiKHQAAgCIodgAAAIqg2AEAACiCYgcAAKAIih0AAIAiKHYAAACKoNgBAAAogmIHAACgCIodAACAIih2AAAAiqDYAQAAKIJiBwAAoAiKHQAAgCIodgAAAIqg2AEAACiCYgcAAKAIih0AAIAiKHYAAACKoNgBAAAogmIHAACgCIodAACAIih2AAAAiqDYAQAAKIJiBwAAoAiKHQAAgCIodgAAAIqg2AEAACiCYgcAAKAIih0AAIAiKHYAAACKoNgBAAAogmIHAACgCIodAACAIih2AAAAiqDYAQAAKIJiBwAAoAiKHQAAgCIodgAAAIqg2AEAACiCYgcAAKAIih0AAIAiKHYAAACKoNgBAAAogmIHAACgCIodAACAIih2AAAAiqDYAQAAKIJiBwAAoAiKHQAAgCIodgAAAIqg2AEAACiCYgcAAKAIih0AAIAiKHYAAACKoNgBAAAogmIHAACgCIodAACAIih2AAAAiqDYAQAAKIJiBwAAoAiKHQAAgCIodgAAAIqg2AEAACiCYgcAAKAIih0AAIAiKHYAAACKoNgBAAAogmIHAACgCIodAACAIih2AAAAiqDYAQAAKIJiBwAAoAiKHQAAgCIodgAAAIqg2AEAACiCYgcAAKAIih0AAIAiKHYAAACKoNgBAAAogmIHAACgCIodAACAIih2AAAAiqDYAQAAKIJiBwAAoAiKHQAAgCIodgAAAIqg2AEAACiCYgcAAKAIih0AAIAiKHYAAACKoNgBAAAogmIHAACgCIodAACAIih2AAAAiqDYAQAAKIJiBwAAoAiKHQAAgCIodgAAAIqg2AEAACiCYgcAAKAIih0AAIAiKHYAAACKoNgBAAAogmIHAACgCIodAACAIih2AAAAiqDYAQAAKIJiBwAAoAiKHQAAgCIodgAAAIqg2AEAACiCYgcAAKAIih0AAIAiKHYAAACKoNgBAAAogmIHAACgCIodAACAIih2AAAAiqDYAQAAKIJiBwAAoAiKHQAAgCIodgAAAIqg2AEAACiCYgcAAKAIih0AAIAiKHYAAACKoNgBAAAogmIHAACgCIodAACAIih2AAAAiqDYAQAAKIJiBwAAoAiKHQAAgCIodgAAAIqg2AEAACiCYgcAAKAIih0AAIAiKHYAAACKoNgBAAAogmIHAACgCIodAACAIih2AAAAiqDYAQAAKIJiBwAAoAiKHQAAgCIodgAAAIqg2AEAACiCYgcAAKAIih0AAIAiKHYAAACKoNgBAAAogmIHAACgCIodAACAIih2AAAAiqDYAQAAKIJiBwAAoAiKHQAAgCIodgAAAIqg2AEAACiCYgcAAKAIih0AAIAiKHYAAACKoNgBAAAogmIHAACgCIodAACAIih2AAAAiqDYAQAAKIJiBwAAoAiKHQAAgCIodgAAAIqg2AEAACiCYgcAAKAIih0AAIAiKHYAAACKoNgBAAAogmIHAACgCIodAACAIih2AAAAiqDYAQAAKIJiBwAAoAgnWx+AnYodOzL7Kx90npf9lRtNyfnRAPqUoxy5duiT/ZWNgU1zfjiALpEjxVDsrCFHscmjXZPSLqHqQQ3WzJFh0/Y0SzhFQQ3kSHkUO4uwZnIeK03Vo+dBLzSVozSnKM5P0AtyZG8odmajqfBkwbTnUfKgNXrJken5iZMTtIYc2TOKXV7pJT8ZSil5NDzYlq5zlHJy4swE2yJHEIpdruk6P+nR8GATiuWIMxNsghzBFMUuZxTLT3o0PFiB8jnizAQrIEfIEMUuW5TPT3pJDY96BzOywxwlnZk4LcGM7DZHQpSyh2L3GHYYIVMM8GAWdp4jTkswCzvPkfBMKXsodpkiQqYY4CF3yJEpTkvIHXJkihxljWKXASKUGeodso8cZYbTErKPHGWGHGWGYvcIIpQd1DtkjRxlB6clZI0cZQc5So9i9xARyinqHdIjRznFaQnpkaOcIkemKHZEKE+od0hCjvKC0xKSkKO8IEdJ7LrYESFzod7ZM3JkLpyW7Bk5MhdyZKfFjghZAvXO3pAjS+C0ZG/IkSXYc44cbH0ANkCKLCrlo++gNnJkUaZ/HB0KI0cWZZ85sq+JHRGyDkZ3aiNH1mHPIwd7QI6sww5zZEcTO1JkZYzulESOrMw+Rw7KI0dWZlc5souJHRGyFUZ3KiFHtmKHIweFkSNbsZ8cqT+xI0U2x+hOAeTI5uxq5KAqcmRz9pAjlSd2REg7GN3pFznSDvsZOaiHHGmH8jlSdmJHijSI0Z3ukCMNsoeRg2LIkQYpnCM1ix0p0iy6nY6QI81S+JykHnKkWarmSLVLsURI+7gsq33kSPsMm7YrfC1JDeRI+5S8LKvUxI4U6QijO80iR3ph2LRd1ZGDAsiRjiiWI3WKHSnSHbqdBpEj3VHsnKQGcqQ7KuVIkWJHinSKbqcp5EinVDonKYAc6ZQyOVKh2JEiXaPbaQQ50jVlzkl6R450TY0c6b7YkSIF0O1sjhwpQI1zkq6RIwUokCN9FztSpAy6nQ2RI2UocE7SL3KkDL3nSMfFjhQphm5nE+RIMXo/J+kUOVKMrnOk12JHipREt7MycqQkXZ+T9IgcKUm/OdJlsSNFCqPbWQ05Uph+z0m6Q44UptMc6a/YkSLl0e2sgBwpT6fnJH0hR8rTY450VuxIkZ2g21kUObITejwn6Qg5shO6y5Geih0psit0OwshR3ZFd+ckvSBHdkVfOdJNsSNFdohuZ3bkyA7p65ykC+TIDukoR/oodqTIbtHtzIgc2S0dnZO0jxzZLb3kSAfFjhTZObqdWZAjO6eXc5LGkSM7p4scab3YkSII3S7PyBFEJ+ckLSNHED3kSOvFDgAAANmk6WLH0yOkYGiXa+QIKbQ/bNAscoQUGs+RdosdKUIadLtcIEdIQ+PnJG0iR0hDyznSaLEjRcgQ3S5HyBEypOVzkgaRI2RIsznSaLEDAABATmmx2PH0CFlgaJdN5AhZ0OywQWvIEbKgzRxprtiRIjwW3e6xyBEeS5vnJE0hR3gsDeZIW8WOFCGb6HZZIEfIJg2ek7SDHCGbtJYjbRU7AAAA5JqGih1Pj5AjDO0yRI6QI1obNmgEOUKOaCpHWil2pAi5QLdLgxwhFzR1TtICcoRc0E6OtFLsAAAAkEeaKHY8PUKuMbRLQY6Qa9oZNtgcOUKuaSRHmih2AAAAyDvbFzueHiGPGNoJOUKeaWTYYFvkCHmkhRzZuNiRIpiFnXc7cgSz0MI5yYbIEczC5jmy/cQOAAAAZmHLYsfTI5iR3Q7tyBHMyObDBlshRzAj2+aIiR0AAIAibFbseHoEs7PDoR05gtnZ4dCOHMHsbJgjJ1vdMPTAGHb+00VHN16MKRBQcdTAWnU9RETibwTP+fHwtrB8lZrWe7tLCS+T9WN2rO+w6KrJArceo3uNLJfBnhPjE41ODo4WPXoAAOyNbSZ2PD3ShbhLY4Yt/8VQ9pX/VS1/dkuriQcvicj9c68MXb21ZM0xfcrGr/yt28oI0y3yBVR5s1/jh//rWuTuBUPRQiLx4au/+6vHm8tHLLscmrRe5MnB4/aGWOKY7WpoR45gIXY1tCNHsBBb5YiJHTJ17MTPzvX/HVG9hkibctHru538+16dwTsOfOdV50i/8lVEGg49X+yzIyd6NK+avIVj0RKdiiZ9Gb911sYiw/v28ZKQRb+/F93i+5H5D36/4mWfF5a1dtix4GDR5/qUt9X9AgBAVTaY2PH0SC9KVJ31Ro3KSV8nJMYV8CrtJrdv3zOWKFhaREScShcqF3LzWGIGmxrP7hh1qtrcjgVEpHjt8gX27pj6w7YvzhfuUCmf8eLuaffqTaxmscuwdjK0I0ewKDsZ2pEjWJRNcsTEDpkq6tevqIhIQljIpAk7o57u3dJBXCsVK/b1qd+javT3SDz9z4l9ic43I0UKptkyZtmCo1WfHVpRRERcqzVb91mNI5ceFC3nW9zl3s9vX+44sknaLQAAQN5Ze2LH0yOdiTu5dk2DgasOtOq9ZWhJVxGp3fy7rpFv9P6i/sDPe57OX9uQL79ruo0u7v/obOVRTZ1TFjh4eNWu4lvcRe7v2fJTmRbDCtz4Yd6KrmNXT98XldG8L8+UH9qRI1iB8kM7cgQrsH6O+Bw7ZO7Bjs8XNv8pcdQXL/89oGTyu1/d240acunXfr/OefHIG2VdPQqUTFfsjq45fLddjSfS7y8xdOYvia8NKLr7i+XfudedPLDEwRl/LYxIvx4AAMgVih0ydW7Xi6sKff1N50H++VIX3jn+3Ev/bXb2CijsePz3IxfaVQtMu9nNpRsSujQrkn5/1/7cdKhtyw5ucUGXHFu1LVO3erVu/hFnblrwHgAAYF+s+ho75t66cmNvyPHEezNHL5zzcEHxN+e06+Zd6blqW5/ue7FGsZjghMo/zyzjLCJydXyvHy+MGPdLS5HQ8xtvlBxVId3uos++s7n4e7M8RKRb79Ktx/18rNT944n1/wiwzNHvmiSNplhm17ZGjmA1hk3bjYFNbX0UFkGOYDVWzhFvnkBmCrR4clPFOJMFbhVdRMSpxYihp7uHnjd6VS/tlvwyusJDJz8TUypprfLTv6hYM/0D60HB4ePKVHcQEfFu2mFnlVvHw/JVCPDyZGgMAIC5GIxGo3VuiadHuXbQeZ6tD0Gv1BvakaNcc+3Qx9aHoFfqDe3IUa6Ro1yzWo4YlwAAACiCYgcAAKAIKxU75t6wCcU+0I4cwSYU+0A7cgSbsFqOmNgBAAAogmIHAACgCGsUO+besCFlrsaSI9iQMldjyRFsyDo5YmIHAACgCIodAGiKUXb8KCdiHlkWtks2Xshqo7AzcjMuqxUA2AeLFzvm3rA5Ba7GkiM7cnmZTPlagmNTlzy4Kd/Plf3Xk7+/J4vHycC+MmS0HIwQEZEw+eQHccyXbl/mpMDVWHIEm7NCjpjYAYBmJITI9HXi55y65Ni3MnCobLyeuiTyP/k7QH5cIlPry9d/iYjs/lbKD5FC1j5YABpEsQMAjYiTHz6URmMlwDF1WfUX5Nc/5PmqqUs820i7YBncXybtkxc6SsIZWeYkfQKsf7gANCj932oHANjC0a/kSGuZEyAzsl7PQ/p/JP2Tv1s5XdqPFRdLHxwAfbDsxI4XNEAjdP0yO3JkF+4flLlBMq5Pzv5Vjtoiu2tImwSZP1z69JRRn8odSx2g6PxlduQIGmHpHDGxAwANuHlCbp6WFzuIiMTcl43/k5OTZEyzLLeJkx9XysCP5cincqOLLG4na8bKkuPycjWrHDEsJTr86nGHYk94OoiI3Dzca0Nw5MOfFBzSuXUfj4y2MSY+EAdng9WOEdpFsQMADfDvL6uSL69Of1Kq/iZdvB6zyZXlEtFVquWTwyIODmIwiJOTJCRY+khhWQ8uj/l15bmGQ9fUdhWR6NBLW5z8FtbxFRER50ouIiI3Lh2atv/SDbfiLzap16aAQSRh8/pV66t3fb8Yr5sHb54AAF2KkAUH5LlWIiK1Bkjh36VfH1lbQvrVtPWBIS9iV6/b+ldC6qn5fHhkuVIVOgSU6RBQpkNAiXL5RGJODVsXFti0+Tvlo99effS6SGLYwZlxtd6i1UFERAxGo9FCu+YFDeZy0HmerQ9BEY2m2PoIco4cmYtrhz62PgRFGAOb2voQckwvObp56u9OQRVGOq5bXOLZpInd6tVfTpNa1R/cinAp1KVe/QFFncV456sV/yxxKuQXHXY7oM3KBvkX/b7R2LbTcxleojU3cmQulssRl2IBANCAuydf3Ok09ZmAxA3rkhdFh0Qk3M7v0rV+XcfQY6MXr7r73NMve3kPe/rpLqF3brkWrFHA+V7IxuW+Df+wSquDLlDsAACwNWPEF2sO+rfu1d5V1qYudX+13yuvJn3p53PhzIKfTkW83NBLJF8J3yIlRMR4a8ZeeaOH97nTu6Ydv+VQpMqkxgFlHDO8AdgLLskDAGBrYae/vBp3YOdf7ZasmBQSf3DfXx12XBHjg2vh0fceruFW0cc16sEjfxH40qHdwdUatoo80H93bK8mT3S6t2fQ/ggbHDy0hIkdAAC2VrD6b/3LJf2F4G3bli72bTy9uq8YQj/4eW30U4MWlHeWhCsrg6VpB+/UTWJDJgcXfbeXuzE4IrJ45Q7FiibGFJt4KkLkce+nhtIsVez08kpV2JVdk3T2/glyBA0ybNqur/dP6CNHTu6VirgnfXnF1VCgQKGans4iJSd0rNRz9cK6hX0LRN6Jr9T299TrrIl7dh72adSlgoj4V++7c12rlT6JYfcHdmphozuAnLFcjpjYAQCgIQ0bdS3l7Jz0dfFyzXe8XP9s2N1ED+8K+fOZvHwqvlCVVu8UdRQRcSr2br/e/UPvOngVKufKK6zsHcUOAAAN8fEp4WPyrcHJrUJRt3RrOQcUdU79zsG1fFFXyx8adIBqDwAAoAiKHQAAgCJ0eCnWeH9/UPC28HhvX/+eZb09RCTq0rzD1yJNVikXULdfMdO79uB0SPC6sITixf07lyzgYuUDBgAAsArdTexilq/6pfe+0MgHkcv/W9p009UYEUlMuBcXl/y/qA0HD2+ONt3kwX///frU9qvhcZFL1i7reSg8aWnErZCv9x766UJk7MPV7i/fsGPdA+veGwAAAPPR28Tu/tkvz5aYNapVD0dJLGsIWHF6T2CJFp5lxgeWSfr5vYubAsOavV/O5H5FHp98yGPS8DbPu4uUM1RZfnhnrZYNb+7utC6qd/UiNw+s6RvReWVNj+iLe753qvK7c4a3CgAAoAMWmdhZ8EODHEuO79mwtaOISJzR6Oju5mv608TQjzbcfv7JKkVNF94OP+NdvFHSxwMVLdUi7sqWOyJGEYPB0dEhn4PBaDSK8fbs3fEjGxXRW89Fju2aZOsjyDZ9fPgW7JJh03ZbH0J2kSNoloVypLdLsc7ercv4eN0Lfv23ZbVXXu3XoW4Vkx/ePL5rRYknnvd8dBOvAv63r+28LyIiNy5vjb1/6744FGv4V7sy+WLiitfusKRWgWtHdx2r1LAd4zoAAKBn+hxRuRZ7rmFN/xMHP9t4vE//OtUeLg2du+vekN5+ad8b4V19YrVjr/7296nSLscuXjW6itEoIlLQt9ywpHHfgwvvnS78Ti/3q1dOL78aVyGg0lM++ax4ZwAAAMxDZxO7xKiba69HGR3z1/Sv+Fq7Ri1unfzr9sMfxV84vsi9ct+C6Tdy7dax37qW5ar7ln6/Z8sOTu5F85v+1Hhg1yHXBnX9zm/qtOFyrNydv3LV1/wNZQAAoEM6K3YOd06+sPro8aRvYu9eepDP4+H1U+OGk0FlK5UvkX4b462FG06Elag4uHpA3ehL/zqVbmta/iKPTQuvNNHf6VjIlTqNW735ROMZlWP+uRxv8XsCAABgbnq7FFuy5iSP33v8EtXHz+VM0JkL1Vo/45H0g9CNF5ya1PIwWTVh3dbVfxRq9WlVnwrGNQOXhz/r77jz+MUWrXvUMaSs82Dl1pBmzboWFnEuU+Lg9s2fRLhvOOXy1NN6+7UAAADor9g5eL/0dO+6IZd233lQKbDngnI+BZKWJxhqVG9SrYjpqgbfQsVqFnAScWjSutfiU0E77ruO7tmgTSGT1+DFhd8v1+QVHxERz3KtVjufWX4t7qXuXTtkcD0XAABA6/RW7ETEyaNBhSoN0ix09O3X1PfRRQ61qzaonfSlwb1BlVppNxGRfEX6pb6r1qFEqcojSpn3WAEAAKxHZ6+xAwAAQGZ0OLEDoEUJ+4/vX/vwL/aJU5EK4yp4i4jER245c/5gjGutsmUDvfkgIQCwLCZ2AMwiYs2uY1vvJf/V5nijiEjCzamLlo0Ojr5/7/yYxSumX08QEZH44PMn5u099ndY8t9mjrs8ef2ZG7Y6cABQCBM7rZvaaP5U7z5JX+9c39S2BwNkLvLc3WJD2zTt5Zi6KOLUnumJNQ50eaKCSH/XpfX3nh/RJeD6gb8GXS75bGnHX9f8caxD7zcLGw/v3h8W0LFo5rs2i5i1v7p26GPhGwEAG6PYadrURvNNv23cJvXvylHyoC1RESEuLlHnjs0Jl9KlynYvnt9RJCTsjmfJWhVERMSvdCmfvVeOSICviBgcnBwcnAySaBS5e3xqWMXPm3GVFtA6nhrpApditStNq0ujcZvtSf+z2vEAWYmIPH835PugqPuxNz9bubjroQijiJ+X550r14JEROTSpSvno++HigTU7bywumtkrFPPDt3G+D5YtTW4YbPKvo/Zu3nErP3VKrcDADbDxE6jsm51plK6HTM82JJP9Z+frdOgiIezyGu+D0qtO7i9dmCzqvVH7lvTe8X9pzzvbr4UXUQ8jCIiTuXLVhtVVkQk7tqOL53rrygUe+D06a338reqWr6mq2UPkwuyANRGsdOi7Lc6U1yohS25eTdze/ile+lS9aJDLsWL5Cv50aA+nc5dCTGWH1bjWu/l0cUe2Sbyix13n+tY/MiW314ODxhY+MrgX8N+HNSwhiH93gHYGM+I9IJLsZqTu1aXBhdqYWUhB/6ouTY46dNOYi5fPVjQp4qTSPjZSXtvV6pUZVDlkg4Xz98s61/TZJPbp3Zt9G/Uy+3u+hDXEe2fGNWk1TCXCxvuWvxQuSALQGFM7LTFLK3OFBdqYR1lK1SttXdz57+ud/aJ33z0WtdWvWuLSMFC7kHLe96p1tHj9tIzbh/19k+90JpwbeqR/BN6eYkktikT88q6/fcKh34XU/rbAra7DwCgfwaj0Wj2ncaOHWn2fdqDzFrdVO9t5r0hOy95jabY+giyR3c5Msbc/jf40skY58r+5Z4q7JK88NafJy9dcvJuEeBf0z31Imt8WPCS2FIDSiStFrP/1Omt0flbVS1fy8KvsUvBdaU8Mgbq458R3eVIm8iLhVgiR0zstMLss7osMMaDJRhcfdpX82mfdmHhbnUKp1/ZqVDAgNTvXOtVrlXPskeXFu+iAKAkip0mWLPVmaLhAQCyxlMgfbHImydcps+zxG7t1sQ7zSx9E/bzZgu9XIcVcmR5vIsi1/RyHVbIETTMQjliYmd7thrXZYjPTIFd4YIsAMVQ7GxMU60uDS7UAoCd45mP7lDsbEnLrc4UDQ8KY2gHQCUUO5vRS6szxYVaKIluB0AZFDvb0GOrS4MxHgCojSc8ekSxswEFWp0pGh4UwNAOsL3oKxIUmvyNm1SpJM6ZLETmKHbWplirM8WFWgBA7h36XqYfk+IeIiLiJ1PeFd9MFkqMnLsgBUqLr9vDbY//Ja7tJSCfbY5cSyzyJ8WS8Idc0st1qzP7XxWzGk01PB19iF0KcmRNDO2yQ0cfYpeCHOWCDeKwfLiEviovVX3MQuMleftdcSotdy5Ii7eld3lJOC3jVsvUN8TFykece5bLkUU+oBgZUnhWlwX7+ehjKICPLAZs6do1KeopZ0/K9aisFkYdl2uVZMiLMrCubD8oIvLHd/LU8zpqdRbFpVgrsc9WZ4oLtQCAzD2Qa7dk4zgp7SNXzkqNUTKxvRgyWligvYwMlwVzxbO6TOgpdzfJntryYUFbH79WUOysgVaXBu+3gGbxLgpAbPOyhAfS+AUZ3Esq5pfYfdJ/nGxuLoGJGS10lzp9pU5fERGJk8/+kEEfS8JdCbkmxQPEw9HqR64tXIq1OFpdFrhQCw3igixgCx7SebBUzC8i4lJXGrnImYuZLDRxealEdpeAIBk1VL77Rl4eIWfibHDsWmLBYsefXhZaXbalNDzLlTw9vnNCyBE0Ro/vnBBypAsJx+XDbyQ86Zu7cj1KChfOeGGqcFlwWIa0lAMrpcIY+XCGjPSXVQetfug5ZtEccSnWgmh1ucOFWtgcF2Rhz2zz4HcsIdEr5WNn6V1Xrvwp55rJpMLi6JjBwhQHFkipIVJE5H4pObVVggvIztNS/BkbHLyWcCnWUmh1eceFWtgQF2QB6/KWdxdItTBZ9pOcKSWfTRavzBYmiZC9BaVfRRER/37S102++1Y8Bsv/StvqDmgEEzuLoNWZF++oBQDrsOWs2qm49H8jWwtFRLxk2JDkr/NJ4IsSaMFD0xHLTuzs82UNtDqLyt0YT6cvsEtinznSAoZ2aej0BXZJyBE0wtI5YmJnZrQ6q+GleAAApEGxMydanU1woRaWw7soYFd4tCuAYmc2tDotYIwHs6PbAdARi78rlpc1mMXEO81sfQg6Y/pSPF2/wC4JOYLN6foFdknIEWzOCjni407Mg3GdZvFpKcg73kUBe8BkWg0UOzOg1QHKo9sB0AVrFDu1p9+0Oo1T4PpRErVzBI0jR0DeWSdHvHkiT1JaXdQ5ifMT73wPlyfekF1/yeXrkr+CNOoqhVwf2Sr+quz4Xa7dk2JNpG+vJa8AACAASURBVGnTTP8bGONFnMRgsYMHkCO8iwIK47GtDC7F5l7qrC5GtnwkQXeTfxAtq8bLBQ9p0FGcj8hPX0qM6WYRsmKC3CkvTdpL5G/y504REYmV44vktw9l/QZ5kLTaLVn6idy3zj0BkD1ckAWgcVYqdupNv1NaXdBP8u0rcuBW6o+Mh+V4CenWR8rUkSdfFJddcsmY+tOo7RLkL0+2E7+a0q6nBK2VeyJXf5NDrtJmkBTYJRuOiYic+kX8+oq7Ne+SipS5fpREvRxBF8gRkHdWyxETu9wwfV1dqaekx/tSxS31p4llpNuA5E6WIIk+4mlyPfVeuLgUlaTVXUqI+0UJE/GpIXG7ZOsyORYp5UpJwhk54CENSlrjvgDIEYZ2UA/XYVXCa+xyLM27Jdx8xU3E1aQhOxaXasVFRBJuyt9zxbOTFDVZ3ydAEr+R87FS1kUubpaIu3LPKK61ZVB5uXlDPP3F3Ul2fSZ1R4mjVe4OgJzixXYANMt6Ezs1pt/Zfw/s1TXy9RiJ7STPdHpkeb4G0qmWrHhZFoySf8PEJ584G0REHDykWIC4O0n0VrlQWyobZc/X8usM2XfG3PfBbih2/SiJGjmCjpAjIO+smSMmdjmQ/VYX9LmsuiBdZkgF33Q/M0i1EVIpQqJECobJpxelgOlP42TLBmkxUU59IucrSstmsuET8Z4rAfnS7QeA7TC0gzJ4JCvGqq+x0/WTpOy3usRjsvqo/G9qRq1ORK7Lr5MlzEMKeknIOnFqLkVMfnjrT4nvKMUd5fZ1CWguxapKgJeE3c1oP8iSkmOGJLrOEfSFHAF5Z+UcMbHLlhx9CnHoUbkXK+snJ3/vIK2mSGmRVQPEZYK0qyq1CsmiEVLEUyLc5enBJluGy6az0v5pEZHqneWXdyXER8J8pb+Pue4HALNhaAcF8BhWD8Xu8bLT6hq/Lfk8H37tFSgDapr8zCBJk7smE8RQRkSk8gjxuyRRzuJb9JGRaaJRWgx7eGXWs4UMqS637kphf+EyLKBNdDsAWmMwGo2PX8usYseOtPIt5oV2/mLYVO9ttj4E3VD4+lEKfeVIYQoXO3JkDxR+AGuE9XPE59hlRTutDoA28bF20C9anZJsUOz08pJVWp1O2cOYQfSTI3ugZLcjR0De2SRHTOwyRqsDAAC6Y5tip/EnSbQ6/bKTMUMSjefIrig2tCNH9oDrsJZmqxwxsUuLVgcgFxTrdgB0ymbFTptPktRtdYkSek7O35LM3gMdFy5BQXIv0aoHZW52NWZIos0cQdfIEZB3NswRE7tUGm91E+80y/a60TL3AwlP+TZM5j4vo2bI7Ffk+Y8lLM3KRtkzX/q/JPOmSv+XJDi52929IkEXJT5lrSvy5568HD+gPIZ20AuuwyrMlsWOJ0kWkChXT8jSD+XPC6mdbM83sq2q/PCFzPtZ6u+XJUcf2SJ2h0w/IXN+lvkLZZS/rNolIrLvUxk9UxbPlxEfPSyI674WY2lr3pOcssMxQxJyBDMiR0De2TZHTOwe0vi4Ltvuyprf5FS4OJosO35MGrUUZxHJJ61byPZHP+h4zzqp0FlKxMvNMGn5trzWRERk/25pNVReGCo+hyTEKLH7ZV1J6VzMincE0CWGdgBsy8bFTiNPklRpdSLiJS9MlknPiqvJsvz55erVh19fuSy3H70WG3pLorbJoEEycbj0GSknokVEnp0iUcvk00XS7H2pY5Qlv0jvgY+URY2x2zFDEo3kCEn02+3Ika0PwUq4DmtRNs+R7Sd2Ns+SQq0uE4Gd5MwSWbpJNi+RJQflQcwjb6G4Hy2n78qsJfL1YuknMutXERGXMjL0HfngPelQUW6tlnOtpIGbhAVLSNoX6GmBzVOkBTbPEfSOHAk5Qp5pIUe2L3a2pX6rE5EiXWX+ULmyTQ5Gywt9xMtbDCY/9fKS2u2kiIOIg7RoISEn5YHpxtHy3SZ5vqOsf1dGz5P5I+WDjVY+fEB39Du0A6B3mih2tnqSZBetTkTCTkh4ZXltorz2vNw/JxUrPfLTsuXk8qWHX1+/Jm7e4mzy0zM/iWs/8b8hy8Jkzicye45c+TXd+2ptSQtPjzSCYYOm6KvbkaMUyueI67CWo5EcaaLY2YS9tDoR8bgnH02QtXtlz1L5Llj6PSkiIlfk16VyU6RaZ/H4TxZukaNbZME/0rWLyZbXZeFVGVxPxFuK35JNx+X4ZokoIR62uR8AACBrWil2Vn6SpHqrKyA1Kkq+5O9cnpB3e8iepbL+hrz1qdRyERGRcNl3QO6KSFmZ8a5EbpIlG6TpFHmxeupugrZIk6HiJSKuMuptOf+L/HpRJrwmLla+O5nSyNMj7VB+2KAvehnakaM0FM4R4zrL0U6ODEZjZn+LwAZix460wq3ot9VN9d72+JXshnZSpDXWyRGySeOnUnKUGSVzpPFHo35pKkdamdhZjX5bHQAAQNa0VewsPQCn1SlDU0+PtEbhC0l6pOULsuQoC+rliHGdhWgtR9oqdmLJLNHqlKG1FGmQeuckmB05eixyhMfSYI40V+zEMlmi1SlDgynSJs5J2qHBoR05yiZyhCxoM0daLHZmR6sDYFsa7HawK1yHtR8aLXZmfJJEq1OJNp8eaRbDBmSIHOUIOUKGNJsjjRY7MVOWaHUq0WyKtIxzknZoZGhHjnKBHCENLedIu8VO8pwlWp1KtJwijeOcpB0273bkKNd0nSOuw5qXxnOk6WKXF7Q6AABgb7Re7HL3JIlWpxiNPz3SPl0PGxRjw6EdOcojcgTRQ460Xuwk51mi1SlG+ynSBc5J2mGTbkeOzEKPOeI6rBnpIkc6KHaSkyyp3eom3mlm60OwNl2kSC/0eE6CWZAjMyJHdksvOdJHsROyZJf0kiIdIUcaYc2hHTkyO3Jkh3SUI90UO8lGltQe19kbHaVIXzgn2RVyZCF6yRHXYc1CXznSU7GTLLNEq1OJvlKkO3o5J6nNCkM7cmRR5MhO6C5HOit2kkmWaHUq0V2K9IhzkhZYtNuRIyvQeI4Y1+WdHnOkv2In6bJEq1OJHlOkUxo/JyEvyJHVkCOF6TRHuix2YpIlWp1KdJoi/eKcZHOWGNqRIysjR0rSb44MRqPR1seQe5NWGmx9CNY21XubrQ/BUvSbIr2LHTvS1odg78x4yYwc2YrWcsR12LzQdY70OrFLMqWHjlspTOk6RXrHvEEZ5MiGyJEy9J4jfRc7odspQe8pUgDnJNsyywVZcmRz5EgBCuRI98VO6HY6p0CK1MA5ybby2O3IkUZoJEdch80dNXKkQrETup1uqZEiZWjknIScIkeaQo50SpkcKVLshG6nQ8qkSCWck2wod0M7cqRB5Eh3VMqROsVO6Ha6olKKFMM5yYZy2u3IkWbZMEdch80pxXKk7487yYzCH4OiwMedKBYhhWnt4xvsRDbPyuRIL6yfI4pd9imZI6UmdikY3WmWkilSFaM7m8jO0I4c6Qg50ixVc6RmsRO6nSapmiKFcU7SIHKkO9bMEeO6bFI4R2peijWl2GVZnV6KVThCdoLLslaW4emZHOmdFXJEsXss5XOk7MQuBaM7m1M+RfaA0Z2Vpb8gS44UQI5szh5ypP7ELoUaozt9TezsIUL2htGd1aSMXsiReiyUI8Z1WbCfHKk/sUvB6M7K7CdFdoWRg9UkDe3IkZLIkZXZVY7saGKXQu+jO+0P7ewqQnaL0Z2lce63B2bMEeO6DNnh+ciOJnYpGN1ZlB2myD5ROyyKX6+d4D+0Rdnn+cgeJ3YpdDq60+zEzj4jBEZ35sWZ3j7lPUdM7EzZ8/nIrotdEt3VOw0WO3uOEJJQ7/KOSodc54hWl4LzEcXuIR3VO00VOyIEU9S73KHSwVQuckSxE85HySh2j9BFvdNIsSNCyAz1LvuodMhMjnJk58WO85Epil0GNF7vbF7siBCyg3qXNSodsiM7ObLnVsf5KD2KXaY0W+9sWOyIEHKKepcelQ45lXWO7LPYcT7KDMXuMTRY72xS7IgQ8oJ6l4RKh7zILEf2Vuw4H2WNYpctmqp3Vi52RAjmYs/1jkoHc0mTI7tqdZyPsoNilzNaaHjWKXbkB5ZjPw2PPgfLScqRPRQ7zkc5QrHLJRs2PIsWO/IDa1K14dHnYE2GTdttfQgWwfkodyh2eWX9hmeJYkd+YFtqNDz6HGxLjYbH+SiPKHZmY7WGZ8ZiR36gNXpsePQ5aI0eGx7nI3Oh2FmERUteHosd4YFeaLnkUeagF1oueZyPLIFiZw3m7Xk5LXYkB2qwbc+jyUENtu15nI+sgGJnG3mpelkXO2ID+2G5qkeNg/2wXNXjfGQTFDsdSGqBU3rwXwrIpdixI+lqQB4lVUDqmsZR7AAAABThYOsDAAAAgHlQ7AAAABRBsQMAAFAExQ4AAEARFDsAAABFUOwAAAAUQbEDAABQBMUOAABAERQ7AAAARVDsAAAAFEGxAwAAUATFDgAAQBEUOwAAAEVQ7AAAABRBsQMAAFAExQ4AAEARFDsAAABFUOwAAAAUQbEDAABQBMUOAABAERQ7AAAARVDsAAAAFEGxAwAAUATFDgAAQBEUOwAAAEVQ7AAAABRBsQMAAFAExQ4AAEARFDsAAABFUOwAAAAUQbEDAABQBMUOAABAERQ7AAAARVDsAAAAFEGxAwAAUATFDgAAQBEUOwAAAEVQ7AAAABRBsQMAAFAExQ4AAEARFDsAAABFUOwAAAAUQbEDAABQBMUOAABAERQ7AAAARVDsAAAAFEGxAwAAUATFDgAAQBEUOwAAAEVQ7AAAABRBsQMAAFAExQ4AAEARFDsAAABFUOwAAAAUQbEDAABQBMUOAABAERQ7AAAARVDsAAAAFEGxAwAAUATFDgAAQBEUOwAAAEVQ7AAAABRBsQMAAFCEk60PIK3YsSPNu0OX6fPMu0MAAABtYmIHAACgCIodAACAIih2AAAAiqDYAQAAKIJiBwAAoAiKHQAAgCIodgAAAIqg2ImIRJ5c+fEbL/Tt2rnXi+M+/nbV8XBbH1Bae6d37DLnQAY/yPTI4/6b0GrAd8HmO4TEjZNa/e/LkxbYMwAAMBOKnYStGVqrxqAv9ob71g9sVCpm97xBjZ8ct82C3e7WD30r9F5wJSeb3D61dduZO2mXZnXkxrj7UfdiE8x0yCJivHls087gSAvsGQAAmInm/vKE1QV/+fa3cc/9ffqb9vmTFowdPLZu3afHP3n9izYGi9xiQsSVs5fD4/O8nyyP3LnDnL0d8nwTGbHcngEAQN5Q7M6dOyeuT3q7pyxwqTP0g2nOEQXjRJxFRKKOLpn7zdp9weH5K7YYOOrVp8q4iIiIMXTrV7N/2nA8rECl5v979dX2/k6yZXKrnyotnOi1aNrc0A6rZvdwyWDbG4tf6vnpUbl5rW+rm5PWTu/omtn+Y8/+Pmf+8l0XDGUb9xlRNcdHHvffhHY/lP/25+cDRBKvb/xs5g8bzxnKtR4yzG/FizueWju9vev2KW0XlJ33cuTnc9edC3co8US/t8Y8XcldROTBxb8/n7ds56kr9/OXrttl+GsDahU0udnUPWe+h4xv0az/3QAAQDpciq3Xpo1X8GcDn37769X7L0UlioiU7zZhyqB6ziIi0RvfqN/4jX9jA1q2b1biwoI+zQf+cilRRMJWDKr75Ad7XWs3r+6wf073+t2/Py8SenzTpp9ff/bjc6W6PFkjX8bbFmz24ptPlZZS7V6bPKCuc2b7N5755Ml6veYf8qjbuGLC+ldbvPxvbA6PPPHG0U27Q6JEJOKvoQ07zjpauGGzivd+69t86MJNR64niMitE5tXvd9zwK9Sr2OnJ5z3ze3Vdtz2RBG5tqB7jS5zD+ar1LJNncJB3w9u8dyi26Y3m7rnzPaQ2S0CAADLYmLn0+/HrQlTpny26M2uHwxz9K7QsNVT3QcPf7Fr5QIicvbT0Z+4jju2YWIVJxEZOaROV/8Xxqzssdh/xuhfikw+8e+4Sg4ir7X1qNz6s8XnnqsocvZYiW+D5rd0FpGzH2e87dNNynvJvoBGLWsUy2ydBU5T3t1VZfqJDaPLO4qM6V+uZaXXj+bsyFOc+nT8D15vH107saqDyIhm+cq3nJ/ys1uxbZeu+zTQWWRIvajtjTftvChN/UNC87V5b+HPE1q6i8iL5S56j9lxXPo3z/i3l9EeYrK4RQAAYEEUOxGPGgOn/zZwenx48L7N6/9ds+ynd7p9s3Dcuj0fNjPs23PE8ZZhZIfND1e9d854v+CZy1ev7Dhfo0evSknzznzN5wTHiYgsF3Fu+WQLZxERic1kWxHPlFvObJ1Td3dHNB4xqLyjiIgYyg7s3/T1cTk68pTLnpE7th8r0+HHqg8PtUXn9gXmX0r+oWtg+5ZJR+tQoUKArE9IEDE0GffHipgbJ3f9d+bsqUP//rAtMbFvYqa/uwz2kOUtAgAAC7L7Yndp0zd/Rrca3rG8walgQKNuAY26vThuwKjKLWbO+mNys+6JiYlejYZMHFHNZAt3/xK3vrslBQsWzGB33j4+ye+4yGRbkajU7zNZJ3JmpHh5pRZAnyJF0v+HyvLI+ySvFRYWJgUKpM7wTL8WD0/PdO8PSTi/7NWeQ34Oq9yiaZ0qVZoE1l2yMIP7mcUesrxFAABgQXb/GrvIzbNffX3hEdOZlHP9RnWcHJ2cRFzKly8VHulaveVDDTyC1v53MS5/2XLlJOjUqeSN7m2eMXj496dERAyG5J6T2bamN57ZOuXKlZMjBw+lHFTQyZPp30Sb5ZGn8Ctb1ul8UFBc8kZBQTey/H0Yt89+9RvD+H3ndq/+5cuZk15sWizL1TOQ01sEAADmYvfFrtqAl5pfmf3M4K+2htyJSZTEmGv7F748eVWhp59p7SLyxJAXq2+cNOSrY3cTjfcurB7z7LCvzjoVlwJdB/eMXjhh2u5wo8Rc3TrlrQkrb7qVSLPnTLZNcv/+/SzWKd3n2cAb377+zqbQeEmMODhn5CfpX2H3mCNP5vTU8wPdl039cE+kURJvbpg46+/Mr6uKiCRGRt4VJwdDoojEX9/y3tw1cXEPHhhz8BvN6S0CAABzsftiJwGj1u7+pN6RcW3K+bi7ODu7lWj45rEnZq76sqe3iBgqjV78Xdug0TV8vLwKl+m+rOC4pfO7eYj49P5s6SiZ38TXq2Bhv5ZfJwz96ZO+nml2nNm2UrhuvdJHP2xd8cWl0ZmtU/L5734eGD2vVTFPT0/fFouqvT28Wg6PPIVHuxlLxjp92qhwQR+vim+69R/kJ66umX/yiGP7N6e1vvR2lUKlAkoXazg5YeBLTe4tGtzly3PZ/43m8BYBAIC5GIzGnExjLC927Ejz7tBl+rxsrRcXcfHUmetG71Jl/Et45nvkR8bYsHMnz0UVKFc5oJDJMEziIy+dPBPq4le5fFF3BxG5dWJzkEudxgGej9vWeP/a8aOX3SrUD/A2ZLH/B7eDT1xI9KtaoVDihX0HHpRvWiGjl/VlfOTGm8e2nPeo36CMw82zFx/4lna9fvqGV6Wqhf8ZWGCY1783PmsuYSe3nM5Xq0l5r6T7cu3w9qs+Dev5uYrI/ZunT1544Fupip+nkzH66vGTkUVrVJagzcFudRuV80jec/5M9mDM7BYBAIBlUexUd+ubNsXGFll89pfePoa4C191q/1+2RXnP2uV7/Fb6ucWAQCAiPCuWPUVHjznk9VdB5co9k4lnzvBt8q+8MM3Fu5Y1r9FAAAgIkzs7ERc+PmTQaFOxSpW8POyTsey/i0CAAAmdnYhX8EyNZ8oo/YtAgAA3hULAACgCIodAACAIih2AAAAitDcmycAAACQO0zsAAAAFKG5d8VOWmkw7w6n9GAkCZgHn0YEWJN5E0fc7AQTOwAAAEVQ7AAAABRBsQMAAFAExQ4AAEARFDsAAABFUOwAAAAUQbEDAABQBMUuYVEPQwY6LLiX9POYs0sn9GlVN6Bwfo8i5Wq1f/mzHTcSkrfdPqK4wWCo9WHQo7s07hxVymAw1Jh2OoPb+2Ogm+ntOLoVLlO781u/nYqx4H1MkhgVeiU0KuHxKwJacWH6E4biIzY/unDzy76GxrMuZLmh5R7t/zxXwPvl9SKxi7oYKkw6bIFbAGxk/cu+huaf3EhdEB+0oFsxxxI9vj8bL3Llk+YZnCvdBq5OXf/u8WXv9Gleo6xvfncfvyqNeoz5fk9oouktPHYFmIHmPqDYNlpN2fXhk48s8SrnJiL3d4xv2G72vdaj35g6tnKh6JDti2aNbfXkxS27pjd0f7iiwXBs6dIz4ydUTNnUuHv5imsOWVXmRuM3zO3mLiKS+ODuld1fjR/3TFeXgFNT61m0Z4cu6FlqSffzO0f7W/JWAC2w/KPdsUqPN4eU8LXU7gFbS7yweGCbYbvqf77p1+fKP+wKXj0+/2dsXdO1DIUrPfzq/s5xTVvOut3qldffeaOmv1vE6c0/z3m16R87/9j+dcfC2VoB5kGxExGRQhUaNmyYbmnC4ekvz4j83+oD37X3SVrSrme38p3KPffuL2/9/cLDx+ETTZscXLYseMKEgOTN9ixfcbdp02pb72R6cwUDGjRsmD/5u+YtHPd69fx73bmp9cqb6w7lXUz0fRd3NzP/GRBAk3LzaHeq+/yMuo9fzaLHAFiK8eofL7QZtK7y7I3Lh1VxTlmcr2jVjE6WInLrr5f/NzNu+Pqjc1t6Jy1p/VSvPs2eq9H51VnDgj+sZ3jsCha/T3aDS7GZM/47e9bx5m9NSW51IiLi22va97O6lL6bssCna682x5YuO5uy4MDy5aFdnm7nkv0bcvb0dDUUKuQtIhJ5+OuX2tb0L5jfq2SdLhNWX344pDaGbv2wb+OKvp6FAp7oPmHV5aRLTBmuvPstP9/nv/y2X40S3u6eJWq0efnHk/dFLkx/othr22TXm2WSrjIbw3bMHNisSnFPzxJVmw2Ysf1W0h9eW97bUOej3f9NbF+x6DO/PcjVrw2wqgxSkPbRnkms0j7aMwmFxJ797bWnapUqWLB0nQ5j/7z08Ppu6qXYP57JV/21z8c39/N2z+9bufXoNVfOL3utbdVingWK1Og5b39MpseZ/hgSz696u1sdP2/3/IXLNuo/d2eEdX+XQJLQf0e27bPS74N1f4yskb1zWcSKeT/G9v94WnJpS+LT6eM1a+d38YnNxgowG4qdiIjcDt7/qOPXYuT8sWNR/k2alHx0VbfavUa90q5s6gKfrr3aHE9tdgeXLw/t3Ku9W/Zu2Pgg8uK2j2esSGzb56lCIpc+79F00ukG43/auPO/r5/zXNm3/eubwkUST0xt13Ze+FMf/bF+2fvto77rFjhmR3xmK4vI7R9HvRPWe96fm1bP/5/bX8+3GPJnpP/YvdfnNpNGM88b1w7Jbwz6uEPgtDP1xy/euPGXcfWDprXq+HHyHbi+8PVZsYO+3fbZ0znopoDFxN08/Wg6T9+MS/lhhil49NGeRVJMH+2ZhSL6v1da9l2S0GXWn//8MqnZ0VdfXRKd/hhPfDoreODiXXvXTKx0enbPCk0+9Xxt6eb/vuomv78xYXF4ZseZ7hgufj6g1/yIp2b/vmX9D8OLbXi91/s7Lf7rBR5lvLNlXLvun55vO2fFW3XSnsnSpXH/mdB4EZED+w8YnwhsmT/t3nxrPdWpSVnXbKwAs+FSrIiIbHi7fv23TRdUn3zsaIvz56V48eKP3dine682L32xLGTcuLIih5Yvv9l5Wnu381Oz2OLvFzwML6R+61h26Ko9z/uJ7J734dbmH9/84JmCIlKn5o8eJ0u2nr58cvPiM2cEPfXVhnd7eos0ecI//tqwLUFXd6/McOXA6iKJ+XvOX/rO054i0qC+x7kyHWctnNF1hMl/682fzDxQ/71zcweVFpF69X67vTvgg3lbxs5rISLX3bvMm9G/Qg5+e4Alhf02rP5vaRc2Svq/TCITOMRk1azWMXm0b8o4FO+emfnD/f8tWz61u6eINPzN8VTp7n+lO0bHzu9992IzD5GKr/f46M/1z3/6fudqItXeHjz72wUXr4iczt4xrD9+PK7S+Bd6tgxwlPpfr6m6M6ZEPP9Kw6pCvurRNdy/U7Nzv8+cvf+ZKfUefYIfsWJY/RWPLAn89NbGVwqFBweHuZUu6pm8NHhanfITDz38xqnvsrgv7jxmhcVPW+b+2CMmdiIi0muJ8VFH360mpUuXlmvXrqVdN/ZWyOmzN+6bLCnUvXerw0uXnROR48uXX+rcq0Pqs4/lzyS/d8jkTbKNxm/YlWTnxhXzXm8Z9+2zo9fek6iTJy/H/TPEN99Drm2+vpF49er180eP3q3evHnyALvc4AXrFgz2yWRlERGpGxiYnB+3Vm2byIkTp0zvQujJk7fKtmxZOvl7/5Yt/ENPngwTERGPunVpddCQYq9uejSdm15KfqF1ZpEx3TzLdVIf7ZmF4tTx4wkN2rZJzpNH2ycbZXCMBf38PERExODtXVD8y5RJWuzt7Z2TY5Amz77eKGR8Nb+GPV6aNH9zTM0nG/vT6mBdV26V+WD9n0sXz2xz+aNB7+5Lc4208Etp0mjc+EohESlYpozX/bNnryavV+rZHx+e5D7v5SPZWQHmQ7HLXNmqVd0ubN9++dGl95YNrVr52cXhpssKdevV6sjSZcFyYvnyCx2e7uCe+qOuC8If2vFmSl8qGNCgYZJGgT1GzP7prUY3167dL25eXvmKvbI5LkV8gtF4eFKVuLg4cXR0fPQ4Mls53d1wcHR0iI+PN11kNBrTrOPgIMnreHh4ZP93BNhSdlKQ5Tqpj/bMQpE2eS5ubmmimJ7BkOZ14Nk8BnFr+M72kKD/PhlQXY58/2qL8tWG/H4ze78JwEwajft2eFVnKTX0m9mtL88c/M6u7L38rXrNmg4H//kn9OG3LiVrJJ3jyifeup29FWA2FLvMObR7ZXjAlo/GffxcgwAAH9FJREFUr7qV+i9+fMjC79bF1mz/5KNXaAt179XqwLLla5cvP9+xl2mvk3zuXg8VcMnsl12iRAm5ffu2OFarXuXGlo0nkz/V58qiZxt1mHNEylWt6nJs587kN2xc+qa7f4P3jmSysoiIHNy8OXntmM3rd8RXrfrIqa5IlSqFzm/dmlJZL23Zer5QtWpFs/+7AbQgs8jkdB3JPBQVqlRx3PPf+uQ8PdixdXfOPx0vm8cQtmHO6x9tdW/We8SUL/84EPx9h2s//PhffNq1AEtycno4JfZ7/tvZT16eOXjSzvtZbyEiIsWeG/+sx5pJY0zPlyKR/02auTmbK8BsmPOLyMM3Tzy6yDugXrkmk7985Y/2PeveGj12cGBV7/ATm36cMWNT8Vc3vVEtzQ6K9OjVYvjUYddCO87olO7FoY/l6Vkg4VpYhFQcNq7Hx4O7D/T+5LVGzqeWvP/6L1EvjawmzlVHjyre6LX+013eae8ZvOqD91a59B1TvWLlDFeWfSIStey1vjUcJ7UvevOfj0Z8H9v7t+dKi4Q6OMjNkFNXw32LB44aXbv6O/97s9CMgdWMxxeOeXdP7bHft8z1rw+wkUwikzRve/hoz3SdR2USikINRj87qf3wXpUT3u3qF755xuuL77qnP45cH+cjChrO/jJx/tUCn49tVezuieXfbYmv9WZt/o2GrfgN+WbWsurDBk/ofnhOk6S3UcTdPL1//6MXdZyKVKrl5yEFOnzw1bBdz/SocWH4W8+2re3vEnZi48IZiyJbdAz4RUTk8SvAXPhHQ0QyePOEOPX9PW5xN49W8w8cqDdpwlefv/VlSLhbyQq1OszfMWVYw/TdzbdHrxbDX9nRaXqnAjm/+SpVqsiiL78MGjL2mZ+2x7w5cs7QJ8dEFagU+NLvC9+r6ygi9aatX+0y4u2R7aeEupRu0mfRX1Ob5BPJZGUR6Tj9y1rrx/eeHhJXol63b7fM7llQRHw7DOr92dgeFa59Hbp8wNi1Gx1HjXnv6a+vGkrUbDN+09w3K/ApQtAfn0xS8MijPfOkmDBUzDgU7m0/27J4zCvTXus4J7F0w74fLRr1Tp/Q9AeSy+N8hGOrab/Pvjt6zuAWb0a5Fqvc4qVlC8ZUzdWvBTCLUi98M2dp9SHPju9+aG5LkYzePCHFXtt2bU5TESna9cu9e5u/N/n7Re/+PDmqcJX6jTt/vnN8+8uz76wvIpKdFWAehnSvLLGxSSvN3C+m9NDWHbS43W/5NTo7M35ln8e+EgjIodixI827Q5fp88y7Q0Al5k0ccbMTvMYOAABAERQ75bh4lypTJOev8wMAALrHa+yUU3vCzhBbHwMAALAFJnYAAACKoNgBAAAogmIHAACgCIodAACAIjT3OXYAAADIHSZ2AAAAitDcx50YNm037w6NgU3Nu0PAbhFPwJrMmzjiZieY2AEAACiCYgcAAKAIih0AAIAiKHYAAACKoNgBAAAogmIHAACgCIpdqriIS8f27jtx8XZMtj6zOfT45l3n7pp+Ye0953C3gI4ZY24FHTxw8vq9hExXCTu5ZWdwpBWPCVBLxNmdmw9efpB28YPLBzcfupJuMbSKYiciEn3q11eaFM1fsHTNhg2q+RcqWr33zG23H7NNwoaJgb2/OGHyhRX3nJvdZirm1L+L/jl5z4obAjkQe+6Pt9qW8/LwrVivftXinp7FGw5dcDj5+Y7pg3Db+216fnrUdgcK6NyhT3oGDv81NO3i0F+HB45YFioidw6vWrTpXKI5bsuMu0IaFDuRq9/3bDBgTdG3/jp2NepBdFjQhqk1Do/vNvCnK1rds5l3G/HPlAGT19yy4oZAdiUcndmy1tM/G/st2BF0KzrmTsi+X18vtX54g3qj1keL8CAELK9Qj5l/Te9aSEQuLhszYNbWOHPs1Iy7QhoUu/AVb47dUP2jDSvebFetuLuTq0/5ViN+/uYFlzVzfwxKXcsYGxZyaO/BoFuxNt9z9nabeP/G6QMHztyITn1CFHZyy76L9xOjrgcd3rP36KXIBBGR2EsHdpyNkMjgXZuPXn+4blzE5ZP7dh84eTXq0WdTifeunjiQ1YaAmYXMf+WdI/Vm7Vw3tXeD8oVcnQuWqdN5zPK9P3cP+/SVafujMn4QPrhz7tC+IxcjHrlom0HQbp3YvP9SjMj9yzv2hSStHBMWfHjP3mMXwjnfAMkcnN093J0d7p7bve9CtISd2rL1VPIVosTom0EH9h4+fyclMfFXD28+ej1BEu+e3XbwatIyY+ztC0f37Dl85ub9pAWP7ir20oHNh6+m7CHy7M4tp8Iy2VUGt4i07L7YPfj356VRPd94uazBZKFT82nbDn33jG/Sd1GHv+lXtZBvlcD2zasUL9Vk4qY7ttxzNnZ7e+fMzmW9itUMDKxZrGDZTh/vSMrgtvfbDJo267ma5Rp07NKqZmm/+m+ui5DwbV/P/PuiXP537uSfDzwQufjn641L+vjVbtW2cdVSxSsP/uVC0uv3wrZ/1N6voF/9Fg38vYs2GfPvtdtpNgTM7+TP32z1HjrllTKP/kNVqPe012uc/vrbv9M9CI1hmyc39Qto/lRgbX+/uqPXRYhIpkHb/G7gC9+t/rhZyUpdPtmfIHfWjW/qX6x8o9aBtcsUq9D9y1PxVr63+H979x5Q8/0/cPzVvaTcS0XoQiIJuWTIIvfb5rLv3O/3aXPd5lIuyTa3MbONmctirnPJpVAod5ZLcheKRFEimur3R0V32Ri/t+fjr/U5n/N6f07OWc8+5xLeTRlPxUb6z1u8/55c/HP69I0XRST2oE876xJmjm7uDazNKraefTJJRCR+k4frOF+/MU4WNXssvyTy9/lVfWqalqpU1+3DOlXMzGt5+N0TyT4q2newq8em568mOjP/I7cZwXmMymdF5PTeh92l8PBnVtWrF8m+VbOklaNjxeIiIk8Ofv2Rx+W2G28mPIh7GHdiqumSjp0WX3l7k186NmnP+M4TrrZYG5GQkBAfsdb96lddv9ybcfcPX7JKb9GVe1HRd8JmOp6dPWvdfdP/Ld44wkHsh64JnNVaX4K8+82L67ot+kFCQnzM1h4pK7x+PisiiTu++GhafJ/ddxIexl7f1fPx7M4ewe2zXRF4/f4OC7uk4eRcO/eftLapU6f4vfDo5jnvhHd+X3Fr8ql7UdH3Li1uEjH3u40JBT/QLs/75sqE0PuxKzvrnpozzOfWx1vuPEhMiD06tsTWL7y3U3bAC1WHrF7S01IaTPDf9VV9ebDps07fJA08cOfh/fsPowOHPZrWqu+6O+l7Hpz1vc7CywkR85rIkw0TB/jqf3bkbvzDhPiI39xvzPdefTvHqIJkGVXQisjqvQ+7u3fvirGxcf47hKxccbNexzaGEcdCQg6eeVijbVODfb5+t/LY8174vgzHbyS91smvcsCpActX3m/vOadjBT0RvQod50xpG7ti+e6M9802HzOrpZmWaBja9+rsLLGxOd5xkVZjzK5Te2a1MjXQ0NA3qmJVTjMuLk4kacsvq6S7t2eT0toaRco38/pp6fgmpRNfcqDAv5UQG/u3XokSBnlcVKZMabl7N9ervKX5mFktymqK6Fu3a14tLS7ufsEPtMS6A3zaWuqKiCQmJopG0oPYpBStYs6eh+7f/rGl1hu9dcD/Y093rfgj2a1T478vHA4JOXwxtUkbl4frV+9Jf/YmtdnwqY1MNEVEtJpODzm1+eu6pXRF06C0nZWJxMW9yjv9XowqcEVklftX4fdMlSpVZGNEhIh5ts0xh303Xa7UqUcD/evXH2hd3ug1MeD5ZXauNsXyui+FLvcYu0tERKyHb1o/4DVOfpUDTr569WklN4eimRcYOThUfBJ0Nf3HWAkrqxIZ2/X09HIP1yhZzvjAnLEz9p++ePlixCNtzTSxF5Grly6lVnGtrpOxl3G9Pl/XExF+VcKbVcrOrvSTY5eiRMrluOTvixcjDBwdbUX8s20vbWubeQ/X0krvsoSCHmglbWwy9284bun4sOEDq5mOrFy3iVvrboOGdXV8A7cJUELU9espKQmrJk1c/3xTvUZVijwRERFLG5vMHxc6phbaAd+OmHjo7KXLlyKTdVKk+Cut82JU/ivq/qtboqD3PuzMatY0ve23+di3Ls5Zfj2/s2Zcj9HGv37ao4GRmZmhvsWswF/cMi5KTX6clKprKHIk56hmPn/95ZPl69c3+VUO2KBYMYmPj39xUWJi4vNTfJk/6/Jz67eOtYff6z5v2txpTnZWZfcPNOgjIlK0aFF5/Pjx892S7l67lWpiXeAo4N+r0aBBkcW+v18YNb5K1teUSsxa3z0ptb2cc/3/S1s71ybjfB9oJ0R0dV/8TDBr7eN/zSvmTPDe3X6/L+rrvP5K6Ikv7V/3TQKUUNbMTEzbLA6clvnrT8rTR09E31AeSLbHVbhP0/rfGX42b/IPLg52FYqsa2vkk/fAFx49evHDJsuofFdETu/9U7HSaMK0FrcWjpgSEvv8U37v7Ji88KBhi84tjESkVkOXJ9tWbMk4d/z0yJfVi9aaGf4WJ79krFEd5yq3tm44kvGOoeTD67ZEVXGuY1SYI352cOfupy3Hzx3YyrlyWSOt6xcupb+DsLyzc9nTm/+8lr5g6tEpLlbtfrhemInAv1HiE59JTqHefSfvi3nxpteE0Pn9xu6ynDBncPlCDSncAy3Gt59Tq9ln9Uwc3D75fM4fXi00T504zYvsgLwVqd/Q6dr65QczGixhR79KRq1/js652719O08YdvP06eFWy8qkSNrFC9fymKWpqSkR1zJ+wDw7tmd/nh8zXsgVwRk7ESk/YNEPu90HNa2xp0O7JvYlHpzb57f1aEq7JYt6m4mImPacPn5B008bPx7Ru77u+U2/rnzY6bfBDiL5f/79m578krEOHrN6Le3S1jXBo2u1tLDV83/7u896D4cC5ukbGmpe2b3c17l7V0dnpzTPWUNmJbiViTu2epV/lFF88tbtF2u3njCrdc1Bbm2ihrkbX9m+bHlMi+/724v+wedXbGzNPQlvgKb9aN8VEd2GuFXf1qp1U6eK2tFhB3ZsOSnNF/hPddaXbPfexvnMyO+Blp2Jk51mv8ldO98d0LTsw3N+y7YadVrpyr0a751bQYtnZ3261NJ9hMvzLwwNDeWs309/2H3crd5InyFL2rV1jR3RrXry8TW/rjMYuLOnhUhstmmlajtbxa2YMtImop7hrf3L1wY91osJ2bjvhm2TLKOc69UvN27J0N7Fh3xY4orvrCUxeZ+Fs81nReTEGTsRDatef5wOXTva1eTR+WNn44vXH778r/Ob+lhmPPejU9cz5Oivn5hFhew9k1x37Pajqz81FxENk+quDayNs/zHfzb5JWOLd1gWGuTtqnFmz+4wTbeZ+0OXti8uIlLavklD2+ePVx3zmq51LA1EpNjHk+Z9ZHj498CIFFuPzf7TXZ6GbPwz5G6Vr3ed3r5ggOWF7X89EPNeG0K3eVSLO7T7WLzNoNUnNg2okO2Kr+9fA8hGx+aTxYfP7PnmfzUM74Qeu5Jc1m3cxvCL2wZXTY+urHfC0vZNXGyKZV5R18LJtXZ5fcn3gSZlqrlm2b/qmJ37F3Qsdm3v1oAzj2p84Xd05Sdl/+sbC7xVxW1dXK0Sg7dldeTmM73ytVydyumJiE0v76lNk3f8cShaxMh94Yng2e6Gl4L2XTZoPj340KJmxUVEx6Kma50KmZ/boOE8LWDLKPuYPRu2nkxqMD/o1LrZHxse9Q9PyjpKs7F3wJpRtR7uWzZ78SFrn+2LejepWjrXqHxXRE4aaWnv1t8Z1QgKeb0D01wbvt6BwHuLhyfwX3q9jzgebu8JztgBAAAogrADAABQBGEHAACgCMIOAABAEYQdAACAIgg7AAAARRB2AAAAinjnPscOAAAA/wxn7AAAABTxzv0txMOTXvPA+tNe80AAAIB3E2fsAAAAFEHYAQAAKIKwAwAAUARhBwAAoAjCDgAAQBGEHQAAgCIIu6hdc72mLj/2ONvG+wd+8vol5P5bOqRXlHJmtdeCwDuvcpXkW0e3rFo857uFvjuPXo59+m9GFSwt7A+vuQFRb2AyAADIhbCL3DnHc0rf7p6Hn2bZGLd/sefPwXFv7aBeSNg9a+BM/9iCdnl22tfz+73RhZ2YcunXjytb1+v6xQ/bgnYumdCxupnjoC3Raf9k1Mulnl3jOcc/8g1MBgAAuRB2IiJGRaPnDvY+/extH0ceksK2LdlyJrGgXfS6rUs8NcmhkAPvLhk2NMBqeljsnTN7t23bGxoR/ovdn13a+Jx99VGv4s1NBgAAmd65vzzxVtT9Yqr+vPGD5316cEwVjTwuT7t/fo//vtNRGuVqNG7uZldCQ0TkwoZpe02GdikStCn46gNNc+cOXVwt9XJdNSpgrm9yp/4WJ9bvu5JoYO3cvFWjSkVERMLXee03/6xPhfANKy5UGtu3gU4eq8SH/PLdzhsS6T/Pq2yvr7o76Yg8uha8K+j4lQeGlRt3aFPbRFtEUsI3fRdUdsjIpqaFOaTTx48nW49pa2+YcUN1K/Ue5/H7+LjrT6W63otRIpISE7rDb/9VDasmbV2f+c0+X/Or7o46BSyREn1yx+4j56OSDC1rtezgmn4zMxXqIPNc8R//qwIA8N7hjJ2IiIZZjx9mNjrtOeTnG7kvTAlb3L6GY+dJqwMDfSd2rlGj3Y9nU0REzq+fOmt6f/ePpm89esxvztAPqzWbfyX3tSP950yd1KOp69AFW4L8Fg51q15vuF+MiMi5tZ6LVi/s2bz/khN3UzTyXSWrtKu+PWpWb//1ysBDu38e3rBy89mnn0rWZzkLc0hWtrZaoT+Nnx9wMT5zAZev/A9820Yv26iUc4vb1XQZuHDHbt8JLWq3GunluSo0uYAlkg5Mrm/n0mfG6sADmxeOalm59tjgJ1mXfflB5rciAAAoLMIunUaFwYumVD80YfiKnK/vv71s1Bf7aywIPRe0dWvQub8WOOwf/flvt9Mvu37wyaDgE1t+XxcUtqq7ZvDa7bfzmp34V6TTynOn92zfczp8e++kRcO8g9OD5czKQ27+Z/duGPeBdt6rFGs4cExLSynn7jGlu5POXd9RQwLrrwo/t3fzRr9jlw55xH/V5quDqdkXe+khVRr525J+ZkfGuVcpXcrapeOgSQs2noj5O+cxx6waPfqkm++ZEzu2+J86MdHo7LWCl0gLXLboisv3oeH7/bbtCzvqXffimg2n8v1u5zWhoBUBAEChEHaZNCt/8dOX5XeO/nxjtncqJO/fFajf5fMBFbVFRLQrDfTorBe460B6CWm16t+vvJaISJGGDZ3k8ePHIuEbvDL8eCBjkn3/8e1Ka4iIlGo2cViDGzt2hImIiIZ7777pVy9wlQyph3btfmKiGbbEe8aMGTNmzNkea2QUGRxyPfvNyOuQsjOw77P0SOTda0c2zh5Yr8jltZO71qlQrc/6bOcqn+7Z4m/cZUjH0iIiWhZ9B7bL8oxuXktotP71Xtz2QeXSkuMjz+7bE3ov9cmTbGfsXnaQBa4IAAAKhbB7QdtxwuJRJdd8Ni7g4YuNtyMjUy3Kl3/+fdK0tLRIiYxMPw9W3MREN2O7hkbGa9aSHz3IkPAk47lOS0vL5/PMK1XSu3nzpoiIlDY31y3EKhkex8U9KWKgl/IsU+kPR3r2djbMfiPyOqSsEqPOhd1MEC3jinXb9Z8w13fvhajweU3uLB/kFZDl3F/kjRupZmZmz78z5uZlXrLE04trPne3NzU2qdqsz9SAO9pauZcucEKBKwIAgELhzRNZ6bl4LR64zm3opA8+ydxU1sJC81ZkZKpUSa+utKio25oWFqYiknc6OfaaO7fXiy+vicjNmzdF7NK/jrl586m5i7nIdRFNzcySy3+VB88nFa1QoWRy0dZfTumYEUUxoTuPS1WT7MvnWXNZhH/fqu6B0VEHPzPP3FLEdmj/lmMCbt9+LJL5TgUzc3ON6OhoESsREUmJjr5X8BJhPh3+t7bWyuDbn9oW1RTZ2lu/XwEHkceEAlcEAACFwhm77Ayb+izo8XDhuJ8iMjboNW7h+njdvN9upIiIpNxYOndtkqt7o1d6njBs6bfb0z8T7/6+GT8Em7u7V8+xR8GrpKamiojU79ix2JZvvjmZ/uRq9MqBrm1mhuR6h8VLOLZpW+7QdyN/Pvv8rGTy5WXL92rUdWtc9MVeRVxbNrq//pdt90VEUm+vWLI1/+dVRUSehYdf0q3Vur1tUU2RtDh/v4NP09LSXuW4XnVFAACQG2fscirRfvacdn49/pSK6V+b950/58+WQ50cVje2S7sQdOBho/k7+pkXOCInu/JXBjs4VattGnMs8HzJgaunfKifc5d8VzG2sCh6ZunIQQYTFoxoPvXHns27N666rUkN3fCgg3FOMwMHmIs8zWPJfOk2/nbXsviPhtWy9K5qb22q++j6mZM3jdr+6O9RMetu5Qct/HajWzeH+m51S94493c1FzPR0dXNZ6aIdtNPupbqMqpBq531y8SdPP6omk3FWL8ZnruqTir0gb3iigAAIDctT0/Pt30M2UQGvuaB5T4s+HINDQ1jmwaujubPPzDNsEajWsWMrRu4uta2NBDRNHH+tG/nWuWK6Jeo1naEz+yJLStoZlyxtH2ThrbFM8fomNd0rWNpkH18VMDcJZoTz/851EJXu5RT1/Gzvx3RoETGFcpUc3WxKZa+X36r6FR2aWZrlKZt4ehStUxxu3Z9uje2KWFgbNNskPci74+tddInaZo6uNarVLRQh6RTpuZH/fq2a1TD2qSEiV3D9v0nL5jv8UGZzNuUMUrTtG73ni0qGWiZ1Os/1bPaKe/N+n0mdrDO71aXtOvUu52tsaZuGYc2n3tPG9qlqb2xlpFNnZrlDDRMqrs2sDZ+6UHmtyIAACgsjVd8xuyNO1z4czyFU3/aax74ao6MLV//8nfPNnUr8N0E75ybi1rUXdbQ7/DkWloiycEedi2veUVu7llcpRUBAFAOT8UiL+U/HtB0Zt8OH5xtW7dUzP61gaW+3NntzTbWf78iAADK4YzdG3bv+Dq/h869mlYs+M2q76KH4Ts3BJy5q122Sv3mrWuV/Q9+B/jvVwQAQCmEHQAAgCL4uBMAAABFEHYAAACKIOwAAAAU8c69xg4AAAD/DGfsAAAAFEHYAQAAKIKwAwAAUARhBwAAoAjCDgAAQBGEHQAAgCIIOwAAAEUQdgAAAIog7AAAABRB2AEAACiCsAMAAFAEYQcAAKAIwg4AAEARhB0AAIAiCDsAAABFEHYAAACKIOwAAAAUQdgBAAAogrADAABQBGEHAACgCMIOAABAEYQdAACAIgg7AAAARRB2AAAAiiDsAAAAFEHYAQAAKIKwAwAAUARhBwAAoAjCDgAAQBGEHQAAgCIIOwAAAEUQdgAAAIog7AAAABRB2AEAACiCsAMAAFAEYQcAAKAIwg4AAEARhB0AAIAiCDsAAABFEHYAAACKIOwAAAAUQdgBAAAogrADAABQBGEHAACgCMIOAABAEYQdAACAIgg7AAAARRB2AAAAiiDsAAAAFEHYAQAAKIKwAwAAUARhBwAAoAjCDgAAQBGEHQAAgCIIOwAAAEUQdgAAAIog7AAAABRB2AEAACiCsAMAAFAEYQcAAKAIwg4AAEARhB0AAIAiCDsAAABFEHYAAACKIOwAAAAUQdgBAAAogrADAABQBGEHAACgCMIOAABAEYQdAACAIgg7AAAARRB2AAAAiiDsAAAAFEHYAQAAKIKwAwAAUARhBwAAoAjCDgAAQBGEHQAAgCIIOwAAAEUQdgAAAIog7AAAABRB2AEAACiCsAMAAFAEYQcAAKAIwg4AAEARhB0AAIAiCDsAAABFEHYAAACKIOwAAAAUQdgBAAAogrADAABQBGEHAACgCMIOAABAEYQdAACAIgg7AAAARRB2AAAAiiDsAAAAFEHYAQAAKIKwAwAAUARhBwAAoAjCDgAAQBGEHQAAgCIIOwAAAEUQdgAAAIog7AAAABRB2AEAACiCsAMAAFAEYQcAAKAIwg4AAEARhB0AAIAiCDsAAABFEHYAAACKIOwAAAAUQdgBAAAogrADAABQBGEHAACgCMIOAABAEYQdAACAIgg7AAAARRB2AAAAiiDsAAAAFEHYAQAAKIKwAwAAUARhBwAAoAjCDgAAQBGEHQAAgCIIOwAAAEUQdgAAAIog7AAAABRB2AEAACiCsAMAAFAEYQcAAKAIwg4AAEARhB0AAIAiCDsAAABFEHYAAACKIOwAAAAUQdgBAAAogrADAABQBGEHAACgCMIOAABAEYQdAACAIgg7AAAARRB2AAAAiiDsAAAAFEHYAQAAKIKwAwAAUARhBwAAoAjCDgAAQBGEHQAAgCIIOwAAAEUQdgAAAIog7AAAABRB2AEAACiCsAMAAFAEYQcAAKAIwg4AAEARhB0AAIAiCDsAAABFEHYAAACKIOwAAAAUQdgBAAAogrADAABQBGEHAACgCMIOAABAEYQdAACAIgg7AAAARRB2AAAAiiDsAAAAFEHYAQAAKIKwAwAAUARhBwAAoAjCDgAAQBGEHQAAgCIIOwAAAEUQdgAAAIog7AAAABRB2AEAACiCsAMAAFAEYQcAAKAIwg4AAEARhB0AAIAiCDsAAABFEHYAAACKIOwAAAAUQdgBAAAogrADAABQBGEHAACgCMIOAABAEYQdAACAIgg7AAAARRB2AAAAiiDsAAAAFEHYAQAAKIKwAwAAUARhBwAAoAjCDgAAQBGEHQAAgCIIOwAAAEUQdgAAAIog7AAAABRB2AEAACiCsAMAAFAEYQcAAKAIwg4AAEAR/wfH1f35H9/fZAAAAABJRU5ErkJggg==",
      "text/plain": [
       "plot without title"
      ]
     },
     "metadata": {
      "image/png": {
       "height": 420,
       "width": 420
      }
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "CellChatDB <- CellChatDB.mouse\n",
    "showDatabaseCategory(CellChatDB)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b2c4e368",
   "metadata": {},
   "source": [
    "Pick the subset of the database that you actually want to use"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "72780083",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "\n",
    "# use a subset of CellChatDB for cell-cell communication analysis\n",
    "CellChatDB.use <- subsetDB(CellChatDB, search = \"Secreted Signaling\") # use Secreted Signaling\n",
    "# use all CellChatDB for cell-cell communication analysis\n",
    "# CellChatDB.use <- CellChatDB # simply use the default CellChatDB"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3f1c55ee",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Create a CellChat object from a data matrix\"\n",
      "Create a CellChat object from spatial transcriptomics data... \n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in CellChat::createCellChat(object = sample_data$data, meta = sample_data$meta, :\n",
      "“The 'meta' data does not have a column named `samples`. We now add this column and all cells are assumed to belong to `sample1`! \n",
      "”\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Set cell identities for the new CellChat object \n",
      "The cell groups used for CellChat analysis are  B naive activated, B naive resting, CD4, CD8, Dendritic Cells, Endothelial, Macrophage, Mesenchymal, Monocytes, Neutrophils, NK_cell, Plasma Cells, T-helper, Treg, Tumour \n",
      "[1] \"Create a CellChat object from a data matrix\"\n",
      "Create a CellChat object from spatial transcriptomics data... \n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in CellChat::createCellChat(object = sample_data$data, meta = sample_data$meta, :\n",
      "“The 'meta' data does not have a column named `samples`. We now add this column and all cells are assumed to belong to `sample1`! \n",
      "”\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Set cell identities for the new CellChat object \n",
      "The cell groups used for CellChat analysis are  B naive activated, B naive resting, CD4, CD8, Dendritic Cells, Endothelial, Macrophage, Mesenchymal, Monocytes, Neutrophils, NK_cell, Plasma Cells, T-helper, Treg, Tumour \n",
      "[1] \"Create a CellChat object from a data matrix\"\n",
      "Create a CellChat object from spatial transcriptomics data... \n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in CellChat::createCellChat(object = sample_data$data, meta = sample_data$meta, :\n",
      "“The 'meta' data does not have a column named `samples`. We now add this column and all cells are assumed to belong to `sample1`! \n",
      "”\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Set cell identities for the new CellChat object \n",
      "The cell groups used for CellChat analysis are  B naive activated, B naive resting, CD4, CD8, Dendritic Cells, Endothelial, Macrophage, Mesenchymal, Monocytes, Neutrophils, NK_cell, Plasma Cells, T-helper, Treg, Tumour \n",
      "[1] \"Create a CellChat object from a data matrix\"\n",
      "Create a CellChat object from spatial transcriptomics data... \n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in CellChat::createCellChat(object = sample_data$data, meta = sample_data$meta, :\n",
      "“The 'meta' data does not have a column named `samples`. We now add this column and all cells are assumed to belong to `sample1`! \n",
      "”\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Set cell identities for the new CellChat object \n",
      "The cell groups used for CellChat analysis are  B naive activated, B naive resting, CD4, CD8, Dendritic Cells, Endothelial, Macrophage, Mesenchymal, Monocytes, Neutrophils, NK_cell, Plasma Cells, T-helper, Treg, Tumour \n",
      "[1] \"Create a CellChat object from a data matrix\"\n",
      "Create a CellChat object from spatial transcriptomics data... \n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in CellChat::createCellChat(object = sample_data$data, meta = sample_data$meta, :\n",
      "“The 'meta' data does not have a column named `samples`. We now add this column and all cells are assumed to belong to `sample1`! \n",
      "”\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Set cell identities for the new CellChat object \n",
      "The cell groups used for CellChat analysis are  B naive activated, B naive resting, CD4, CD8, Dendritic Cells, Endothelial, Macrophage, Mesenchymal, Monocytes, Neutrophils, NK_cell, Plasma Cells, T-helper, Treg, Tumour \n",
      "[1] \"Create a CellChat object from a data matrix\"\n",
      "Create a CellChat object from spatial transcriptomics data... \n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in CellChat::createCellChat(object = sample_data$data, meta = sample_data$meta, :\n",
      "“The 'meta' data does not have a column named `samples`. We now add this column and all cells are assumed to belong to `sample1`! \n",
      "”\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Set cell identities for the new CellChat object \n",
      "The cell groups used for CellChat analysis are  B naive activated, B naive resting, CD4, CD8, Dendritic Cells, Endothelial, Macrophage, Mesenchymal, Monocytes, Neutrophils, NK_cell, Plasma Cells, T-helper, Treg, Tumour \n"
     ]
    }
   ],
   "source": [
    "\n",
    "\n",
    "cellchat_list <- lapply(names(ovVis_input), function(sample_name) {\n",
    "  sample_data <- ovVis_input[[sample_name]]\n",
    "  \n",
    "  cellchat <- CellChat::createCellChat(\n",
    "    object = sample_data$data,\n",
    "    meta = sample_data$meta,\n",
    "    group.by = \"cell_type_collapsed\",                 # adjust to the column name in meta for cell types\n",
    "    coordinates = sample_data$spatial,\n",
    "    datatype = \"spatial\",\n",
    "    #scale.factors = sample_data$scale.factors,\n",
    "    spatial.factors = data.frame(ratio = 1, tol = 32.5),  # adjust tol if needed\n",
    "    do.sparse=TRUE\n",
    "  )\n",
    "  \n",
    "  # Set CellChat database (optional: replace with the specific database you want to use)\n",
    "  cellchat@DB <- CellChatDB.mouse  # or CellChatDB.use if you loaded a custom DB\n",
    "  \n",
    "  return(cellchat)\n",
    "})\n",
    "\n",
    "names(cellchat_list) <- names(ovVis_input)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3b0eb1dd",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'B naive activated'</li><li>'B naive resting'</li><li>'CD4'</li><li>'CD8'</li><li>'Dendritic Cells'</li><li>'Endothelial'</li><li>'Macrophage'</li><li>'Mesenchymal'</li><li>'Monocytes'</li><li>'Neutrophils'</li><li>'NK_cell'</li><li>'Plasma Cells'</li><li>'T-helper'</li><li>'Treg'</li><li>'Tumour'</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 'B naive activated'\n",
       "\\item 'B naive resting'\n",
       "\\item 'CD4'\n",
       "\\item 'CD8'\n",
       "\\item 'Dendritic Cells'\n",
       "\\item 'Endothelial'\n",
       "\\item 'Macrophage'\n",
       "\\item 'Mesenchymal'\n",
       "\\item 'Monocytes'\n",
       "\\item 'Neutrophils'\n",
       "\\item 'NK\\_cell'\n",
       "\\item 'Plasma Cells'\n",
       "\\item 'T-helper'\n",
       "\\item 'Treg'\n",
       "\\item 'Tumour'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 'B naive activated'\n",
       "2. 'B naive resting'\n",
       "3. 'CD4'\n",
       "4. 'CD8'\n",
       "5. 'Dendritic Cells'\n",
       "6. 'Endothelial'\n",
       "7. 'Macrophage'\n",
       "8. 'Mesenchymal'\n",
       "9. 'Monocytes'\n",
       "10. 'Neutrophils'\n",
       "11. 'NK_cell'\n",
       "12. 'Plasma Cells'\n",
       "13. 'T-helper'\n",
       "14. 'Treg'\n",
       "15. 'Tumour'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       " [1] \"B naive activated\" \"B naive resting\"   \"CD4\"              \n",
       " [4] \"CD8\"               \"Dendritic Cells\"   \"Endothelial\"      \n",
       " [7] \"Macrophage\"        \"Mesenchymal\"       \"Monocytes\"        \n",
       "[10] \"Neutrophils\"       \"NK_cell\"           \"Plasma Cells\"     \n",
       "[13] \"T-helper\"          \"Treg\"              \"Tumour\"           "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>157</li><li>42</li><li>212</li><li>127</li><li>134</li><li>311</li><li>1095</li><li>84</li><li>144</li><li>19</li><li>176</li><li>197</li><li>48</li><li>69</li><li>6461</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 157\n",
       "\\item 42\n",
       "\\item 212\n",
       "\\item 127\n",
       "\\item 134\n",
       "\\item 311\n",
       "\\item 1095\n",
       "\\item 84\n",
       "\\item 144\n",
       "\\item 19\n",
       "\\item 176\n",
       "\\item 197\n",
       "\\item 48\n",
       "\\item 69\n",
       "\\item 6461\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 157\n",
       "2. 42\n",
       "3. 212\n",
       "4. 127\n",
       "5. 134\n",
       "6. 311\n",
       "7. 1095\n",
       "8. 84\n",
       "9. 144\n",
       "10. 19\n",
       "11. 176\n",
       "12. 197\n",
       "13. 48\n",
       "14. 69\n",
       "15. 6461\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       " [1]  157   42  212  127  134  311 1095   84  144   19  176  197   48   69 6461"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'data.raw'</li><li>'data'</li><li>'data.signaling'</li><li>'data.scale'</li><li>'data.smooth'</li><li>'images'</li><li>'net'</li><li>'netP'</li><li>'meta'</li><li>'idents'</li><li>'DB'</li><li>'LR'</li><li>'var.features'</li><li>'dr'</li><li>'options'</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 'data.raw'\n",
       "\\item 'data'\n",
       "\\item 'data.signaling'\n",
       "\\item 'data.scale'\n",
       "\\item 'data.smooth'\n",
       "\\item 'images'\n",
       "\\item 'net'\n",
       "\\item 'netP'\n",
       "\\item 'meta'\n",
       "\\item 'idents'\n",
       "\\item 'DB'\n",
       "\\item 'LR'\n",
       "\\item 'var.features'\n",
       "\\item 'dr'\n",
       "\\item 'options'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 'data.raw'\n",
       "2. 'data'\n",
       "3. 'data.signaling'\n",
       "4. 'data.scale'\n",
       "5. 'data.smooth'\n",
       "6. 'images'\n",
       "7. 'net'\n",
       "8. 'netP'\n",
       "9. 'meta'\n",
       "10. 'idents'\n",
       "11. 'DB'\n",
       "12. 'LR'\n",
       "13. 'var.features'\n",
       "14. 'dr'\n",
       "15. 'options'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       " [1] \"data.raw\"       \"data\"           \"data.signaling\" \"data.scale\"    \n",
       " [5] \"data.smooth\"    \"images\"         \"net\"            \"netP\"          \n",
       " [9] \"meta\"           \"idents\"         \"DB\"             \"LR\"            \n",
       "[13] \"var.features\"   \"dr\"             \"options\"       "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "cellchat <- cellchat_list[[1]]\n",
    "levels(cellchat@idents)\n",
    "groupSize <- as.numeric(table(cellchat@idents)) # cells per type\n",
    "groupSize\n",
    "slotNames(cellchat) #data.raw is only populated if you provide the raw counts which aren't actually required by cellchat\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "36526208",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "\n",
    "library(presto)\n",
    "\n",
    "# Subset data to relevant genes and ligand-receptor pairs\n",
    "cellchat <- subsetData(cellchat)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c9e46088",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "cellchat@spatial\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "aa74880f",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The number of highly variable ligand-receptor pairs used for signaling inference is 1255 \n",
      "triMean is used for calculating the average gene expression per cell group. \n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in max(data):\n",
      "“no non-missing arguments to max; returning -Inf”\n"
     ]
    },
    {
     "ename": "ERROR",
     "evalue": "Error in aggregate.data.frame(as.data.frame(x), ...): no rows to aggregate\n",
     "output_type": "error",
     "traceback": [
      "Error in aggregate.data.frame(as.data.frame(x), ...): no rows to aggregate\nTraceback:\n",
      "1. aggregate(t(data.use), list(group), FUN = FunMean)",
      "2. aggregate.default(t(data.use), list(group), FUN = FunMean)",
      "3. aggregate.data.frame(as.data.frame(x), ...)",
      "4. stop(\"no rows to aggregate\")",
      "5. .handleSimpleError(function (cnd) \n . {\n .     watcher$capture_plot_and_output()\n .     cnd <- sanitize_call(cnd)\n .     watcher$push(cnd)\n .     switch(on_error, continue = invokeRestart(\"eval_continue\"), \n .         stop = invokeRestart(\"eval_stop\"), error = NULL)\n . }, \"no rows to aggregate\", base::quote(aggregate.data.frame(as.data.frame(x), \n .     ...)))"
     ]
    }
   ],
   "source": [
    "\n",
    "cellchat <- identifyOverExpressedGenes(cellchat)\n",
    "cellchat <- identifyOverExpressedInteractions(cellchat)\n",
    "cellchat <- computeCommunProb(cellchat, raw.use = FALSE)\n",
    "cellchat <- filterCommunication(cellchat, min.cells = 10)\n",
    "cellchat <- computeCommunProbPathway(cellchat)\n",
    "cellchat <- aggregateNet(cellchat)\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.4.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
