{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "69e902d0",
   "metadata": {},
   "source": [
    "## 1. Load libraries\n",
    "This loads the required R packages for:\n",
    "- **CellTrek** (mapping scRNA → spatial)\n",
    "- **Seurat** (single-cell integration)\n",
    "- **CellChat** (cell–cell communication, though not used later in this script)\n",
    "- Visualization (`ggplot2`, `pheatmap`, `patchwork`, etc.)\n",
    "- Clustering/statistics (`ConsensusClusterPlus`, `RColorBrewer`)."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "896be9d0",
   "metadata": {},
   "source": [
    "Need to install this too in env beforehand\n",
    "#mamba install -c conda-forge -c bioconda bioconductor-singlecellexperiment\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "d68f5195",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "“replacing previous import ‘data.table::first’ by ‘dplyr::first’ when loading ‘CellTrek’”\n",
      "Warning message:\n",
      "“replacing previous import ‘MASS::select’ by ‘dplyr::select’ when loading ‘CellTrek’”\n",
      "Warning message:\n",
      "“replacing previous import ‘data.table::last’ by ‘dplyr::last’ when loading ‘CellTrek’”\n",
      "Warning message:\n",
      "“replacing previous import ‘data.table::between’ by ‘dplyr::between’ when loading ‘CellTrek’”\n",
      "Warning message:\n",
      "“replacing previous import ‘dplyr::as_data_frame’ by ‘igraph::as_data_frame’ when loading ‘CellTrek’”\n",
      "Warning message:\n",
      "“replacing previous import ‘dplyr::groups’ by ‘igraph::groups’ when loading ‘CellTrek’”\n",
      "Warning message:\n",
      "“replacing previous import ‘dplyr::union’ by ‘igraph::union’ when loading ‘CellTrek’”\n",
      "Warning message:\n",
      "“replacing previous import ‘igraph::groups’ by ‘plotly::groups’ when loading ‘CellTrek’”\n",
      "Warning message:\n",
      "“replacing previous import ‘magrittr::set_names’ by ‘purrr::set_names’ when loading ‘CellTrek’”\n",
      "Warning message:\n",
      "“replacing previous import ‘data.table::transpose’ by ‘purrr::transpose’ when loading ‘CellTrek’”\n",
      "Warning message:\n",
      "“replacing previous import ‘igraph::simplify’ by ‘purrr::simplify’ when loading ‘CellTrek’”\n",
      "Warning message:\n",
      "“replacing previous import ‘igraph::compose’ by ‘purrr::compose’ when loading ‘CellTrek’”\n",
      "Warning message:\n",
      "“replacing previous import ‘purrr::partial’ by ‘randomForestSRC::partial’ when loading ‘CellTrek’”\n",
      "Warning message:\n",
      "“replacing previous import ‘data.table::melt’ by ‘reshape2::melt’ when loading ‘CellTrek’”\n",
      "Warning message:\n",
      "“replacing previous import ‘data.table::dcast’ by ‘reshape2::dcast’ when loading ‘CellTrek’”\n",
      "Warning message:\n",
      "“replacing previous import ‘purrr::discard’ by ‘scales::discard’ when loading ‘CellTrek’”\n",
      "Warning message:\n",
      "“replacing previous import ‘igraph::as_data_frame’ by ‘tibble::as_data_frame’ when loading ‘CellTrek’”\n",
      "Warning message:\n",
      "“replacing previous import ‘igraph::crossing’ by ‘tidyr::crossing’ when loading ‘CellTrek’”\n",
      "Warning message:\n",
      "“replacing previous import ‘magrittr::extract’ by ‘tidyr::extract’ when loading ‘CellTrek’”\n",
      "\n",
      "Attaching package: ‘dplyr’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    filter, lag\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    intersect, setdiff, setequal, union\n",
      "\n",
      "\n",
      "Attaching SeuratObject\n",
      "\n",
      "Loading required package: viridisLite\n",
      "\n",
      "Loading required package: igraph\n",
      "\n",
      "\n",
      "Attaching package: ‘igraph’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:dplyr’:\n",
      "\n",
      "    as_data_frame, groups, union\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    decompose, spectrum\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:base’:\n",
      "\n",
      "    union\n",
      "\n",
      "\n",
      "Loading required package: reticulate\n",
      "\n",
      "\n",
      "Attaching package: ‘future’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:igraph’:\n",
      "\n",
      "    %->%, %<-%\n",
      "\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[1] ‘4.1.3’"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "[1] ‘4.3.0’"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "[1] ‘1.6.3’"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "#remove.packages(c(\"Seurat\",\"SeuratObject\",\"Matrix\"))\n",
    "#install.packages(\"remotes\")\n",
    "#remotes::install_version(\"Matrix\", version = \"1.6.3\")   \n",
    "#remotes::install_version(\"SeuratObject\", version = \"4.1.3\", dependencies = FALSE)\n",
    "#remotes::install_version(\"Seurat\", version = \"4.3.0\") #BUT MAKE SURE YOU SELECT NONE WHEN IT ASK WHICH PACKAGES TP UPDATE/UPGRADE\n",
    "\n",
    "library(CellTrek)\n",
    "library(dplyr)\n",
    "library(Seurat)\n",
    "#library(SeuratDisk)\n",
    "library(viridis)\n",
    "library(ConsensusClusterPlus)\n",
    "library(RColorBrewer)\n",
    "library(ggplot2)\n",
    "library(ggpubr)\n",
    "library(pheatmap)\n",
    "library(CellChat)\n",
    "library(patchwork)\n",
    "library(sceasy)\n",
    "library(future)\n",
    "library(future.apply)\n",
    "\n",
    "packageVersion(\"SeuratObject\")  # 4.1.3\n",
    "packageVersion(\"Seurat\")        # 4.3.0\n",
    "packageVersion(\"Matrix\")  # 1.6.3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c5fba913",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "#to clear up session at any point\n",
    "list=ls()\n",
    "list\n",
    "rm(list); gc()\n",
    "#gc()\n",
    "#graphics.off()\n",
    "#closeAllConnections()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1d21950d",
   "metadata": {},
   "source": [
    "## 2. Load input data\n",
    "Loads the **Visium spatial transcriptomics** object and the **scRNA-seq** Seurat object. In this script:\n",
    "- `ovVis` = Visium data (early-gen, spot-based)\n",
    "- `ovSc` = single-cell RNA-seq data (annotated with cell types).\n",
    "\n",
    "Each of the datasets have already been filtered and subset to contain only spots/cells or genes that pass the filters. These subset versions of the datasets are loaded and used here but we use the raw counts (for the cells/spots remaining) for the workflow as required by CellTrek."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f566f868",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "#cluster\n",
    "cell2loc = '/mnt/scratchc/fmlab/lythgo02/OV_visium/emily/cell2location/'\n",
    "\n",
    "\n",
    "ovVis <- readRDS(paste0(cell2loc, \"mito_norm_emily.rds\"))\n",
    "ovSc <- readRDS(paste0(cell2loc, \"upk10_sc_400_newlyAnnot.rds\"))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f13db27f",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "\n",
    "#local\n",
    "cell2loc = '/run/user/1804238067/gvfs/sftp:host=clust1-sub-1,user=lythgo02/mnt/scratchc/fmlab/lythgo02/OV_visium/emily/cell2location/'\n",
    "ov_visium = '/run/user/1804238067/gvfs/sftp:host=clust1-sub-1,user=lythgo02/mnt/nas-data/fmlab/group_folders/lythgo02/OV_visium/emily/cell2location/'\n",
    "\n",
    "#### Load Visium Data\n",
    "ovVis <- readRDS(paste0(ov_visium, \"/mito_norm_emily.rds\"))  #version of Ollie's filtered against mine\n",
    "#ovVis <- readRDS(\"/home/lythgo02/Documents/OV_visium/final_adata_vis_clean_seurat.rds\")\n",
    "\n",
    "#### Load scRNA-seq\n",
    "ovSc <- readRDS(\"/home/lythgo02/Documents/OV_visium/upk10_sc_400_newlyAnnot.rds\")\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0c245af1",
   "metadata": {},
   "source": [
    "If using my rds list the cell below is a method to convert the images slot to an actual SeuratImage"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "78991b91",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "ovVis <- readRDS(\"/home/lythgo02/Documents/OV_visium/split_by_sample/rds/seurat_list_all.rds\")\n",
    "\n",
    "if (!methods::isClass(\"SeuratImage\")) {\n",
    "  methods::setClass(\n",
    "    \"SeuratImage\",\n",
    "    slots = c(\n",
    "      image         = \"ANY\",\n",
    "      scale.factors = \"list\",\n",
    "      coordinates   = \"data.frame\",\n",
    "      assay         = \"character\",\n",
    "      key           = \"character\"\n",
    "    )\n",
    "  )\n",
    "}\n",
    "\n",
    "for (smp in names(ovVis)) {\n",
    "  obj      <- ovVis[[smp]]\n",
    "  coords   <- obj@misc$spatial_coords[[smp]]\n",
    "  img_info <- obj@misc$spatial_images[[smp]]\n",
    "\n",
    "  # coords -> Seurat style, aligned to cells\n",
    "  colnames(coords)[1:2] <- c(\"imagecol\",\"imagerow\")\n",
    "  coords <- coords[match(colnames(obj), rownames(coords)), , drop = FALSE]\n",
    "  rownames(coords) <- colnames(obj)\n",
    "\n",
    "  # build minimal SeuratImage S4 (class defined earlier)\n",
    "  simg <- methods::new(\"SeuratImage\")\n",
    "  simg@image         <- as.array(img_info$image)      # or img_info$images$hires\n",
    "  simg@scale.factors <- img_info$scalef               # or img_info$scale.factors\n",
    "  simg@coordinates   <- coords\n",
    "  simg@assay         <- DefaultAssay(obj)\n",
    "  simg@key           <- paste0(smp, \"_\")\n",
    "\n",
    "  # >>> attach directly to the images slot (avoid [[<-)\n",
    "  if (is.null(obj@images)) obj@images <- list()\n",
    "  obj@images[[smp]] <- simg\n",
    "\n",
    "  ovVis[[smp]] <- obj\n",
    "}\n",
    "\n",
    "# sanity\n",
    "lapply(ovVis, function(o) class(o@images[[1]]))  # should print \"SeuratImage\"\n",
    "# 1) Make sure each spatial object uses RNA as its default assay\n",
    "\n",
    "for (i in names(ovVis)){\n",
    "    obj <- ovVis[[i]]\n",
    "    obj[[\"RNA\"]] <- obj[[\"originalexp\"]]\n",
    "    DefaultAssay(obj) <- \"RNA\"\n",
    "    obj[[\"originalexp\"]] <- NULL\n",
    "    ovVis[[i]] <- obj\n",
    "}\n",
    "ovVis <- lapply(ovVis, function(o) { DefaultAssay(o) <- \"RNA\"; o })\n",
    "\n",
    "# 2) If you created SeuratImage manually, ensure its assay slot is \"RNA\"\n",
    "for (smp in names(ovVis)) {\n",
    "  im <- ovVis[[smp]]@images[[1]]\n",
    "  if (!is.null(im)) im@assay <- \"RNA\"\n",
    "  ovVis[[smp]]@images[[1]] <- im\n",
    "}\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e1ee0db9",
   "metadata": {},
   "source": [
    "Need to rename assays, note when converted to seurat object earlier (code in markdown cell run in r_spatial env) the \"raw_counts\" and \"X\" is stored under \"originalexp\" in the seurat objects \n",
    "\n",
    "For spatial we did:   \n",
    "\n",
    "seurat_obj <- as.Seurat(\n",
    "            adata,              # your SingleCellExperiment object\n",
    "            counts = \"raw_counts\", # assay to use as raw counts\n",
    "            data = \"X\" # optional: assay to use as normalized data\n",
    ")\n",
    "3. Extract sample per spot\n",
    "sample <- colData(adata)$sample\n",
    "seurat_obj@meta.data$sample <- sample\n",
    "\n",
    "4. Extract coordinates and image info\n",
    "coords <- reducedDim(adata, \"spatial\")\n",
    "coords_list <- split(as.data.frame(coords), sample)\n",
    "image_list <- metadata(adata)$spatial\n",
    "\n",
    "\n",
    "5. Assign images and coordinates to Seurat in a CellTrek-compatible way\n",
    "for (sample_name in names(coords_list)) {\n",
    "  sp_coords <- coords_list[[sample_name]]\n",
    "  img_info <- image_list[[sample_name]]\n",
    "\n",
    "  seurat_obj@images[[sample_name]] <- list(\n",
    "    image = as.array(img_info$image),\n",
    "    scale.factors = img_info$scalef,\n",
    "    coordinates = sp_coords\n",
    "  )\n",
    "}\n",
    "1. Store original coordinates for reference\n",
    "seurat_obj@misc$spatial_coords <- coords_list\n",
    "seurat_obj@misc$spatial_images <- image_list\n",
    "\n",
    "For single cell we did:  \n",
    "\n",
    "seurat_obj <- as.Seurat(\n",
    "       singlecell,              \n",
    "       your SingleCellExperiment object#  counts = \"counts\", # assay to use as raw counts\n",
    "       data = \"log_counts\" # optional: assay to use as normalized data\n",
    ")\n",
    "\n",
    "\n",
    "so:  \n",
    "\n",
    "ovVis[[\"originalexp\"]]$counts # = raw_counts \n",
    "ovVis[[\"originalexp\"]]$data #= X\n",
    "\n",
    "ovSc[[\"originalexp\"]]$counts # = counts \n",
    "ovSc[[\"originalexp\"]]$data #= log_counts\n",
    "\n",
    "But mine lacks the imaging info stored in correct format so filtering Ollie's against mine below"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "01f122c9",
   "metadata": {},
   "source": [
    "The following silenced chunks are for filtering Ollie's mito_nrom.rds by subsetting to match the colnames (spot IDs) and rownames (genes) present in my rds\n",
    " - can't convert the adata to rds in the correct format so taking format from ollie's with processing from mine \n",
    " - saved the end product of subsetting as mito_nrom_emily.rds and that is now loaded at the beginning "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c7b70a7c",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "#ovVis <- readRDS(paste0(ov_visium, \"/mito_norm_emily.rds\"))\n",
    "# 1) Split emily object by sample ID (assuming \"orig.ident\" marks samples)\n",
    "# adjust 'split.by' if your sample column is named differently\n",
    "#emily_list <- SplitObject(ovVis_emily, split.by = \"sample\")\n",
    "#cells_ollie <- lapply(ovVis_ollie, colnames)\n",
    "\n",
    "# Get barcodes from each Seurat object in em_list\n",
    "#cells_emily <- lapply(emily_list, colnames)\n",
    "\n",
    "#head(cells_emily[[1]])\n",
    "#head(cells_ollie[[1]])\n",
    "#need to append sample name to barcodes in Ollie spatial data \n",
    "\n",
    "# (Optional) make a backup first\n",
    "#ovVis_ollie_orig <- ovVis_ollie\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f254d1ee",
   "metadata": {},
   "source": [
    "Need to append sample name to the colnames/spot barcodes (id) in Ollie's rds to match the formatting of mine"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "156cde19",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "#for (i in names(ovVis_ollie)){\n",
    "#    obj <- ovVis_ollie[[i]]  #overwrite names\n",
    "#    colnames(obj) <- paste0(i, \"_\", colnames(obj))\n",
    "#    ovVis_ollie[[i]] <- obj #save back to list\n",
    "#}\n",
    "# Check one sample to confirm\n",
    "#head(colnames(ovVis_ollie[[\"OV_1\"]]))\n",
    "\n",
    "# 2) Match sample names between ollie list and split emily\n",
    "#common_samples <- intersect(names(ovVis_ollie), names(emily_list))\n",
    "\n",
    "#common_samples\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "36471258",
   "metadata": {},
   "source": [
    "Now filter Ollie's spatial data by the list of spot id and genes present in my sample \n",
    " - to ensure the filtering applied is consisted accross workflows worked on by me\n",
    " - spots are filtered to match exactly\n",
    " - genes are filtered to approximately match, exact numbers may differ due to naming conventions but assuming largely correct "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d83db049",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "#ovVis_ollie_filtered <- lapply(common_samples, function(i) {\n",
    "#  obj <- DietSeurat(ovVis_ollie[[i]], graphs = TRUE) # remove the NN graphs ollie previously generated as they can disrupt subsetting\n",
    "#  # --- cells ---\n",
    "#  emily_barcodes <- colnames(emily_list[[i]])  #get list of spot ids from my rds (which has filtered out low quality spots)\n",
    "#  cells_keep <- intersect(colnames(obj), emily_barcodes) #keep corresponding cells in Ollie's object \n",
    "#  obj <- subset(obj, cells = cells_keep)\n",
    "#  # --- genes ---\n",
    "#  emily_genes <- rownames(emily_list[[i]])  #get gene names remaining after filtering \n",
    "#  genes_keep <- intersect(rownames(obj), emily_genes)\n",
    "  obj <- subset(obj, features = genes_keep)\n",
    "#})\n",
    "\n",
    "#names(ovVis_ollie_filtered) <- common_samples\n",
    "\n",
    "#sapply(ovVis_ollie_filtered, ncol)\n",
    "#sapply(emily_list, ncol)\n",
    "#sapply(ovVis_ollie_filtered, nrow)\n",
    "#sapply(emily_list, nrow)\n",
    "\n",
    "#saveRDS(ovVis_ollie_filtered, paste0(ov_visium, \"/mito_norm_emily.rds\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "77d0ecb1",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "“Cannot add objects with duplicate keys (offending key: originalexp_), setting key to 'rna_'”\n"
     ]
    }
   ],
   "source": [
    "# Copy originalexp → RNA and set RNA as default\n",
    "ovSc[[\"RNA\"]] <- ovSc[[\"originalexp\"]] #copy\n",
    "DefaultAssay(ovSc) <- \"RNA\" #set to default\n",
    "ovSc[[\"originalexp\"]] <- NULL  # Remove the old assay\n",
    "\n",
    "#not needed unless using my rds\n",
    "#for (i in names(ovVis)){\n",
    "#    obj <- ovVis[[i]]\n",
    "#    obj[[\"RNA\"]] <- obj[[\"originalexp\"]]\n",
    "#    DefaultAssay(obj) <- \"RNA\"\n",
    "#    obj[[\"originalexp\"]] <- NULL\n",
    "#    ovVis[[i]] <- obj\n",
    "#}\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d76cd8ca",
   "metadata": {},
   "source": [
    "Note first had to convert the outputs from cell2location workflow to rds objects:\n",
    " - activate r environment \n",
    "   - mamba activate r_spatial\n",
    "   - open R\n",
    "   - \n",
    "library(zellkonverter)\n",
    "adata <- readH5AD(\"/run/user/1804238067/gvfs/sftp:host=clust1-sub-1,user=lythgo02/mnt/scratchc/fmlab/lythgo02/OV_visium/emily/cell2location/final_adata_vis_clean.h5ad\")\n",
    "# Convert SCE to Seurat\n",
    "seurat_obj <- as.Seurat(\n",
    "  adata,              # your Spatial/singleCellExperiment object\n",
    "  counts = \"raw_counts\", # assay to use as raw counts\n",
    "  data = \"X\" # make sure to use X as this contains filtered data \n",
    ")\n",
    "saveRDS(adata, file=\"/run/user/1804238067/gvfs/sftp:host=clust1-sub-1,user=lythgo02/mnt/scratchc/fmlab/lythgo02/OV_visium/emily/cell2location/final_adata_vis_clean.rds\")\n",
    "\n",
    "#same with single cell data\n",
    "#write to local directory if it doesn't allow you to write to cluster (h5ad files can be tricky)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "0ee7ec3b",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'orig.ident'</li><li>'nCount_originalexp'</li><li>'nFeature_originalexp'</li><li>'nCount_RNA'</li><li>'nFeature_RNA'</li><li>'percentMt'</li><li>'nCount_SCT'</li><li>'nFeature_SCT'</li><li>'SCT_snn_res.1'</li><li>'seurat_clusters'</li><li>'pANN_0.25_0.19_161'</li><li>'DF.classifications_0.25_0.19_161'</li><li>'doublet_classification'</li><li>'CellAssign'</li><li>'Treatment'</li><li>'Cell_Type'</li><li>'newAnnot'</li><li>'newAnnot_1'</li><li>'inferCNV_Score'</li><li>'inferCNV_Class'</li><li>'Final_Annotation'</li><li>'CopyKat'</li><li>'newAnnot_2'</li><li>'pANN_0.25_0.05_197'</li><li>'DF.classifications_0.25_0.05_197'</li><li>'pANN_0.25_0.19_227'</li><li>'DF.classifications_0.25_0.19_227'</li><li>'pANN_0.25_0.06_117'</li><li>'DF.classifications_0.25_0.06_117'</li><li>'pANN_0.25_0.28_314'</li><li>'DF.classifications_0.25_0.28_314'</li><li>'SCT_snn_res.0.5'</li><li>'ident'</li><li>'updated_annotation'</li><li>'X_indices'</li><li>'X_scvi_batch'</li><li>'X_scvi_labels'</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 'orig.ident'\n",
       "\\item 'nCount\\_originalexp'\n",
       "\\item 'nFeature\\_originalexp'\n",
       "\\item 'nCount\\_RNA'\n",
       "\\item 'nFeature\\_RNA'\n",
       "\\item 'percentMt'\n",
       "\\item 'nCount\\_SCT'\n",
       "\\item 'nFeature\\_SCT'\n",
       "\\item 'SCT\\_snn\\_res.1'\n",
       "\\item 'seurat\\_clusters'\n",
       "\\item 'pANN\\_0.25\\_0.19\\_161'\n",
       "\\item 'DF.classifications\\_0.25\\_0.19\\_161'\n",
       "\\item 'doublet\\_classification'\n",
       "\\item 'CellAssign'\n",
       "\\item 'Treatment'\n",
       "\\item 'Cell\\_Type'\n",
       "\\item 'newAnnot'\n",
       "\\item 'newAnnot\\_1'\n",
       "\\item 'inferCNV\\_Score'\n",
       "\\item 'inferCNV\\_Class'\n",
       "\\item 'Final\\_Annotation'\n",
       "\\item 'CopyKat'\n",
       "\\item 'newAnnot\\_2'\n",
       "\\item 'pANN\\_0.25\\_0.05\\_197'\n",
       "\\item 'DF.classifications\\_0.25\\_0.05\\_197'\n",
       "\\item 'pANN\\_0.25\\_0.19\\_227'\n",
       "\\item 'DF.classifications\\_0.25\\_0.19\\_227'\n",
       "\\item 'pANN\\_0.25\\_0.06\\_117'\n",
       "\\item 'DF.classifications\\_0.25\\_0.06\\_117'\n",
       "\\item 'pANN\\_0.25\\_0.28\\_314'\n",
       "\\item 'DF.classifications\\_0.25\\_0.28\\_314'\n",
       "\\item 'SCT\\_snn\\_res.0.5'\n",
       "\\item 'ident'\n",
       "\\item 'updated\\_annotation'\n",
       "\\item 'X\\_indices'\n",
       "\\item 'X\\_scvi\\_batch'\n",
       "\\item 'X\\_scvi\\_labels'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 'orig.ident'\n",
       "2. 'nCount_originalexp'\n",
       "3. 'nFeature_originalexp'\n",
       "4. 'nCount_RNA'\n",
       "5. 'nFeature_RNA'\n",
       "6. 'percentMt'\n",
       "7. 'nCount_SCT'\n",
       "8. 'nFeature_SCT'\n",
       "9. 'SCT_snn_res.1'\n",
       "10. 'seurat_clusters'\n",
       "11. 'pANN_0.25_0.19_161'\n",
       "12. 'DF.classifications_0.25_0.19_161'\n",
       "13. 'doublet_classification'\n",
       "14. 'CellAssign'\n",
       "15. 'Treatment'\n",
       "16. 'Cell_Type'\n",
       "17. 'newAnnot'\n",
       "18. 'newAnnot_1'\n",
       "19. 'inferCNV_Score'\n",
       "20. 'inferCNV_Class'\n",
       "21. 'Final_Annotation'\n",
       "22. 'CopyKat'\n",
       "23. 'newAnnot_2'\n",
       "24. 'pANN_0.25_0.05_197'\n",
       "25. 'DF.classifications_0.25_0.05_197'\n",
       "26. 'pANN_0.25_0.19_227'\n",
       "27. 'DF.classifications_0.25_0.19_227'\n",
       "28. 'pANN_0.25_0.06_117'\n",
       "29. 'DF.classifications_0.25_0.06_117'\n",
       "30. 'pANN_0.25_0.28_314'\n",
       "31. 'DF.classifications_0.25_0.28_314'\n",
       "32. 'SCT_snn_res.0.5'\n",
       "33. 'ident'\n",
       "34. 'updated_annotation'\n",
       "35. 'X_indices'\n",
       "36. 'X_scvi_batch'\n",
       "37. 'X_scvi_labels'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       " [1] \"orig.ident\"                       \"nCount_originalexp\"              \n",
       " [3] \"nFeature_originalexp\"             \"nCount_RNA\"                      \n",
       " [5] \"nFeature_RNA\"                     \"percentMt\"                       \n",
       " [7] \"nCount_SCT\"                       \"nFeature_SCT\"                    \n",
       " [9] \"SCT_snn_res.1\"                    \"seurat_clusters\"                 \n",
       "[11] \"pANN_0.25_0.19_161\"               \"DF.classifications_0.25_0.19_161\"\n",
       "[13] \"doublet_classification\"           \"CellAssign\"                      \n",
       "[15] \"Treatment\"                        \"Cell_Type\"                       \n",
       "[17] \"newAnnot\"                         \"newAnnot_1\"                      \n",
       "[19] \"inferCNV_Score\"                   \"inferCNV_Class\"                  \n",
       "[21] \"Final_Annotation\"                 \"CopyKat\"                         \n",
       "[23] \"newAnnot_2\"                       \"pANN_0.25_0.05_197\"              \n",
       "[25] \"DF.classifications_0.25_0.05_197\" \"pANN_0.25_0.19_227\"              \n",
       "[27] \"DF.classifications_0.25_0.19_227\" \"pANN_0.25_0.06_117\"              \n",
       "[29] \"DF.classifications_0.25_0.06_117\" \"pANN_0.25_0.28_314\"              \n",
       "[31] \"DF.classifications_0.25_0.28_314\" \"SCT_snn_res.0.5\"                 \n",
       "[33] \"ident\"                            \"updated_annotation\"              \n",
       "[35] \"X_indices\"                        \"X_scvi_batch\"                    \n",
       "[37] \"X_scvi_labels\"                   "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>OV_16</li><li>OV_17</li><li>OV_18</li><li>OV_19</li><li>OV_20</li></ol>\n",
       "\n",
       "<details>\n",
       "\t<summary style=display:list-item;cursor:pointer>\n",
       "\t\t<strong>Levels</strong>:\n",
       "\t</summary>\n",
       "\t<style>\n",
       "\t.list-inline {list-style: none; margin:0; padding: 0}\n",
       "\t.list-inline>li {display: inline-block}\n",
       "\t.list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "\t</style>\n",
       "\t<ol class=list-inline><li>'OV_16'</li><li>'OV_17'</li><li>'OV_18'</li><li>'OV_19'</li><li>'OV_20'</li></ol>\n",
       "</details>"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item OV\\_16\n",
       "\\item OV\\_17\n",
       "\\item OV\\_18\n",
       "\\item OV\\_19\n",
       "\\item OV\\_20\n",
       "\\end{enumerate*}\n",
       "\n",
       "\\emph{Levels}: \\begin{enumerate*}\n",
       "\\item 'OV\\_16'\n",
       "\\item 'OV\\_17'\n",
       "\\item 'OV\\_18'\n",
       "\\item 'OV\\_19'\n",
       "\\item 'OV\\_20'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. OV_16\n",
       "2. OV_17\n",
       "3. OV_18\n",
       "4. OV_19\n",
       "5. OV_20\n",
       "\n",
       "\n",
       "\n",
       "**Levels**: 1. 'OV_16'\n",
       "2. 'OV_17'\n",
       "3. 'OV_18'\n",
       "4. 'OV_19'\n",
       "5. 'OV_20'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] OV_16 OV_17 OV_18 OV_19 OV_20\n",
       "Levels: OV_16 OV_17 OV_18 OV_19 OV_20"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "'RNA'"
      ],
      "text/latex": [
       "'RNA'"
      ],
      "text/markdown": [
       "'RNA'"
      ],
      "text/plain": [
       "[1] \"RNA\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>Tumour 1</li><li>Macrophage</li><li>NK_cell</li><li>Treg</li><li>Tumour 2</li><li>Plasma Cells</li><li>CD4</li><li>CD8</li><li>B naive activated</li><li>Mesenchymal</li><li>B naive resting</li><li>Monocytes</li><li>Dendritic Cells</li><li>Neutrophils</li><li>Th1</li><li>Th17</li><li>Endothelial</li><li>Tumour 3</li><li>Cycling Plasma Cells</li></ol>\n",
       "\n",
       "<details>\n",
       "\t<summary style=display:list-item;cursor:pointer>\n",
       "\t\t<strong>Levels</strong>:\n",
       "\t</summary>\n",
       "\t<style>\n",
       "\t.list-inline {list-style: none; margin:0; padding: 0}\n",
       "\t.list-inline>li {display: inline-block}\n",
       "\t.list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "\t</style>\n",
       "\t<ol class=list-inline><li>'B naive activated'</li><li>'B naive resting'</li><li>'CD4'</li><li>'CD8'</li><li>'Cycling Plasma Cells'</li><li>'Dendritic Cells'</li><li>'Endothelial'</li><li>'Macrophage'</li><li>'Mesenchymal'</li><li>'Monocytes'</li><li>'NK_cell'</li><li>'Neutrophils'</li><li>'Plasma Cells'</li><li>'Th1'</li><li>'Th17'</li><li>'Treg'</li><li>'Tumour 1'</li><li>'Tumour 2'</li><li>'Tumour 3'</li></ol>\n",
       "</details>"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item Tumour 1\n",
       "\\item Macrophage\n",
       "\\item NK\\_cell\n",
       "\\item Treg\n",
       "\\item Tumour 2\n",
       "\\item Plasma Cells\n",
       "\\item CD4\n",
       "\\item CD8\n",
       "\\item B naive activated\n",
       "\\item Mesenchymal\n",
       "\\item B naive resting\n",
       "\\item Monocytes\n",
       "\\item Dendritic Cells\n",
       "\\item Neutrophils\n",
       "\\item Th1\n",
       "\\item Th17\n",
       "\\item Endothelial\n",
       "\\item Tumour 3\n",
       "\\item Cycling Plasma Cells\n",
       "\\end{enumerate*}\n",
       "\n",
       "\\emph{Levels}: \\begin{enumerate*}\n",
       "\\item 'B naive activated'\n",
       "\\item 'B naive resting'\n",
       "\\item 'CD4'\n",
       "\\item 'CD8'\n",
       "\\item 'Cycling Plasma Cells'\n",
       "\\item 'Dendritic Cells'\n",
       "\\item 'Endothelial'\n",
       "\\item 'Macrophage'\n",
       "\\item 'Mesenchymal'\n",
       "\\item 'Monocytes'\n",
       "\\item 'NK\\_cell'\n",
       "\\item 'Neutrophils'\n",
       "\\item 'Plasma Cells'\n",
       "\\item 'Th1'\n",
       "\\item 'Th17'\n",
       "\\item 'Treg'\n",
       "\\item 'Tumour 1'\n",
       "\\item 'Tumour 2'\n",
       "\\item 'Tumour 3'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. Tumour 1\n",
       "2. Macrophage\n",
       "3. NK_cell\n",
       "4. Treg\n",
       "5. Tumour 2\n",
       "6. Plasma Cells\n",
       "7. CD4\n",
       "8. CD8\n",
       "9. B naive activated\n",
       "10. Mesenchymal\n",
       "11. B naive resting\n",
       "12. Monocytes\n",
       "13. Dendritic Cells\n",
       "14. Neutrophils\n",
       "15. Th1\n",
       "16. Th17\n",
       "17. Endothelial\n",
       "18. Tumour 3\n",
       "19. Cycling Plasma Cells\n",
       "\n",
       "\n",
       "\n",
       "**Levels**: 1. 'B naive activated'\n",
       "2. 'B naive resting'\n",
       "3. 'CD4'\n",
       "4. 'CD8'\n",
       "5. 'Cycling Plasma Cells'\n",
       "6. 'Dendritic Cells'\n",
       "7. 'Endothelial'\n",
       "8. 'Macrophage'\n",
       "9. 'Mesenchymal'\n",
       "10. 'Monocytes'\n",
       "11. 'NK_cell'\n",
       "12. 'Neutrophils'\n",
       "13. 'Plasma Cells'\n",
       "14. 'Th1'\n",
       "15. 'Th17'\n",
       "16. 'Treg'\n",
       "17. 'Tumour 1'\n",
       "18. 'Tumour 2'\n",
       "19. 'Tumour 3'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       " [1] Tumour 1             Macrophage           NK_cell             \n",
       " [4] Treg                 Tumour 2             Plasma Cells        \n",
       " [7] CD4                  CD8                  B naive activated   \n",
       "[10] Mesenchymal          B naive resting      Monocytes           \n",
       "[13] Dendritic Cells      Neutrophils          Th1                 \n",
       "[16] Th17                 Endothelial          Tumour 3            \n",
       "[19] Cycling Plasma Cells\n",
       "19 Levels: B naive activated B naive resting CD4 CD8 ... Tumour 3"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "#check if integrated\n",
    "\n",
    "colnames(ovSc@meta.data)\n",
    "\n",
    "#sample names for single cell \n",
    "unique(ovSc@meta.data$orig.ident)\n",
    "ovSc@active.assay \n",
    "\n",
    "unique(ovSc$updated_annotation)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "55fddc86",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<dl>\n",
       "\t<dt>$OV_1</dt>\n",
       "\t\t<dd><style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'assays'</li><li>'meta.data'</li><li>'active.assay'</li><li>'active.ident'</li><li>'graphs'</li><li>'neighbors'</li><li>'reductions'</li><li>'images'</li><li>'project.name'</li><li>'misc'</li><li>'version'</li><li>'commands'</li><li>'tools'</li></ol>\n",
       "</dd>\n",
       "\t<dt>$OV_2</dt>\n",
       "\t\t<dd><style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'assays'</li><li>'meta.data'</li><li>'active.assay'</li><li>'active.ident'</li><li>'graphs'</li><li>'neighbors'</li><li>'reductions'</li><li>'images'</li><li>'project.name'</li><li>'misc'</li><li>'version'</li><li>'commands'</li><li>'tools'</li></ol>\n",
       "</dd>\n",
       "\t<dt>$OV_3</dt>\n",
       "\t\t<dd><style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'assays'</li><li>'meta.data'</li><li>'active.assay'</li><li>'active.ident'</li><li>'graphs'</li><li>'neighbors'</li><li>'reductions'</li><li>'images'</li><li>'project.name'</li><li>'misc'</li><li>'version'</li><li>'commands'</li><li>'tools'</li></ol>\n",
       "</dd>\n",
       "\t<dt>$OV_4</dt>\n",
       "\t\t<dd><style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'assays'</li><li>'meta.data'</li><li>'active.assay'</li><li>'active.ident'</li><li>'graphs'</li><li>'neighbors'</li><li>'reductions'</li><li>'images'</li><li>'project.name'</li><li>'misc'</li><li>'version'</li><li>'commands'</li><li>'tools'</li></ol>\n",
       "</dd>\n",
       "\t<dt>$OV_5</dt>\n",
       "\t\t<dd><style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'assays'</li><li>'meta.data'</li><li>'active.assay'</li><li>'active.ident'</li><li>'graphs'</li><li>'neighbors'</li><li>'reductions'</li><li>'images'</li><li>'project.name'</li><li>'misc'</li><li>'version'</li><li>'commands'</li><li>'tools'</li></ol>\n",
       "</dd>\n",
       "\t<dt>$OV_6</dt>\n",
       "\t\t<dd><style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'assays'</li><li>'meta.data'</li><li>'active.assay'</li><li>'active.ident'</li><li>'graphs'</li><li>'neighbors'</li><li>'reductions'</li><li>'images'</li><li>'project.name'</li><li>'misc'</li><li>'version'</li><li>'commands'</li><li>'tools'</li></ol>\n",
       "</dd>\n",
       "</dl>\n"
      ],
      "text/latex": [
       "\\begin{description}\n",
       "\\item[\\$OV\\_1] \\begin{enumerate*}\n",
       "\\item 'assays'\n",
       "\\item 'meta.data'\n",
       "\\item 'active.assay'\n",
       "\\item 'active.ident'\n",
       "\\item 'graphs'\n",
       "\\item 'neighbors'\n",
       "\\item 'reductions'\n",
       "\\item 'images'\n",
       "\\item 'project.name'\n",
       "\\item 'misc'\n",
       "\\item 'version'\n",
       "\\item 'commands'\n",
       "\\item 'tools'\n",
       "\\end{enumerate*}\n",
       "\n",
       "\\item[\\$OV\\_2] \\begin{enumerate*}\n",
       "\\item 'assays'\n",
       "\\item 'meta.data'\n",
       "\\item 'active.assay'\n",
       "\\item 'active.ident'\n",
       "\\item 'graphs'\n",
       "\\item 'neighbors'\n",
       "\\item 'reductions'\n",
       "\\item 'images'\n",
       "\\item 'project.name'\n",
       "\\item 'misc'\n",
       "\\item 'version'\n",
       "\\item 'commands'\n",
       "\\item 'tools'\n",
       "\\end{enumerate*}\n",
       "\n",
       "\\item[\\$OV\\_3] \\begin{enumerate*}\n",
       "\\item 'assays'\n",
       "\\item 'meta.data'\n",
       "\\item 'active.assay'\n",
       "\\item 'active.ident'\n",
       "\\item 'graphs'\n",
       "\\item 'neighbors'\n",
       "\\item 'reductions'\n",
       "\\item 'images'\n",
       "\\item 'project.name'\n",
       "\\item 'misc'\n",
       "\\item 'version'\n",
       "\\item 'commands'\n",
       "\\item 'tools'\n",
       "\\end{enumerate*}\n",
       "\n",
       "\\item[\\$OV\\_4] \\begin{enumerate*}\n",
       "\\item 'assays'\n",
       "\\item 'meta.data'\n",
       "\\item 'active.assay'\n",
       "\\item 'active.ident'\n",
       "\\item 'graphs'\n",
       "\\item 'neighbors'\n",
       "\\item 'reductions'\n",
       "\\item 'images'\n",
       "\\item 'project.name'\n",
       "\\item 'misc'\n",
       "\\item 'version'\n",
       "\\item 'commands'\n",
       "\\item 'tools'\n",
       "\\end{enumerate*}\n",
       "\n",
       "\\item[\\$OV\\_5] \\begin{enumerate*}\n",
       "\\item 'assays'\n",
       "\\item 'meta.data'\n",
       "\\item 'active.assay'\n",
       "\\item 'active.ident'\n",
       "\\item 'graphs'\n",
       "\\item 'neighbors'\n",
       "\\item 'reductions'\n",
       "\\item 'images'\n",
       "\\item 'project.name'\n",
       "\\item 'misc'\n",
       "\\item 'version'\n",
       "\\item 'commands'\n",
       "\\item 'tools'\n",
       "\\end{enumerate*}\n",
       "\n",
       "\\item[\\$OV\\_6] \\begin{enumerate*}\n",
       "\\item 'assays'\n",
       "\\item 'meta.data'\n",
       "\\item 'active.assay'\n",
       "\\item 'active.ident'\n",
       "\\item 'graphs'\n",
       "\\item 'neighbors'\n",
       "\\item 'reductions'\n",
       "\\item 'images'\n",
       "\\item 'project.name'\n",
       "\\item 'misc'\n",
       "\\item 'version'\n",
       "\\item 'commands'\n",
       "\\item 'tools'\n",
       "\\end{enumerate*}\n",
       "\n",
       "\\end{description}\n"
      ],
      "text/markdown": [
       "$OV_1\n",
       ":   1. 'assays'\n",
       "2. 'meta.data'\n",
       "3. 'active.assay'\n",
       "4. 'active.ident'\n",
       "5. 'graphs'\n",
       "6. 'neighbors'\n",
       "7. 'reductions'\n",
       "8. 'images'\n",
       "9. 'project.name'\n",
       "10. 'misc'\n",
       "11. 'version'\n",
       "12. 'commands'\n",
       "13. 'tools'\n",
       "\n",
       "\n",
       "\n",
       "$OV_2\n",
       ":   1. 'assays'\n",
       "2. 'meta.data'\n",
       "3. 'active.assay'\n",
       "4. 'active.ident'\n",
       "5. 'graphs'\n",
       "6. 'neighbors'\n",
       "7. 'reductions'\n",
       "8. 'images'\n",
       "9. 'project.name'\n",
       "10. 'misc'\n",
       "11. 'version'\n",
       "12. 'commands'\n",
       "13. 'tools'\n",
       "\n",
       "\n",
       "\n",
       "$OV_3\n",
       ":   1. 'assays'\n",
       "2. 'meta.data'\n",
       "3. 'active.assay'\n",
       "4. 'active.ident'\n",
       "5. 'graphs'\n",
       "6. 'neighbors'\n",
       "7. 'reductions'\n",
       "8. 'images'\n",
       "9. 'project.name'\n",
       "10. 'misc'\n",
       "11. 'version'\n",
       "12. 'commands'\n",
       "13. 'tools'\n",
       "\n",
       "\n",
       "\n",
       "$OV_4\n",
       ":   1. 'assays'\n",
       "2. 'meta.data'\n",
       "3. 'active.assay'\n",
       "4. 'active.ident'\n",
       "5. 'graphs'\n",
       "6. 'neighbors'\n",
       "7. 'reductions'\n",
       "8. 'images'\n",
       "9. 'project.name'\n",
       "10. 'misc'\n",
       "11. 'version'\n",
       "12. 'commands'\n",
       "13. 'tools'\n",
       "\n",
       "\n",
       "\n",
       "$OV_5\n",
       ":   1. 'assays'\n",
       "2. 'meta.data'\n",
       "3. 'active.assay'\n",
       "4. 'active.ident'\n",
       "5. 'graphs'\n",
       "6. 'neighbors'\n",
       "7. 'reductions'\n",
       "8. 'images'\n",
       "9. 'project.name'\n",
       "10. 'misc'\n",
       "11. 'version'\n",
       "12. 'commands'\n",
       "13. 'tools'\n",
       "\n",
       "\n",
       "\n",
       "$OV_6\n",
       ":   1. 'assays'\n",
       "2. 'meta.data'\n",
       "3. 'active.assay'\n",
       "4. 'active.ident'\n",
       "5. 'graphs'\n",
       "6. 'neighbors'\n",
       "7. 'reductions'\n",
       "8. 'images'\n",
       "9. 'project.name'\n",
       "10. 'misc'\n",
       "11. 'version'\n",
       "12. 'commands'\n",
       "13. 'tools'\n",
       "\n",
       "\n",
       "\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "$OV_1\n",
       " [1] \"assays\"       \"meta.data\"    \"active.assay\" \"active.ident\" \"graphs\"      \n",
       " [6] \"neighbors\"    \"reductions\"   \"images\"       \"project.name\" \"misc\"        \n",
       "[11] \"version\"      \"commands\"     \"tools\"       \n",
       "\n",
       "$OV_2\n",
       " [1] \"assays\"       \"meta.data\"    \"active.assay\" \"active.ident\" \"graphs\"      \n",
       " [6] \"neighbors\"    \"reductions\"   \"images\"       \"project.name\" \"misc\"        \n",
       "[11] \"version\"      \"commands\"     \"tools\"       \n",
       "\n",
       "$OV_3\n",
       " [1] \"assays\"       \"meta.data\"    \"active.assay\" \"active.ident\" \"graphs\"      \n",
       " [6] \"neighbors\"    \"reductions\"   \"images\"       \"project.name\" \"misc\"        \n",
       "[11] \"version\"      \"commands\"     \"tools\"       \n",
       "\n",
       "$OV_4\n",
       " [1] \"assays\"       \"meta.data\"    \"active.assay\" \"active.ident\" \"graphs\"      \n",
       " [6] \"neighbors\"    \"reductions\"   \"images\"       \"project.name\" \"misc\"        \n",
       "[11] \"version\"      \"commands\"     \"tools\"       \n",
       "\n",
       "$OV_5\n",
       " [1] \"assays\"       \"meta.data\"    \"active.assay\" \"active.ident\" \"graphs\"      \n",
       " [6] \"neighbors\"    \"reductions\"   \"images\"       \"project.name\" \"misc\"        \n",
       "[11] \"version\"      \"commands\"     \"tools\"       \n",
       "\n",
       "$OV_6\n",
       " [1] \"assays\"       \"meta.data\"    \"active.assay\" \"active.ident\" \"graphs\"      \n",
       " [6] \"neighbors\"    \"reductions\"   \"images\"       \"project.name\" \"misc\"        \n",
       "[11] \"version\"      \"commands\"     \"tools\"       \n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "#spatial info stored in images slot \n",
    "lapply(ovVis, slotNames)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "72c35eef",
   "metadata": {},
   "source": [
    "## 3. Integrate scRNA-seq samples\n",
    "Here multiple scRNA samples are merged and integrated using Seurat's anchor-based pipeline. This corrects for batch effects and aligns datasets into a shared space.\n",
    " - SelectIntegrationFeatures finds highly variable genes shared across all samples which are used as anchors for sample integration. \n",
    " - Anchors are pairs of cells across datasets that are similar in expression.\n",
    " - object.list = ovSc tells Seurat which samples to integrate.\n",
    " - anchor.features = features uses the previously selected genes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "c81408b4",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'OV_16'</li><li>'OV_17'</li><li>'OV_18'</li><li>'OV_19'</li><li>'OV_20'</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 'OV\\_16'\n",
       "\\item 'OV\\_17'\n",
       "\\item 'OV\\_18'\n",
       "\\item 'OV\\_19'\n",
       "\\item 'OV\\_20'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 'OV_16'\n",
       "2. 'OV_17'\n",
       "3. 'OV_18'\n",
       "4. 'OV_19'\n",
       "5. 'OV_20'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] \"OV_16\" \"OV_17\" \"OV_18\" \"OV_19\" \"OV_20\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "\n",
    "#already have single object with multiple samples \n",
    "# Split the Seurat object by sample and store in list of invdividual samples\n",
    "ovSc.list <- SplitObject(ovSc, split.by = \"orig.ident\")\n",
    "\n",
    "# Check the names\n",
    "names(ovSc.list)\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2740f822",
   "metadata": {},
   "source": [
    "Single cell is currently previous version 3/4 of suerat object which uses slots, need to convert to version 5 which uses layers \n",
    " - define function to convert to assay5"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "86029ff2",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "'Assay'"
      ],
      "text/latex": [
       "'Assay'"
      ],
      "text/markdown": [
       "'Assay'"
      ],
      "text/plain": [
       "[1] \"Assay\"\n",
       "attr(,\"package\")\n",
       "[1] \"SeuratObject\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".dl-inline {width: auto; margin:0; padding: 0}\n",
       ".dl-inline>dt, .dl-inline>dd {float: none; width: auto; display: inline-block}\n",
       ".dl-inline>dt::after {content: \":\\0020\"; padding-right: .5ex}\n",
       ".dl-inline>dt:not(:first-of-type) {padding-left: .5ex}\n",
       "</style><dl class=dl-inline><dt>OV_16</dt><dd>'Assay'</dd><dt>OV_17</dt><dd>'Assay'</dd><dt>OV_18</dt><dd>'Assay'</dd><dt>OV_19</dt><dd>'Assay'</dd><dt>OV_20</dt><dd>'Assay'</dd></dl>\n"
      ],
      "text/latex": [
       "\\begin{description*}\n",
       "\\item[OV\\textbackslash{}\\_16] 'Assay'\n",
       "\\item[OV\\textbackslash{}\\_17] 'Assay'\n",
       "\\item[OV\\textbackslash{}\\_18] 'Assay'\n",
       "\\item[OV\\textbackslash{}\\_19] 'Assay'\n",
       "\\item[OV\\textbackslash{}\\_20] 'Assay'\n",
       "\\end{description*}\n"
      ],
      "text/markdown": [
       "OV_16\n",
       ":   'Assay'OV_17\n",
       ":   'Assay'OV_18\n",
       ":   'Assay'OV_19\n",
       ":   'Assay'OV_20\n",
       ":   'Assay'\n",
       "\n"
      ],
      "text/plain": [
       "  OV_16   OV_17   OV_18   OV_19   OV_20 \n",
       "\"Assay\" \"Assay\" \"Assay\" \"Assay\" \"Assay\" "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "class(ovSc.list[[1]][[\"RNA\"]])\n",
    "sapply(ovSc.list, function(obj) class(obj[[\"RNA\"]]))\n",
    "#shows that each one of the RNA assays is in old Seurat v3/v4 style (slot based)\n",
    "#have got code to update to v5 in emails in needed\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "69ebea85",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "\n",
    "#### Iterate over samples in list to prepare for integration\n",
    "\n",
    "# Make sure every object has HVGs before integration\n",
    "ovSc.list <- lapply(ovSc.list, function(x){\n",
    "  DefaultAssay(x) <- \"RNA\"\n",
    "  # We already have 'counts' + 'data' layers from earlier conversion\n",
    "  # Find HVGs on log-normalised data\n",
    "  x <- NormalizeData(x)\n",
    "  x <- FindVariableFeatures(\n",
    "    x,                # use log-normalized values, not raw counts\n",
    "    selection.method = \"vst\",\n",
    "    nfeatures = 3000\n",
    "  )\n",
    "  x\n",
    "})\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "5182a239",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "\n",
    "# Select features and integrate\n",
    "features <- SelectIntegrationFeatures(object.list = ovSc.list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "2754510d",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Scaling features for provided objects\n",
      "\n",
      "Finding all pairwise anchors\n",
      "\n",
      "Running CCA\n",
      "\n",
      "Merging objects\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 14458 anchors\n",
      "\n",
      "Filtering anchors\n",
      "\n",
      "\tRetained 12442 anchors\n",
      "\n",
      "Running CCA\n",
      "\n",
      "Merging objects\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 14428 anchors\n",
      "\n",
      "Filtering anchors\n",
      "\n",
      "\tRetained 11850 anchors\n",
      "\n",
      "Running CCA\n",
      "\n",
      "Merging objects\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 15424 anchors\n",
      "\n",
      "Filtering anchors\n",
      "\n",
      "\tRetained 12501 anchors\n",
      "\n",
      "Running CCA\n",
      "\n",
      "Merging objects\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 6999 anchors\n",
      "\n",
      "Filtering anchors\n",
      "\n",
      "\tRetained 3423 anchors\n",
      "\n",
      "Running CCA\n",
      "\n",
      "Merging objects\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 6999 anchors\n",
      "\n",
      "Filtering anchors\n",
      "\n",
      "\tRetained 3308 anchors\n",
      "\n",
      "Running CCA\n",
      "\n",
      "Merging objects\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 6960 anchors\n",
      "\n",
      "Filtering anchors\n",
      "\n",
      "\tRetained 3252 anchors\n",
      "\n",
      "Running CCA\n",
      "\n",
      "Merging objects\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 15534 anchors\n",
      "\n",
      "Filtering anchors\n",
      "\n",
      "\tRetained 13010 anchors\n",
      "\n",
      "Running CCA\n",
      "\n",
      "Merging objects\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 15571 anchors\n",
      "\n",
      "Filtering anchors\n",
      "\n",
      "\tRetained 12203 anchors\n",
      "\n",
      "Running CCA\n",
      "\n",
      "Merging objects\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 15726 anchors\n",
      "\n",
      "Filtering anchors\n",
      "\n",
      "\tRetained 12335 anchors\n",
      "\n",
      "Running CCA\n",
      "\n",
      "Merging objects\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 7090 anchors\n",
      "\n",
      "Filtering anchors\n",
      "\n",
      "\tRetained 2828 anchors\n",
      "\n"
     ]
    }
   ],
   "source": [
    "anchors <- FindIntegrationAnchors(object.list = ovSc.list, \n",
    "    normalization.method = \"LogNormalize\",\n",
    "    anchor.features = features)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "61c9c277",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Merging dataset 1 into 5\n",
      "\n",
      "Extracting anchors for merged samples\n",
      "\n",
      "Finding integration vectors\n",
      "\n",
      "Finding integration vector weights\n",
      "\n",
      "Integrating data\n",
      "\n",
      "Merging dataset 2 into 3\n",
      "\n",
      "Extracting anchors for merged samples\n",
      "\n",
      "Finding integration vectors\n",
      "\n",
      "Finding integration vector weights\n",
      "\n",
      "Integrating data\n",
      "\n",
      "Merging dataset 3 2 into 5 1\n",
      "\n",
      "Extracting anchors for merged samples\n",
      "\n",
      "Finding integration vectors\n",
      "\n",
      "Finding integration vector weights\n",
      "\n",
      "Integrating data\n",
      "\n",
      "Merging dataset 4 into 5 1 3 2\n",
      "\n",
      "Extracting anchors for merged samples\n",
      "\n",
      "Finding integration vectors\n",
      "\n",
      "Finding integration vector weights\n",
      "\n",
      "Integrating data\n",
      "\n"
     ]
    }
   ],
   "source": [
    "seuCombined <- IntegrateData(anchorset = anchors)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4bcd7ab1",
   "metadata": {},
   "source": [
    "## 4. Subset to cell types of interest\n",
    "Selects specific immune/stromal/tumor cell types from the integrated scRNA data.\n",
    "Not subsetting because already filtered the dataset in cell2location pipeline to exclude those we aren't interested in. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "27107f3a",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'Tumour 1'</li><li>'Macrophage'</li><li>'NK_cell'</li><li>'Treg'</li><li>'Tumour 2'</li><li>'Plasma Cells'</li><li>'CD4'</li><li>'CD8'</li><li>'B naive activated'</li><li>'Mesenchymal'</li><li>'B naive resting'</li><li>'Monocytes'</li><li>'Dendritic Cells'</li><li>'Neutrophils'</li><li>'Th1'</li><li>'Th17'</li><li>'Endothelial'</li><li>'Tumour 3'</li><li>'Cycling Plasma Cells'</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 'Tumour 1'\n",
       "\\item 'Macrophage'\n",
       "\\item 'NK\\_cell'\n",
       "\\item 'Treg'\n",
       "\\item 'Tumour 2'\n",
       "\\item 'Plasma Cells'\n",
       "\\item 'CD4'\n",
       "\\item 'CD8'\n",
       "\\item 'B naive activated'\n",
       "\\item 'Mesenchymal'\n",
       "\\item 'B naive resting'\n",
       "\\item 'Monocytes'\n",
       "\\item 'Dendritic Cells'\n",
       "\\item 'Neutrophils'\n",
       "\\item 'Th1'\n",
       "\\item 'Th17'\n",
       "\\item 'Endothelial'\n",
       "\\item 'Tumour 3'\n",
       "\\item 'Cycling Plasma Cells'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 'Tumour 1'\n",
       "2. 'Macrophage'\n",
       "3. 'NK_cell'\n",
       "4. 'Treg'\n",
       "5. 'Tumour 2'\n",
       "6. 'Plasma Cells'\n",
       "7. 'CD4'\n",
       "8. 'CD8'\n",
       "9. 'B naive activated'\n",
       "10. 'Mesenchymal'\n",
       "11. 'B naive resting'\n",
       "12. 'Monocytes'\n",
       "13. 'Dendritic Cells'\n",
       "14. 'Neutrophils'\n",
       "15. 'Th1'\n",
       "16. 'Th17'\n",
       "17. 'Endothelial'\n",
       "18. 'Tumour 3'\n",
       "19. 'Cycling Plasma Cells'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       " [1] \"Tumour 1\"             \"Macrophage\"           \"NK_cell\"             \n",
       " [4] \"Treg\"                 \"Tumour 2\"             \"Plasma Cells\"        \n",
       " [7] \"CD4\"                  \"CD8\"                  \"B naive activated\"   \n",
       "[10] \"Mesenchymal\"          \"B naive resting\"      \"Monocytes\"           \n",
       "[13] \"Dendritic Cells\"      \"Neutrophils\"          \"Th1\"                 \n",
       "[16] \"Th17\"                 \"Endothelial\"          \"Tumour 3\"            \n",
       "[19] \"Cycling Plasma Cells\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "\n",
       "   B naive activated      B naive resting                  CD4 \n",
       "                1200                  268                  404 \n",
       "                 CD8 Cycling Plasma Cells      Dendritic Cells \n",
       "                 438                   25                  346 \n",
       "         Endothelial           Macrophage          Mesenchymal \n",
       "                 140                 7977                  364 \n",
       "           Monocytes          Neutrophils              NK_cell \n",
       "                 581                  256                  211 \n",
       "        Plasma Cells                  Th1                 Th17 \n",
       "                 432                  198                   78 \n",
       "                Treg             Tumour 1             Tumour 2 \n",
       "                 286                14301                 1225 \n",
       "            Tumour 3 \n",
       "                2644 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "\n",
       "   B naive activated      B naive resting                  CD4 \n",
       "                1200                  268                  404 \n",
       "                 CD8 Cycling Plasma Cells      Dendritic Cells \n",
       "                 438                   25                  346 \n",
       "         Endothelial           Macrophage          Mesenchymal \n",
       "                 140                 7977                  364 \n",
       "           Monocytes          Neutrophils              NK_cell \n",
       "                 581                  256                  211 \n",
       "        Plasma Cells                  Th1                 Th17 \n",
       "                 432                  198                   78 \n",
       "                Treg             Tumour 1             Tumour 2 \n",
       "                 286                14301                 1225 \n",
       "            Tumour 3 \n",
       "                2644 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "\n",
    "#### Subset single cell rna-seq to cell types of interest\n",
    "unique(seuCombined$updated_annotation)\n",
    "table(seuCombined$updated_annotation)\n",
    "\n",
    "#currently not removing any cell types\n",
    "seuSub <- seuCombined %>% subset(updated_annotation %in% c('Tumour 1','Macrophage','NK_cell','Treg','Tumour 2',\n",
    "                                                            'Plasma Cells','CD4','CD8','B naive activated','B naive resting','Mesenchymal', 'Monocytes',\n",
    "                                                            'Dendritic Cells','Neutrophils','Th1','Th17','Endothelial','Tumour 3','Cycling Plasma Cells'))\n",
    "                                                            \n",
    "table(seuSub$updated_annotation)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "50a47191",
   "metadata": {},
   "source": [
    "## 5. Co-embed Visium with scRNA and run CellTrek\n",
    "Uses `CellTrek::traint()` trains a mapping model between single cell profiles and ST data \n",
    " - project Visium spots and scRNA-seq cells into a **shared latent space**.\n",
    " - first need to handle images as the split object function (from downgraded seurat versions used here) can't handle images"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8985c886",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "class(ovVis)\n",
    "names(ovVis)\n",
    "\n",
    "for (i in seq_along(ovVis)){\n",
    "    name <- names(ovVis)[i]\n",
    "    obj <- ovVis[[i]]\n",
    "    assign(name, obj, envir=.GlobalEnv)\n",
    "    \n",
    "}\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "24c34bab",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "#checking if colnames of spots are clean \n",
    "lapply(ovVis, function(x) any(duplicated(colnames(x))))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "29c7171a",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "\n",
    "# clean up names \n",
    "# make.names() ensures all cell names are valid R names (hanldes special characters)\n",
    "# RenameCells() applies cleaned names to the Seurat object \n",
    "# Clean Visium spot names\n",
    "\n",
    "#no need to apply to visium, already clean and characters are appropriate \n",
    "#ovVis <- lapply(ovVis, function(x){\n",
    "#    RenameCells(x, new.names = make.names(Cells(x)))\n",
    "#})\n",
    "\n",
    "# Clean single-cell names\n",
    "seuSub <- RenameCells(seuSub, new.names = make.names(Cells(seuSub)))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d3fd7f21",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "seuSub@active.assay\n",
    "seuSub@assays$RNA"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a98f5332",
   "metadata": {},
   "source": [
    "CellTrek uses a random forest regressor to learn the mapping between your single cell reference data and spatial transcriptomics data \n",
    " - traint() trains a mapping model between your single-cell reference and spatial sample\n",
    " - The object returned (ovTraintAll[[i]]) contains:\n",
    " - Model parameters\n",
    " - Fitted distances between spots and reference cells\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "180408f8",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "class(OV_1@images[[1]])\n",
    "slotNames(OV_1@assays)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b3c33856",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "#try on one sample first\n",
    "\n",
    "ovtrain_ov1 <- CellTrek::traint(\n",
    "    st_data    = OV_1,               # spatial Seurat object\n",
    "        sc_data    = seuSub,                # single-cell reference\n",
    "        sc_assay   = \"RNA\",         # exact assay name in seuSub\n",
    "        cell_names = \"updated_annotation\"  # column with cell type labels\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c6cb7f52",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "\n",
    "ovTraintAll <- lapply(ovVis, function(i) {\n",
    "    ovTraint <- CellTrek::traint(st_data = i,\n",
    "                                 sc_data = seuSub,\n",
    "                                 sc_assay = \"RNA\",\n",
    "                                 cell_names = \"updated_annotation\")\n",
    "    return(ovTraint)\n",
    "})\n",
    "\n",
    "# Add names so they match the input list\n",
    "names(ovTraintAll) <- names(ovVis)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "340dcaf7",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "#saveRDS(ovTraintAll, paste0(ov_visium, \"cellTrek/ovTrainAll_obj.rds\"))\n",
    "\n",
    "ovTraintAll <- readRDS(paste0(ov_visium, \"cellTrek/ovTrainAll_obj.rds\"))\n",
    "#cluster\n",
    "#saveRDS(ovTraintAll, paste0(cell2loc, \"/cellTrek/ovTrainAll_obj.rds\"))\n",
    "#ovTraintAll <- readRDS(paste0(cell2loc, \"cellTrek/ovTrainAll_obj.rds\"))\n",
    "#seuSub <- readRDS(paste0(cell2loc, \"cellTrek/seuSub.rds\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "356af6a5",
   "metadata": {},
   "source": [
    "## 6. Visualize embeddings\n",
    "Plots the joint embedding to show how Visium and scRNA cells mix, grouped by type or annotation."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "b7057492",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "Attaching package: ‘pals’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:viridis’:\n",
      "\n",
      "    cividis, inferno, magma, plasma, turbo, viridis\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:viridisLite’:\n",
      "\n",
      "    cividis, inferno, magma, plasma, turbo, viridis\n",
      "\n",
      "\n",
      "Only 26 colors are available with 'alphabet'\n",
      "\n",
      "Only 26 colors are available with 'alphabet2'\n",
      "\n",
      "Only 25 colors are available with 'cols25'.\n",
      "\n",
      "Only 32 colors are available with 'glasbey'.\n",
      "\n",
      "Only 22 colors are available with 'kelly'.\n",
      "\n",
      "Only 36 colors are available with 'polychrome'.\n",
      "\n",
      "Only 24 colors are available with 'stepped'\n",
      "\n",
      "Only 12 colors are available with 'tol'\n",
      "\n",
      "Only 16 colors are available with 'watlington'.\n",
      "\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAA0gAAANICAIAAAByhViMAAAACXBIWXMAABJ0AAASdAHeZh94AAAgAElEQVR4nOzcd/zO9f748dfHh5TysbNnSIiQcUp2wyojtLTTonnaaQ91mqJz0jyVBkpTW5ERGdkjMyFJsrJdvz/o+1Op9DkXV72u+/2vy3W9X8/X632dW51H7w8yEolEAADg7y9Hqg8AAEByCDsAgEgIOwCASAg7AIBICDsAgEgIOwCASAg7AIBICDsAgEgIOwCASAg7AIBICDsAgEgIOwCASAg7AIBICDsAgEgIOwCASAg7AIBICDsAgEgIOwCASAg7AIBICDsAgEgIOwCASAg7AIBICDsAgEgIOwCASAg7AIBICDsAgEgIOwCASAg7AIBICDsAgEgIOwCASAg7AIBICDsAgEgIOwCASAg7AIBICDsAgEgIOwCASAg7AIBICDsAgEgIOwCASAg7AIBICDsAgEgIOwCASAg7AIBICDsAgEgIOwCASAg7AIBICDsAgEgIOwCASAg7AIBICDsAgEgIOwCASAg7AIBICDsAgEgIOwCASAg7AIBICDsAgEgIOwCASAg7AIBICDsAgEgIOwCASAg7AIBICDsAgEgIOwCASAg7AIBICDsAgEgIOwCASAg7AIBICDsAgEgIOwCASAg7AIBICDsAgEgIOwCASAg7AIBICDsAgEgIOwCASAg7AIBICDsAgEgIOwCASAg7AIBICDsAgEgIOwCASAg7AIBICDsAgEgIOwCASAg7AIBICDsAgEgIOwCASAg7AIBICDsAgEgIOwCASAg7AIBICDsAgEgIOwCASAg7AIBICDsAgEgIOwCASAg7AIBICDsAgEgIOwCASAg7AIBICDsAgEgIOwCASAg7AIBICDsAgEgIOwCASAg7AIBICDsAgEgIOwCASAg7AIBICDsAgEgIOwCASAg7AIBICDsAgEgIOwCASAg7AIBICDsAgEgIOwCASAg7AIBICDsAgEgIOwCASAg7AIBICDsAgEgIOwCASAg7AIBICDsAgEgIOwCASAg7AIBI5Ez1AdLRqv8mbVT+aXOTMuexsRWTMieE8Oaw05IyZ3zxj5IyJ4SwtHmxZI2qdkXRpMyZuq19UuaEEG59a2myRhW+uW5S5mzu2Topc0IIWe2SM6fS4b2SMyiERnU2JGtU4vI7kzMo55DkzAmh+dCjkzLno375kzInhHDGqq+SNerZl55MypwF3WYkZU4Ioc/+jyVrVKUfM5Iyp1vilqTMCSGEr5I06tJcyZkTQhi8KWmj/qo8sQMAiISwAwCIhLADAIiEsAMAiISwAwCIhLADAIiEsAMAiISwAwCIhLADAIiEsAMAiISwAwCIhLADAIiEsAMAiISwAwCIhLADAIiEsAMAiISwAwCIhLADAIiEsAMAiISwAwCIhLADAIiEsAMAiISwAwCIhLADAIhEuoVdYsrLtw+e9ZsfTxt46yszdmfOso8eeWzkyt3acsPiKZMXr9+94wEAZF+6hd22yS/d9urM3/x46oBbBk3fnTnffNj7PyO+360tFz5z+qlPzNu94wEAZF/OVB9gj9r89ejBQ0bPX71v+Ybt2zUovs9OHy376JH387U5aMabI5fnrtCkXfvaRX9q3MQPU9585eNFeeu2ad2gzP4ZvzNn3dyPnhsyeV3RI0/oUK94zp/e+/j1dyetK/WP41rUK71/RmLawD5Dv/l226OPNLyxR/Pie+3OAYA0FPMTu/n/aXnoSf+euGL9t5/2anX4JR9v2vnDbz7s/c92Lc4d8NW6FSN7HVvnlEHLdnzw+R0n9/xkxbq5/TpVPqbPgt+Zs/jprqc8PGnFd6P+1fKwDs8vCiGE9SN7HtH4yiFLVi8YeFHdRrdM3BhCvtKVDtw394EVKxXdd+/dOQCQliJ+Yrd12dYKNzxz3z+bZoXQ4sdiF41eEJpW3vmC70p0G/HGNZUywtX1Ty97bd/JJ95WI4Sw9OCzp93fKXc4Lffo0p9NDD3K/dacFflOHfXG1RVDuLr+6WWv6DvutF4ln7zhviI3zn/+gmIhdCvaqFSPf58z4rIGx1bP//iWY46rXiAVXwIAkEYiDrvMBhfdm/Xu4IdunzZt3LuvLEtcmfjFBWVbHF0pI4QQ8hzbslGHlydvDDVCCA1btMgdQgglSpbc9vm235lTtnmLiuGn5fP7T/0xfDdufP6Mmo/demsIISwN+adPnxlCmb1ztwAAEYfd6ncuqNNtUuPzzmlz0i3N8s686lcXZGRk7HiVI0eOsHnz1hBCCAUL/uLR2m/N+dnyHDlz5kxs3ry52MF1q1ffL4QQqld/7Lgyhyb/tgAAfkPEv8duzKCX8lzw36duOq9D80o/fv3Nry9Y8NGHc0MIIaz/8P1Py9eokefPzVkw9KN5Py0fUaXWYftkVK9x6Lebi7bs2LFjx47Nco99eux3ByT9rgAAfkvET+yq1K698KGre+Zpvu/0dz+cfsDiVa+OPf+6OjtdkPX1Y506rOhYY/lbj75zRO9xdf7MnNwh5J31YNuOK0459Nu3/vPR0f3GVQ8hnH97uzqnNz7j3OOLL3jjmRE1nrkmbwjhgLx5F73y5AutrjulbpE9f9cAQPqK+Ild6Ytf/+z+lvm35qp4zhPvD3/jkePyr90cctQ46aYOVbZfUK77oJe6HXxAVt3L354w8KQSIYRQvfMtJ1bdsf6n17ucU6zFJX0/HPv8aQftn6/B1e+O++/xhUMIIf9xfScMvb1pkX2LNr556BdPHFcwhBBKnvHok52Krdvwy9/iBwCQXBE/sQshq+rxF/yUaSVOvbxOCCEU79Lz/37jW0beysedVfm4nVZU63RztV+/3uWcHqeGEGqfW+sXe+Y75LgzDznu529V79i9+v9+NwAAvy/qsPs9uYtUKJc3V6pPAQCQPGkbdlWueG9wqs8AAJBMEf8eOwCA9CLsAAAiIewAACIh7AAAIiHsAAAiIewAACIh7AAAIiHsAAAiIewAACIh7AAAIiHsAAAiIewAACIh7AAAIiHsAAAiIewAACIh7AAAIiHsAAAiIewAACIh7AAAIpGRSCRSfQYAAJLAEzsAgEgIOwCASAg7AIBICDsAgEgIOwCASORM9QHS0cCspP1J5F5XJ2dOx0X5kjMohDUjXk7KnHnVn0nKnBDCqEOTNSlsunFOUuYs/OqSpMwJISzPaJasUUfM6ZGUOS/dPT4pc0IIJU5dmJQ55c7cmpQ5IYTDT/lvskaN63RDUuZsWn1PUuaEEP5zQM+kzOl2VcGkzAkhVJ77TLJGFZxdKylzbj/82aTMCSHkuyBXskY1fLF4Uua89XVy/h0eQmj99dCkzEmUOispc0IIuRLXJmvUX5YndgAAkRB2AACREHYAAJEQdgAAkRB2AACREHYAAJEQdgAAkRB2AACREHYAAJEQdgAAkRB2AACREHYAAJEQdgAAkRB2AACREHYAAJEQdgAAkRB2AACREHYAAJEQdgAAkRB2AACREHYAAJEQdgAAkRB2AACREHYAAJEQdgAAkYg47LZ92LPFXSN/8+OPb2l6+/DdmTP38dPOe3HJbm35/ainnxz5XQghhNXjnr7h/M5tjz+5x4MfLtq8W6sBAP4nEYddYtnUYTO++82Pv532yfTluzNn7bwx475av1tbLv+490MfLgshzHyg5dH3LKxx5iVn1F3zTIcWPcds2q31AAD/g5ypPkCSbJj3+oP3vTx6/up9yx913nVXHV16p2Kd+/hpj+Y+ufiwJz9enrtC8/NvvKRJ0YwQQgiJ1ROeuPLBIYvy1u14/kVdaub77TkbZ7/S864XJq8retTZN1x+TMnMEEJYP/3lXg8NmrSu1D/anXVBpxr5tn18y+lPzpmfOOu0Erce8tz0jg8Nu/i4nKFVk/2nl7xg4IRe9Rvs9S8FAEgvkTyxG3tTi9PfOeCEHtdc1Gztk20vfGH1zh+unTfmkfMvGXHI2Vf1OHrbM+2b3z552/YPhtx6y+ImPa7qmue109v1Gv87c2bdf+UL+dv36H5MRv9OTW4cuyWEsOTpE5vcsbhetyvOrDb3tjan9/86kePIKx8+qVzpzg88cmrt42567rIjdkTzxo2bSpcuude+CgAgbcXxxG5r4Za3v3HhSY3LZ4Zm4Y2beyxcGkLWzhfkOPbWZ/7ZJn8IjZ/4skrrfsNv6tMkhLBvy0tvalsvI1Tr1uz+DxeEUOe35mxrev2TV/60/JgnP767Xul+9w8/+vZl5x6dJ4Q61w0pcufjU0+99dAC++fMuaVAgTxF67RvE0IIYc3EPl17DDv6vl6l9/p3AgCknTie2GVWOKzS7P+c2655vUMqn/HK6l9fULZmzfwhhBAyah1ee8WcOWtDCCEc2bBhRggh7H/AAb8/52fLV82fvzoxa9aXmZ/f17pp06ZNm7Z5+IvMxYsX/2LLzYuG3NC86nFPl3x4xIudPbADAPa8OJ7YLe7TvsXTh/R58D/316+45PZDTv7VBatX/1RpP65ZEw44YN8QQgj77pt79+b8bHlmoUIHZOTJs1/Ni54d1DX/T5fkzrvzoE3T+7Rs+q+81wyYftk/CsURzwDAX14c0TFtwoRS7S45vVGlgpnz3vtobiKR+MUF3wz+7/trQghb5j3+9LAGRzX4jZ79rTnfvPbsBz8tH9G46T9yhNqNjpr55tvfZxUqVKjA8n6tK1/49sbtl27bti2Eb5/scV3GzSMHX6HqAIC9J44ndg26nLH11JZHjau933c/lq9Q+6vnbu5/8ssn7XRBqTILr6/b4F/Ffpg4q0TPt7uV+zNzqodQvujUC2v+o3yZlV/MO+jOd84sEUI47b5732zTsOrA6oUWTVpWu9f7HfOGEA4sXXrRzd0vPbjNkuHbZn/f6Yjndswt0uWxNy6rsSe/AQCASMIu69hHJk+/eOpXiZJVDy6279prJ83PWTxkHn3HR3WKbb+gcOu+o7tlTp+/sdghBxfdLyOEEJrd+kmNA3es/+n1LucU7Nb/9cvqVM6cP2NRolTVSoW3//g28+AzBk1vt2DG7JV5KlSrWGifEEIIBU4fMKv25OV5Cyc+rPfDTufLXarc3vkiAIB0FkfYhRByH1ilzo5Oyypfq2YIIeSp1ujA//s8I1eB8jUL7LSgSNXGRXbxehdzytc7NIQQKh5W+Bd75spXrkbdcj97a9/i1esVDyGUrfA/3g8AwJ8VTdj9jkL1Tz5xv6w/vg4A4G8tHcKuVLvbbkj1GQAA9jh/aBMAIBLCDgAgEsIOACASwg4AIBLCDgAgEsIOACASwg4AIBLCDgAgEsIOACASwg4AIBLCDgAgEsIOACASwg4AIBLCDgAgEsIOACASwg4AIBLCDgAgEsIOACASwg4AIBIZiUQi1WcAACAJPLEDAIiEsAMAiISwAwCIhLADAIiEsAMAiETOVB8gHZW7bkwSpy3o1SBZow5/pnqyRoUQ6j/yaLJG9W3TKFmjQghdZnRO4rSXbx2QrFEvrl+UrFEhhE6PvpmsUe/csDZZo0IIxaa0SeK0OnkfS9aoJ2ZXTdaoEEKntt8la9TsVrmTNSqEUL/jJUmcNubhpI1q8F2upM0Kod659yVr1NibTknWqBBCKF0yicOuSdpdhmPrJG1UCGF8z6SNKrGyYdJmhXDKtE+TOO2vyRM7AIBICDsAgEgIOwCASAg7AIBICDsAgEgIOwCASAg7AIBICDsAgEgIOwCASAg7AIBICDsAgEgIOwCASAg7AIBICDsAgEgIOwCASAg7AIBICDsAgEgIOwCASAg7AIBICDsAgEgIOwCASAg7AIBICDsAgEgIOwCASAg7AIBICDsAgEjkTPUB/hYSU16+Y85hPdsfvOuPV8784N2hk5dmlmvcoV2dIpkhhPD1+w88OXrNjs/zN+x2afPie+msAEDa8sRud2yb/NJtr87c9Wffvdv9H0dd8eqC7+e/fdPR1do9tziEELaOfabnwElrd1i3cevePC0AkKY8sUss/+KN1z+Zn1GuUdvjax+YI4SQWDHl7TeGzl5fpHbL9k3K7/fzy7eumPD6oI/nrMtf7bjOrarmzVj2wl1PHXTvwoFnFQnhn7Vbl7jn8aldb6n+1dy5JY//97/uqp2SWwIA0lO6P7FbMuCUw9s/PGnF0mF3t6nd5YWlISwffFbtY27/9Ls18wdcWP/IW8Zt3vnyNe+cf1ib+8auWLv4lcsatLhvVggbih9xxfnHFAkhhFCgUqVC27ZtCyHMmzev5D5L+j9030NPvzXth1TcGACQftL8id34B68Z3enFOfc1yBkur9O1wevDVtWaceOQFk8teLJlnhDOLd3okDsHXD741AN+un7OJx+HU1948q76Yd0JtR6as3FLOLhTrzu2f7Zm/L9ufCnrpHdrhLBm3rzvRr3f59CuR+732Y2Nbh7w1MRnTyiUqnsEANJFeofdygkTljW6skHOEEIo2Pm52Z3Dppfbz/pH52Z5QgghlGzWrEzvKV+GUOunBTU7npq/Rdva807q0Kplx4s7HbLj69uy+L27L7yw96ITnhva87AQQs6jb/v4i4qNDymUEcLZRY8ofetjN55wfeW9f38AQFpJ7x/FbtmyJUeOn38FicROv8jMzNyyZctOb+Sod9u4WUNual5w0YBLG1TtNmR9CGHViJuOqnPR+COemDjmweOKZYQQwn7l6jc5pFBGCCGEkk2bVvrqq6/28J0AAKR52BWpUSNr1KcTt4UQwqqX2uU98pHvDq1R+bOhn6wPIYSwdOjQ+YceutPfcbLpk9tO6DXz0HYX3fLYe6//s8Crr40OG96+rOOAWs+PHXxts1L77Lhsy/sXH1T/tgnbi3DZiBHzq1U7ZK/eFwCQltL7R7HhyCvvqnLkia02nHH4D+8+PfbY+/qVqBbuOPbws1uc3611wSkvPLny0jdPLhDCT39byT4VCq9pf0qjWV1blV/x3tOr2vy7Vvj4soGbCnb+6L7rP9p+SdaRF1zXpsmZ5+Y84cRjl57WvPDC15754fIXzymZslsEANJGmoddKHfm4HHVBr/x6cKtlw2+u2O9A0MIHf47seJbrw2dvaF97886NKuwTwghR42TbspTJYRQ5qK3x1ce+NaEpVsqXfPOv9rWLLBxYourLquw08B9c+UIYZ+6130yttmb742av7HKvZ8+1qTcvqm5OwAgraR72IWQWbzuiefX3fmdjEI12p5T42fvHNql56HbX+5XocXpl7T4v09y1+p6c62wC7lK1e9wTv2knxYA4Del9++xAwCIiLADAIiEsAMAiISwAwCIhLADAIiEsAMAiISwAwCIhLADAIiEsAMAiISwAwCIhLADAIiEsAMAiISwAwCIhLADAIiEsAMAiISwAwCIhLADAIiEsAMAiISwAwCIREYikUj1GQAASAJP7AAAIiHsAAAiIewAACIh7AAAIiHsAAAiIewAACKRM9UHSEcZGdlfmwjZX5zxv/zNNhl1sr00EW7P/rYdW2V/31uyvTRkjMv+l9Xv1ebZXtvtjRuzvfbaJpnZXvtilxnZXrvgwPOyvfaNcv/M9tomDZdne23eu5dke+38yz/K9trmJbL/39IdlpyY7bU5uzfJ9tr7mj6Y7bWjOszJ9tpPemR7abim9dPZXnvXhM+yvbbUTY9le+2VP2b/3znfHZ/9+51TdF22175zdfdsr+1xbbaXhtUvFs/22rwHLM3+xn8TntgBAERC2AEARELYAQBEQtgBAERC2AEARELYAQBEQtgBAERC2AEARELYAQBEQtgBAERC2AEARELYAQBEQtgBAERC2AEARELYAQBEQtgBAERC2AEARELYAQBEQtgBAERC2AEARELYAQBEQtgBAERC2AEARELYAQBEQtgBAERC2G3t3z5X19d3//pJ11Wqdc/cPXceAIBsSruwS2z4dsakGd9uSHw3Y9SXP/zy020/Lps9Ycy46Ut/3OnN9d/OGj92wpwVm3aasnXNoikTZy5bn9jpuq1rl0z74svvN4cQwuq5oz9f8NOQ1XNGj90DtwIA8DM5U32AvWrj1L7Ht77j2/IV13+zX+nc4ys+sOLfzf//p6s/vrHZyY+tO6ha3sWTFlS45dOPLj04Y8v0vm2PuXVOmSpZC6auavvfzx5rG0JYNey6Rk/Nz1X4h0lzSl715pDr6+0fEosGn9vywk/yVS22fPYPTfsMu2fVbc1eP2vZq53zhLD0ubOOG/Xpqv4pu20AID2k1RO7Zc9cfkvu2yZO+OTTSS83+Gbytp9/unXE4FdK3jx2xshPxs56sdXkZ99YEEKY/Mz9UzoNnjVq+Phxjx67YMLkLSGEMH9a/ps+H/fJyBljr9x652WPzQ9h7eBrzv2s7VszRg4dOfX1Vp+cefEHR7Q5fOiQoZtCCN++8dqUVh32+s0CAGknncJu26ihw+t37lwsI4TcNTsdX/kXH2e26j3jlVNyfPHpkP6PvjBu1aZNm0IIVerVWzuw55WPvDrix+N6v3dzs5whhFCsw5nHZoUQclU6/6xGY0eN2RLGf/BBnsMrrRs9bNiw0WvK1thn2LDFrdsc8s6QEYnww5uvjT2uY8u9frcAQNpJp7Bbu3JlIm/e/bb/Iisr6xcfb57+eIdKBzW/4j/vzgyVDi62/c08Jz4x+pFWiU97n1OvZNmWfaYkQgghX758O9bkLVgwY926DYkVK1ZmLvl80KBBgwYNemVWsS6nNihcrk2bA98eMn7126+Nbt7huDx75xYBgHSWTr/HLqtcubwzZ84NJxwUwuaZM+eFfX/28ZhHe046+f05d9XOCNuGX3TvgBBC2DR1UL/5Na58aMA/H/px9JU1Oz4+uvs5ISycNGlVqJ4vhDBlwsQilc4/ICNv+bJ5qvd45KaqIYStMwc/MLZUuXBwmzb7nDzgvoUjmnR44YAU3C4AkG7SKexC4wsv2tz64jur3dhw5cv3vr8l44SMnT/NyspaPm7I0Ik595321I0Dv/+h47yVmw/e9Hnfs3t/dee17cuvGDh8WflTSoQQwoa3bjy3d+HLanw78KpHc/cYflQIGed1L9Cg60Wl7z25+IL+1131xRljrgqhZps2a5rf++4x/R7Pm5r7BQDSSzr9KDZk1rrl/aebLHj6/senH3XpWWXz588fQsgoWr3xIYVDCDWuG/Bw9akP9bzz+cXH/PeDfi2XvjBgbqh905t9Gq1+t0+vR97Nddpbgy4tFw6oUP+Kfn2Pmvvcvx58N3HmGx9cVy0jhFDx8nfevThrWN9e//74gAuGvHlppRBCaNC69f5bG3Zsmz+ldw0ApIu0emK38qUzj36pet97Hi6/4fO7T9l20vM1Qgg5Wtz+YYsQQgh5DzvrgZfO+unifq+fEUIIoWqXO57qstOQg857/v4QQmh1yc+HZx1+dq9nzv7ZW+u//35Lsw5tC+yBWwEA+JW0CrsCXe5/ZsvTL9zaY37O8k0f/bB7rT1496tnfTLk0Xvfb3hx34J7bhMAgJ2kVdiFjAPrnnZN3dP2xlarv3j11bXHP9u3a5G9sRsAQLqF3V5UqkvvAV3++DIAgKRJqz88AQAQM2EHABAJYQcAEAlhBwAQCWEHABAJYQcAEAlhBwAQCWEHABAJYQcAEAlhBwAQCWEHABAJYQcAEAlhBwAQCWEHABAJYQcAEAlhBwAQCWEHABAJYQcAEImMRCKR6jMAAJAEntgBAERC2AEARELYAQBEQtgBAERC2AEARELYAQBEImeqD5CO1qxdtyfGHlazxp4YO+m/8/bE2D6P9N4TY7vXuGRPjH31mFV7YmwIoXS9pntibM43jt8TY4dP3iP/KXhjzql7YuyS3J33xNgSMzfuibEbq++RL+H8D8rsibGPnnThnhj70VPd98TYQx7ssyfGhhAGnVJ0T4zNdc+yPTF25oMV9sTYnoX2xNTQcvEe6ZOpw2ftibF/KZ7YAQBEQtgBAERC2AEARELYAQBEQtgBAERC2AEARELYAQBEQtgBAERC2AEARELYAQBEQtgBAERC2AEARELYAQBEQtgBAERC2AEARELYAQBEQtgBAERC2AEARELYAQBEQtgBAERC2AEARELYAQBEQtgBAERC2AEARELY/b4lz5932lNz//CybR/2bHHXyJ1fAADsbcLu961fOG7M/LV/eFli2dRhM77b+QUAwN4m7HbT1rnPnt/m4pcWbQshrJ/+8s3dOrU79fJ7Bk5etevLFz1/dsfeU3csHXVHyx577aAAQNoSdrvl2yEXtbx55enXdi6dIyx5+sQmdyyu1+2KM6vNva3N6f2/TuxiQemD8o55/PVZIYSwbcSLjywtupcPDACkIWH3x9aOubV155FdBj/XuXSOEGb2u3/40bf1PvfoRkef++B1h39w1+NTd7WofpvWy4YM+SqEMPq114t26LCXzwwApKGcqT7AX9/Cvje+VbvUwpGjloXDyoTErFlfZo6/r3XT3iGE8OPCzMUFF4dQ9VercjRu03Ja73dWXH74a6/l6fjWry8AAEgyYfeHsjr3+6Bf1vWVT7ruzdP6t83Kk2e/mhc9O6hr/p8+z513l8v2ad6m+WnPvzfi69dydXy12l47LQCQvvwo9g8VKF8+f55mt95Z/+2r7hq7JdRudNTMN9/+PqtQoUIFlvdrXfnCtzfuel2eY9o0HH7/pQMTHTscundPDACkJ2G3m4qc/q/r9u975WMLCp12370FHmlY9ajmDSo07Fei150dd/3ELoR8rVrXmjhhQ8cOdfbqSQGAdOVHsb+vZNcn+m+rGEIIGVUue21Yg+9ybswsd8ag6e0WzJi9Mk+FahUL7RNCCJlH3/FRnWI7vwghhKK1ahUvfVjHuqk6PACQXoTd79u3TJ16P73OXbp249LbX+bKV65G3XI7XXdgtUYH/vxFCNvGvf5WRocndR0AsHf4UeweMrtvh0NbP1HkmouOzEj1UQCANOGJ3R5S+eJXp12c6kMAAGnFEzsAgEgIOwCASAg7AIBICDsAgEgIOzOd7lUAABh2SURBVACASAg7AIBICDsAgEgIOwCASAg7AIBICDsAgEgIOwCASAg7AIBICDsAgEgIOwCASAg7AIBICDsAgEgIOwCASAg7AIBICDsAgEhkJBKJVJ8BAIAk8MQOACASwg4AIBLCDgAgEsIOACASwg4AIBI5U32AdHRqhy7ZWHXvQ/dnY1XBpq3/7JJ9543LxkbFEg2ysarUHdnZa1z1Yn92yRvtO2Rjo+6JbtlY9c9Xj8/GqqeqL8rGqtdGb/uzS8b1yMjGRsf3+W82Vh3x3pZsrOq9/NY/u+SkRkdkY6MLF72TjVXVX+yTjVWtZtyZjVXlZx/yZ5fMa5adfxJzlMrOqs0jF2Zj1Zg++/zZJe++nT8bG9Ue2yobqwrNys7fFDHyrNLZWFVn2GN/dsmYb0plY6OnXhqejVUfLjg3G6semnPbn12y7aq7srHRtHbZ+dfLiET82eOJHQBAJIQdAEAkhB0AQCSEHQBAJIQdAEAkhB0AQCSEHQBAJIQdAEAkhB0AQCSEHQBAJIQdAEAkhB0AQCSEHQBAJIQdAEAkhB0AQCSEHQBAJIQdAEAkhB0AQCSEHQBAJIQdAEAkhB0AQCSEHQBAJIQdAEAkhB0AQCSEHQBAJKIMu6392+fq+vruXv1yp4yTX9mTxwEA2CuiDLsk2rRxU6qPAACwe/5GYTfpukr/uOCGNgcVyVegxKHt7xz+/fa3V372r5MOL1ekYPHKTbq/MHvjziu+eqBeoXM/3BpCCGHtS+3yHvv09yF8P/LuE2uVKVSozGHt7xi2PBFCCOHHaY+f9o8qpUpVPfbCZ2duCSFs7d++yIWPP31K1RJnvbZ1l1u83KlgmwvPOqxU4YJFDjnx4fdeverYQ8sUr9Do3JfmhRBC2DD72QsbVy5epGytlhc/N2PLXvqKAIC09jcKuxDCZ/2HN3j5yxVLRvTM83iHK95aF8KPb1/R7rH9rhv29fJZA06Y0aPzAzMT///yMm3alHxnyJgQQlj/3utDG3RoW3DNKxd3eKnM/eOXLvz4n3kebX/1+1tCCOGN/hPPeXvGgkm3FhhwyUOfbF+86vn/fH7y24ue77zxN7ZY+d7Eg16as2zOE/U/vazNg0UeGT9vyp0lXrvj2akhbB1/2wlXL+06ZN6ymQM7Lbqmw01jNgYAgD3s7xV2+Tpff+3h+XPuV6Hz3ZdUHjTo4xA+HfRKue63dyybOzPrsEtv6PLNoMFzdrq+cps2+w8Z8kUIm95/7d26HU4ovOX9l1457LJbmhXZ54CDTvvXwD6dymwKIYTKJ57XtGBGzkIdj2+8atWq7Ws31+x6fdvy++X47S0an3lBlX0zC7Y8pn5G4zPOq7xPzsKtj6m3cOHCEMb3f37NKVefXXG/HPtVPKNH2xUvDJy4l78pACAN5Uz1Af6UEmXL7jhwmQoVNr6zeNW6fb9ZV6Ztye3v5ShbtvSSr5eEUOH/FtRq02ZN9yFzb1n22pDqHe4tFpZ9/fUBZctm7Zh25CklQgiTQzjssMO2T8jM/P+blSxZMoQQ1n2zqy0qhRAKFCgQQgiZmZmZRYr89DKEEBKLFi1e/dl5NT/Ykc1Fi7bzO/UAgD3u7xV2S776akuomjOEsHjhwlzFOmXtv2+xPIsWLQmhZAgh8fXXi4sXL77zgowGbVovuunNFxa9fUjHO4uHsKlo0TUfLPkxhDwhhEXv3P/m/mdcFELIkSPj15vteHP/Yr+/xS5kFC1apPpVH4y6pEQIIayeN3bePjX+95sHAPh9f68fxa4acPd9k9Zs2/TVa9f3ntaufZOM0LBDh3l9b35j8ebE2ql97nyxUPvjK/5sRY6j2rScfs81r1Ts0L5ECGGfozu2+uyhO0au3PrjwoE9u909bkO+P9z0j7bYhbodOyzpff2A+eu3rpn7/HkN2z019+8V0ADA39LfK+yqdGg894JaxUvUv35l10EPt88Xwv5tH3ztnNW3HlGycOX2Ays+NPCq6r+4o1zNWzdf+U2ZDu3LhhBCKHTqYy+3nHL+oQeWqn/z8nMG9Dom1x9u+odb/Frupr0GX7btgWblCpdv/FCOawff1VjYAQB73N8rOPY95IzHn7nt8Z+9V/Af1wwYf83P3so8dfDmU/9vTeHCeWvVal9+xy9zFGt++5tTb9/p6i4DE11++Trz1MGJ/5uwqy12WpXZZfDmHS/3P+PNNWeEEELIqtX92c+6//k7BADItr9X2P1Z2zav/3bgY4NKnzjyoFQfBQBgT/sbhV3Zkx/qlavon1oy+74mjZ7Y/9x+F1beQ2cCAPjr+BuFXf4arY/9k0uqXDfm2+v2yGEAAP5y/l5/eAIAgN8k7AAAIiHsAAAiIewAACIh7AAAIiHsAAAiIewAACIh7AAAIiHsAAAiIewAACIh7AAAIiHsAAAiIewAACIh7AAAIiHsAAAiIewAACIh7AAAIiHsAAAiIewAACKRkUgkUn0GAACSwBM7AIBICDsAgEgIOwCASAg7AIBICDsAgEjkTPUB0lG//Acnd2CJmtWTOK14w3pJnBZCKNGofnIHvvlSuyROK1GhZhKnhRCKV2mY3IElqjRK4rRLRndJ4rQQQsUSSf4CqxVP5hdYvUQyv70QQpMnFiZ3YM3SJZI4rWHZ4kmcFkJoVC6Zxwsh3FnrzSROq1Izycer1TDJX2CtRsk8Yfcb2iZxWgihTp06f+WBf/Hj/TV5YgcAEAlhBwAQCWEHABAJYQcAEAlhBwAQCWEHABAJYQcAEAlhBwAQCWEHABAJYQcAEAlhBwAQCWEHABAJYQcAEAlhBwAQCWEHABAJYQcAEAlhBwAQCWEHABAJYQcAEAlhBwAQCWEHABAJYQcAEAlhBwAQCWEHABAJYQcAEIn0DbvZverXuXP6Xtho5OWl98IuAABpFHbrv501fuyEOSs2hRDC8ulj5ny/cs7YsfPXhhDC1rVLpn3x5febf7p23fyxk7/ZsGrB5M8nzFq+MfE7b4ZdLQ8hhG3rlkybOG3pj9v2+I0BAIQQQsiZ6gPsHVum9217zK1zylTJWjB1Vdv/fvbvkqPfm/rDqsT770xpVjfnB+e2vPCTfFWLLZ/9Q9M+wx5rVzjM6Xdql8/KbJm7qlTpHybPr9DznTcur7nvLt8MiUWDf7U8rBh63bGdn9hQ9ZDE0lCh2Po3U33/AEA6SJMndpOfuX9Kp8GzRg0fP+7RYxdMmLyt9jk925Yq0eqGm48vOPiacz9r+9aMkUNHTn291SdnXjzwhxBCCLPmlntkyrhPRs4Yc9mPN13+1Ne/8ebaXSxPTLin2+PlH5k6ZfjwKR+fvnHiilTeOgCQNtIk7KrUq7d2YM8rH3l1xI/H9X7v5mY7Pagc/8EHeQ6vtG70sGHDRq8pW2OfYcO+CCGEUKzj2S3zhRByVTr/rIYjhn+2bddv7mr54uHDV5zQ7aSiGSHkKHXG2celyWNRACDF0qQ58pz4xOjM/zz2Yu9zbu66rv497wzpfuiOTxIrVqzMXPn5oEETQwghFOtyav3CIWwNIV++fDsuyVuwYOL779eGsIs3Ezl2sXzNmjVZ+fNnbL8wZ968efbSXQIA6S09wm7T1EH95te48qEB/3zox9FX1uz4+OjuvQtt/yijfPmyear3eOSmqiGErTMHPzC2VLkQ5oawcNKkVaF6vhDClAkT81U4KyuEXbyZUWgXy8sedNC3j09YFuoWDSFMHDZsdevU3DYAkF7SI+z22fR537N7f3Xnte3Lrxg4fFn5U0qEsCUzc8XsyXNXtD+ve4EGXS8qfe/JxRf0v+6qL84Yc1UIIYQt79x89sOFrqi5fOBVfbZd+GHTEKbu6s3MXSzP0+6CU/55dtdbD+zZLAy967Ufsu5N8f0DAGkhPcIu1L7pzT533/dyn15bClQ57a1BPcqFEE68+sRrn7n7rXpPXP7Ou/nu6tO318o8h1ww5M1ulbYvqX71fR2WvnjfRxuLnzzo48vqZP7WmxV3sTxPq3+P+HevXv3/9VCBuhc916dCiu4aAEgvaRJ2Yf+qXe54qsvP3qrYufegzttfHn52r2fO/tWSQ0+9rd2pu/Fm1i6W71Oh/U392u/4Rdv/4eAAALsrTf5ULABA/NLlid2fVO2G0R9mFNidNwEA/iqE3S7lPKBwod17EwDgr8KPYgEAIiHsAAAiIewAACIh7AAAIiHsAAAiIewAACIh7AAAIiHsAAAiIewAACIh7AAAIiHsAAAiIewAACIh7AAAIiHsAAAiIewAACIh7AAAIiHsAAAiIewAACKRkUgkUn0GAACSwBM7AIBICDsAgEgIOwCASAg7AIBICDsAgEgIOwCASORM9QHS0eEH3ZPqI/yeOte0SPUR/kCFOhVSfYTf8+G8gak+wu8ZP75fqo/wB3r9pf/nDS3q/KX/Abnn6L/0v15CCN06XZPqI/yeefNWpvoIv+cvfrwQQoUKf+l/QDp90C3VR9jjPLEDAIiEsAMAiISwAwCIhLADAIiEsAMAiISwAwCIhLADAIiEsAMAiISwAwCIhLADAIiEsAMAiISwAwCIhLADAIiEsAMAiISwAwCIhLADAIiEsAMAiISwAwCIhLADAIiEsAMAiISwAwCIhLADAIiEsAMAiISwAwCIhLADAIiEsAMAiISwS4KvX7z4gv4Lf+PDrf3b59qrpwEA0lXOVB8gBuvmjRq5ZW2qTwEApDtP7P5XW9+9om3vmV8+2q7VQ9PCys/+ddLh5YoULF65SfcXZm9M9dkAgLQi7P5Xmcc98OYlVSpd9NqQy8q/fUW7x/a7btjXy2cNOGFGj84PzEyk+nQAQBoRdkn06aBXynW/vWPZ3JlZh116Q5dvBg2ek+ojAQBpRNglz7pvvllXpkzJ7b/IUbZs6SVLlqT2RABAWhF2ybN/sWJ5Fi3a0XKJr79eXLx48dSeCABIK8IuOTZs2BBCww4d5vW9+Y3FmxNrp/a588VC7Y+vmOpzAQBpxF93kgTljjwuzyltmhT+8JMHX5t+QfcjSp61OV+1dg8NvKp6jhC2pvp0AEC6EHZJkLvJ3ZOW3L399TUDxl/zsw8zTx28OQVnAgDSjx/FAgBEQtgBAERC2AEARELYAQBEQtgBAERC2AEARELYAQBEQtgBAERC2AEARELYAQBEQtgBAERC2AEARELYAQBEQtgBAERC2AEARELYAQBEQtgBAERC2AEARELYAQBEQtgBAEQiI5FIpPoMAAAkgSd2AACREHYAAJEQdgAAkRB2AACREHYAAJHImeoDpKOMjIyU7NulS5eU7Nt53LiU7Nuzylsp2XdNmaNSsu+5C3OlZN8w58WUbFv2+gUp2Xdhpdkp2bfUt3emZN/7yk9Kyb7Th5+Tkn3DaanZ9tsfeqVk3wNXNkjJvmsqp+b/B/MesH9K9t2bPLEDAIiEsAMAiISwAwCIhLADAIiEsAMAiISwAwCIhLADAIiEsAMAiISwAwCIhLADAIiEsAMAiISwAwCIhLADAIiEsAMAiISwAwCIhLADAIiEsAMAiISwAwCIhLADAIiEsAMAiISwAwCIhLADAIiEsAMAiISwAwCIhLADAIiEsAMAiEQ6ht3sXvXr3Dk9hK392+fq+noIYd2TLfc76+3sDQEA+KuIIey2LJ08bOqyRAghrJ03ZtjkpVtCCOHHhZ+PmLUyhG0/Lps9Ycy46Ut/3H718ulj5ny/cs7YsfPX/jRgv1a9hlxVf/uHo+as2rRi7qRx077ZuNMWiQ3fzpg049sNie9mjPryh18N2bRy/qQJ05es3frT9b81BwBgz8mZ6gMkQeaXT3Q+P9/QGbdX2zDkn0d1+eKyMfPurbf5/Rsa9zrs87ufO/rkx9YdVC3v4kkLKtzy6UeXVlo0+r2pP6xKvP/OlMYVdwzY/P6NrT65YP3TrcPQm1v9Z9s/li8MxbdMnLip43Oj+rbMHzZO7Xt86zu+LV9x/Tf7lc49vuIDK/oW+L8hzWqufa5D25vmlK6etWDaptaPv/OfE0qEX875bnrflH5DAEBaiOGJXUaD5s0WjRixPIQJn31Rstz6UaOWhjBx1KjSzZssHPxKyZvHzhj5ydhZL7aa/OwbC0KO2uf0bFuqRKsbbj6+zK6GrRq+7Oh3P3/ngy8+7p7x1MsjQgjLnrn8lty3TZzwyaeTXm7wzeRtIew8JNdTl13742WfTfv048+nv9Lww/Nv/2jjr+fsxS8DAEhfMYRd2KdJ84bjRozavHD06BIXXN5o2qjRm5aOHLm6RYs6rXrPeOWUHF98OqT/oy+MW7Vp06Y/Hla3XYdSGSGEQw6tsWXjxhC2jRo6vH7nzsUyQshds9PxlX9+9baxw0fU73rWQTlDCHmPOuvEAz799Mtfz0ny/QIA7EoUYRfyNW9edcSI0Z99tqxB45OOPHT06NEjR01v1vyILdMf71DpoOZX/OfdmaHSwcV2a1bp0qV/9uu1K1cm8ubdb/svsrKyfn71htWrt2RlHfDTOfLl+/7773c9BwBgD4sj7EL55s0zRjwycELtBrUObNhwv5GPvPRZ3RZNc495tOekk9//cmj/3rdf2rho9h6cZZUrl3fmzLkhhBA2z5w57+ef5ilfvsiUydO2/2L9lClzK1So8L/cCABAtkUSduGwZs3mvjq4SIMGuUPNhg2nDx5crHnzAiErK2v5uCFDJ04e+fwVNw78/odF81ZuDiEzM3PFvMlzV2zYvdGNL7xo88MX3/nWiGHPXXHv+1syMjJC+P9D6nS7OM+j51/54rBR7/U57+r3G110aqk9eZ8AAL8plrDL0bDj2Y3bdG5aKoTMI044o9EJZ7Q6KIQa1w14uPrUh3re+fziY/77Qb+WS18YMDeEiidefeK65+5+a0nR6o0PKRxCyFHs0EZVCoUQwoHVmlQ7MGP7yCJVG1ctEkLIrHXL+083WfD0/Y9PP+rSs8rmz58/hJ2G1Ljxw9e6hiEP3v3EpAp3ffrSKUV2MWfvfx8AQBrKSCQSqT7DX9zKl7o2fKl633tOLr/h87tPubng81/cVet/+1tidjz02+u6dOmSkn07jxuXkn17VnkrJfuuKXNUSvY9d2GulOwb5ryYkm3LXr8gJfsurDQ7JfuW+vbOlOx7X/lJKdl3+vBzUrJvOC012377Q6+U7HvgygYp2XdN5dT8/2DeA/ZPyb57Uwx/j90eVqDL/c9sefqFW3vMz1m+6aMfdv8fqw4AYA8RKX8s48C6p11TN0X/FQcAsLti+T12AABpT9gBAERC2AEARELYAQBEQtgBAERC2AEARELYAQBEQtgBAERC2AEARELYAQBEQtgBAERC2AEARELYAQBEQtgBAERC2AEARELYAQBEQtgBAERC2AEARELYAQBEIiORSKT6DAAAJIEndgAAkRB2AACREHYAAJEQdgAAkRB2AACREHYAAJEQdgAAkRB2AACREHYAAJEQdgAAkRB2AACREHYAAJEQdgAAkRB2AACREHYAAJEQdgAAkRB2AACREHYAAJEQdgAAkRB2AACREHYAAJEQdgAAkRB2AACREHYAAJEQdgAAkRB2AACREHYAAJEQdgAAkRB2AACREHYAAJEQdgAAkRB2AACREHYAAJEQdgAAkRB2AACREHYAAJEQdgAAkRB2AACREHYAAJEQdgAAkRB2AACREHYAAJEQdgAAkRB2AACREHYAAJEQdgAAkRB2AACREHYAAJEQdgAAkRB2AACREHYAAJEQdgAAkRB2AACREHYAAJEQdgAAkRB2AACREHYAAJEQdgAAkRB2AACREHYAAJEQdgAAkRB2AACREHYAAJEQdgAAkRB2AACREHYAAJEQdgAAkRB2AACREHYAAJEQdgAAkRB2AACREHYAAJEQdgAAkRB2AACREHYAAJEQdgAAkRB2AACREHYAAJEQdgAAkRB2AACREHYAAJEQdgAAkRB2AACREHYAAJEQdgAAkRB2AACREHYAAJEQdgAAkRB2AACREHYAAJEQdgAAkRB2AACREHYAAJEQdgAAkRB2AACREHYAAJEQdgAAkRB2AACREHYAAJEQdgAAkRB2AACREHYAAJEQdgAAkRB2AACREHYAAJEQdgAAkRB2AACREHYAAJEQdgAAkRB2AACREHYAAJEQdgAAkRB2AACREHYAAJEQdgAAkRB2AACREHYAAJEQdgAAkRB2AACREHYAAJEQdgAAkRB2AACREHYAAJEQdgAAkRB2AACREHYAAJEQdgAAkRB2AACREHYAAJEQdgAAkRB2AACREHYAAJEQdgAAkRB2AACREHYAAJEQdgAAkRB2AACREHYAAJEQdgAAkRB2AACREHYAAJEQdgAAkRB2AAD/r906kAEAAAAY5G99j68omhA7AICJAIftybJXf3EHAAAAAElFTkSuQmCC",
      "text/plain": [
       "plot without title"
      ]
     },
     "metadata": {
      "image/png": {
       "height": 420,
       "width": 420
      }
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "# ---- build a 19-colour palette----\n",
    "library(pals)\n",
    "pal.bands(alphabet, alphabet2, cols25, glasbey, kelly, polychrome, \n",
    "  stepped, tol, watlington,\n",
    "  show.names=FALSE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ee8b781e",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "\n",
    "#### View co-embedding\n",
    "pals19 <- glasbey(19)\n",
    "\n",
    "cols19 <- c(\n",
    "  \"#E41A1C\", \"#377EB8\", \"#4DAF4A\", \"#984EA3\", \"#FF7F00\",\n",
    "  \"#FFFF33\", \"#A65628\", \"#F781BF\", \"#999999\", \"#66C2A5\",\n",
    "  \"#FC8D62\", \"#8DA0CB\", \"#E78AC3\", \"#A6D854\", \"#FFD92F\",\n",
    "  \"#E5C494\", \"#1B9E77\", \"#D95F02\", \"#B3B3B3\"\n",
    ")\n",
    "\n",
    "# Step 1: Generate plots\n",
    "c2lPlots <- lapply(ovTraintAll, function(obj) {\n",
    "  DimPlot(obj, group.by = \"updated_annotation\", cols = cols19)\n",
    "})\n",
    "\n",
    "# Step 2: Save plots\n",
    "for (i in seq_along(c2lPlots)) {\n",
    "  ggsave(\n",
    "    filename = paste0(ov_visium, \"cellTrek/DimPlot_OV\", i, \".pdf\"),\n",
    "    plot = c2lPlots[[i]],\n",
    "    width = 8,\n",
    "    height = 6,\n",
    "    dpi = 300\n",
    "  )\n",
    "}\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "30b99610",
   "metadata": {},
   "source": [
    "## 7. Map cells into spatial tissue\n",
    "Runs `CellTrek::celltrek()` to assign scRNA-seq cells into Visium coordinates, producing pseudo-single-cell maps. Parameters control interpolation, number of trees, PCs, distance threshold, etc."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9e60a7ab",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "#try one sample locally\n",
    "\n",
    "ovt1 <- [[1]]\n",
    "\n",
    "ovCellTrek_ov1 <- \n",
    "    CellTrek::celltrek(st_sc_int = ovt1,\n",
    "    int_assay = \"traint\",\n",
    "    sc_data = sc_data,\n",
    "    sc_assay = \"RNA\",\n",
    "    reduction = \"pca\",\n",
    "    intp = TRUE,\n",
    "    intp_pnt = 1000, #reducing can speed up but impact spatial resolution and smoothness\n",
    "    intp_lin = FALSE,\n",
    "    nPCs = 15, \n",
    "    dist_thresh = 0.55,\n",
    "    top_spot = 5,  \n",
    "    spot_n = 5,  #increasing will reduce sparsity but increase time to run\n",
    "    repel_r = 25,\n",
    "    repel_iter = 10,\n",
    "    keep_model = TRUE,\n",
    "    ntree=100  #default 1000 but take a long time to run, reducing can reduce mapping accuracy \n",
    "  )\n",
    "    \n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0f5656a4",
   "metadata": {},
   "source": [
    "Run using parallelisation with 4 CPUs on the cluster (interactive session) - moved into scripts for slurm submission\n",
    "\n",
    "sbatch celltrek.sh (which submits run_celltrek.r)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a5696376",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "# --- load libraries ---\n",
    "library(CellTrek)\n",
    "library(future)\n",
    "library(future.apply)\n",
    "library(progressr)\n",
    "library(Seurat)\n",
    "\n",
    "# --- parallelization ---\n",
    "plan(multisession, workers = 6)           # use 6 cores\n",
    "options(future.globals.maxSize = 20*1024^3)  # 20 GB per worker\n",
    "handlers(\"txtprogressbar\")  # to monitor progress\n",
    "\n",
    "# --- logging set up ---\n",
    "log_file <- \"celltrek_progress\"\n",
    "sink(log_file, append=TRUE, split=TRUE)  #redirects console output to log file, split=TRUE keeps output visible in console while also writing to log \n",
    "\n",
    "# --- load your data ---\n",
    "cell2loc = '/mnt/scratchc/fmlab/lythgo02/OV_visium/emily/cell2location/'\n",
    "\n",
    "ovTraintAll <- readRDS(paste0(cell2loc, \"cellTrek/ovTrainAll_obj.rds\"))\n",
    "\n",
    "seuSub <- readRDS(paste0(cell2loc, \"cellTrek/seuSub.rds\"))\n",
    "\n",
    "# --- define celltrek wrapper ---\n",
    "run_celltrek_fast <- function(trained_obj, sc_data, sample_name) {\n",
    "  message(Sys.time(), \"Processing \", sample_name) #adds timestamp to each log entry \n",
    "  \n",
    "  ct_result <- CellTrek::celltrek(\n",
    "    st_sc_int = trained_obj,\n",
    "    int_assay = \"traint\",\n",
    "    sc_data = sc_data,\n",
    "    sc_assay = \"RNA\",\n",
    "    reduction = \"pca\",\n",
    "    intp = TRUE,\n",
    "    intp_pnt = 1000,\n",
    "    intp_lin = FALSE,\n",
    "    nPCs = 15,\n",
    "    dist_thresh = 0.55,\n",
    "    top_spot = 5,\n",
    "    spot_n = 5,\n",
    "    repel_r = 25,\n",
    "    repel_iter = 10,\n",
    "    keep_model = TRUE,\n",
    "    ntree=100\n",
    "  )\n",
    "  message(Sys.time(), \" | Finished CellTrek for: \", sample_name)\n",
    "  return(ct_result$celltrek)\n",
    "}\n",
    "\n",
    "# --- run CellTrek on all samples ---\n",
    "with_progress({\n",
    "  p <- progressor(along = ovTraintAll)\n",
    "  \n",
    "  celltrek_results <- future_lapply(\n",
    "    seq_along(ovTraintAll),\n",
    "    function(i) {\n",
    "      sample_name <- names(ovTraintAll)[i]\n",
    "      message(Sys.time(), \" | Processing sample: \", sample_name)\n",
    "      \n",
    "      result <- run_celltrek_fast(\n",
    "        trained_obj = ovTraintAll[[i]],\n",
    "        sc_data = seuSub,\n",
    "        sample_name = sample_name\n",
    "      )\n",
    "      \n",
    "      saveRDS(result, file = paste0(\"celltrek_results\", sample_name, \".rds\"))\n",
    "      message(Sys.time(), \" | Saved result for: \", sample_name)\n",
    "      \n",
    "      p()  # update progress bar\n",
    "      return(result)\n",
    "    },\n",
    "    future.stdout = TRUE\n",
    "  )\n",
    "})\n",
    "\n",
    "saveRDS(celltrek_results, file = paste0(\"celltrek_results.rds\"))\n",
    "\n",
    "# --- merge results ---\n",
    "merged_celltrek <- merge(celltrek_results[[1]], y = celltrek_results[-1])\n",
    "merged_celltrek$treatment <- ifelse(\n",
    "  merged_celltrek$orig.ident %in% c(\"ov_1\",\"ov_2\",\"ov_3\"),\n",
    "  \"Untreated\", \"Treated\"\n",
    ")\n",
    "\n",
    "# --- downstream co-embedding ---\n",
    "DefaultAssay(merged_celltrek) <- \"RNA\"\n",
    "merged_celltrek <- NormalizeData(merged_celltrek)\n",
    "merged_celltrek <- FindVariableFeatures(merged_celltrek)\n",
    "merged_celltrek <- ScaleData(merged_celltrek)\n",
    "merged_celltrek <- RunPCA(merged_celltrek)\n",
    "merged_celltrek <- RunUMAP(merged_celltrek, dims = 1:25)\n",
    "\n",
    "# --- save results ---\n",
    "saveRDS(merged_celltrek, \"merged_celltrek.rds\")\n",
    "message(\"CellTrek pipeline finished successfully.\")\n",
    "\n",
    "    \n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9b730070",
   "metadata": {},
   "source": [
    "View co-embeddings:\n",
    " - can first merge (which requires normalisation etc again) if you want to compare treated vs untreated in one integrated embedding\n",
    " - or keep as separate objects and plot individual (in which case don't use merged rds)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f574cdbb",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "cell2loc = '/run/user/1804238067/gvfs/sftp:host=clust1-sub-1,user=lythgo02/mnt/scratchc/fmlab/lythgo02/OV_visium/emily/cell2location/'\n",
    "ovTraintAll <- readRDS(paste0(cell2loc, \"cellTrek/ovTrainAll_obj.rds\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cb9ac6c5",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'OV_1'</li><li>'OV_2'</li><li>'OV_3'</li><li>'OV_4'</li><li>'OV_5'</li><li>'OV_6'</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 'OV\\_1'\n",
       "\\item 'OV\\_2'\n",
       "\\item 'OV\\_3'\n",
       "\\item 'OV\\_4'\n",
       "\\item 'OV\\_5'\n",
       "\\item 'OV\\_6'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 'OV_1'\n",
       "2. 'OV_2'\n",
       "3. 'OV_3'\n",
       "4. 'OV_4'\n",
       "5. 'OV_5'\n",
       "6. 'OV_6'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] \"OV_1\" \"OV_2\" \"OV_3\" \"OV_4\" \"OV_5\" \"OV_6\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "names(ovTraintAll)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9f76b68e",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "#load \n",
    "\n",
    "ovCellTrekList <- readRDS(paste0(cell2loc, \"/cellTrek/celltrek_results.rds\"))\n",
    "\n",
    "names(ovCellTrekList) <- names(ovTraintAll)\n",
    "#convert updated annotation to factor with levels organised in alphabetical order for plotting purposes \n",
    "ovCellTrekList <- lapply(ovCellTrekList, function(ovCellTrek){\n",
    "    ovCellTrek$cell_type <- factor(ovCellTrek$updated_annotation,\n",
    "                                   levels=sort(unique(ovCellTrek$updated_annotation)))\n",
    "    return(ovCellTrek)\n",
    "})\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bff1494c",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "Set x and y limits for plotting \n",
    " - setting coordinates for xlim() and ylim() for ggplot to zoom/crop the plot to specific region of interest"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "87049fde",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "xl <- list()\n",
    "xl$OV_1 <- c(15, 585)\n",
    "xl$OV_2 <- c(20, 580)\n",
    "xl$OV_3 <- c(20, 580)\n",
    "xl$OV_4 <- c(20, 580)\n",
    "xl$OV_5 <- c(20, 580)\n",
    "xl$OV_6 <- c(15, 585)\n",
    "xl$OV_7 <- c(15, 585)\n",
    "xl$OV_8 <- c(15, 585)\n",
    "xl$OV_9 <- c(25, 575)\n",
    "xl$OV_10 <- c(20, 575)\n",
    "xl$OV_11 <- c(25, 575)\n",
    "xl$OV_12 <- c(25, 575)\n",
    "\n",
    "names(xl) <- names(ovCellTrekList)\n",
    "\n",
    "\n",
    "yl <- list()\n",
    "yl$OV_1 <- c(15, 585)\n",
    "yl$OV_2 <- c(15, 575)\n",
    "yl$OV_3 <- c(25, 565)\n",
    "yl$OV_4 <- c(25, 565)\n",
    "yl$OV_5 <- c(20, 580)    \n",
    "yl$OV_6 <- c(20, 570)\n",
    "yl$OV_7 <- c(35, 560)\n",
    "yl$OV_8 <- c(15, 585)\n",
    "yl$OV_9 <- c(25, 575)\n",
    "yl$OV_10 <- c(5, 575)  \n",
    "yl$OV_11 <- c(25, 565)\n",
    "yl$OV_12 <- c(25, 565)\n",
    "\n",
    "names(yl) <- names(ovCellTrekList)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "85117611",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "cellTrekPlots <- lapply(names(ovCellTrekList), function(sampN){\n",
    "    ovCellTrek <- ovCellTrekList[[sampN]]\n",
    "\n",
    "    img_fact <- ovCellTrek@images$slice1@scale.factors$lowres\n",
    "    img_temp <- ovCellTrek@images$slice1@image\n",
    "    img_data <- ovCellTrek@meta.data %>% dplyr::select(coord_x, coord_y, cell_type:id_new)\n",
    "    img_data$coord_x_new=img_data$coord_y*img_fact\n",
    "    img_data$coord_y_new=dim(img_temp)[1]-img_data$coord_x*img_fact\n",
    "    img_data$color_var <- factor(img_data[, \"cell_type\"])\n",
    "    img_data$shape_var <- ''\n",
    "\n",
    "    if (length(levels(img_data$color_var))<=9) {\n",
    "        pnt_colors <- brewer.pal(length(levels(img_data$color_var)), \"Set1\")\n",
    "    } else {\n",
    "        pnt_colors <- colorRampPalette(brewer.pal(9, \"Set1\"))(length(levels(img_data$color_var)))\n",
    "    }\n",
    "\n",
    "    if (length(levels(img_data$color_var))<=9) {\n",
    "        pnt_colors <- brewer.pal(length(levels(img_data$color_var)), \"Set1\")\n",
    "    } else {\n",
    "        pnt_colors <- colorRampPalette(brewer.pal(9, \"Set1\"))(length(levels(img_data$color_var)))\n",
    "    }\n",
    "    \n",
    "\n",
    "    ggplot(img_data, aes(x=coord_x_new, y=coord_y_new, fill=cell_type)) +\n",
    "    background_image(raster.img = img_temp)+\n",
    "    geom_jitter(size = 1, shape = 21, stroke = 0.02, color = \"black\") +\n",
    "    xlim(xl[[sampN]]) +\n",
    "    ylim(yl[[sampN]])\n",
    "\n",
    "})\n",
    "\n",
    "names(cellTrekPlots) <- names(ovCellTrekList)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a8275d99",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "cellTrekPlots"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "id": "59bff9e7",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "for (i in seq_along(cellTrekPlots)) {\n",
    "  ggsave(\n",
    "    filename = paste0(cell2loc, \"cellTrek/celltrek_OV\", i, \".pdf\"),\n",
    "    plot = cellTrekPlots[[i]],\n",
    "    width = 8,\n",
    "    height = 6,\n",
    "    dpi = 300\n",
    "  )\n",
    "}\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a2f4c6ac",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "source": [
    "Spatial Colocalisation \n",
    " - CellTrek::scoloc() analyzes how different cell types are spatially arranged relative to one another. It builds a graph-based model of spatial proximity and interaction:\n",
    " - use_method = 'DT': uses Delaunay triangulation to define spatial neighborhoods\n",
    " - boot_n = 200: performs 200 bootstrap iterations to assess statistical robustness"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6a71321c",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "\n",
    "ovSGraphKlList <- lapply(ovCellTrekList, function(ovCellTrek){\n",
    "    CellTrek::scoloc(\n",
    "        ovCellTrek,\n",
    "        col_cell = 'updated_annotation',\n",
    "        use_method = 'DT',\n",
    "        boot_n = 200\n",
    "    )\n",
    "})\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "89fcc107",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "source": [
    "Cellchat "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5b84299c",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "ename": "ERROR",
     "evalue": "Error in CellChat::computeRegionDistance(obCoord, group = ctOb$cell_type, : unused argument (group = ctOb$cell_type)\n",
     "output_type": "error",
     "traceback": [
      "Error in CellChat::computeRegionDistance(obCoord, group = ctOb$cell_type, : unused argument (group = ctOb$cell_type)\nTraceback:\n",
      "1. lapply(ovCellTrekList, function(ctOb) {\n .     obCoord <- ctOb@meta.data[, c(\"coord_x\", \"coord_y\")]\n .     obDist <- CellChat::computeRegionDistance(obCoord, group = ctOb$cell_type, \n .         spot)\n .     return(obDist)\n . })",
      "2. FUN(X[[i]], ...)",
      "3. .handleSimpleError(function (cnd) \n . {\n .     watcher$capture_plot_and_output()\n .     cnd <- sanitize_call(cnd)\n .     watcher$push(cnd)\n .     switch(on_error, continue = invokeRestart(\"eval_continue\"), \n .         stop = invokeRestart(\"eval_stop\"), error = NULL)\n . }, \"unused argument (group = ctOb$cell_type)\", base::quote(CellChat::computeRegionDistance(obCoord, \n .     group = ctOb$cell_type, spot)))"
     ]
    }
   ],
   "source": [
    "distMats <- lapply(cellTrekObs, function(ctOb){\n",
    "    obCoord <- ctOb@meta.data[,c(\"coord_x\", \"coord_y\")]\n",
    "    obDist <- CellChat::computeRegionDistance(obCoord, group = ctOb$cell_type, spot)\n",
    "    return(obDist)\n",
    "})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6ab524ec",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "7fbb2756",
   "metadata": {},
   "source": [
    "## 8. Co-localisation analysis\n",
    "Uses `CellTrek::scoloc()` to analyze spatial proximity between cell types. This builds a graph of neighboring cells, bootstraps it, and extracts consensus MST matrices. Later, differences between pre- and post-treatment are computed and visualized as heatmaps."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "09d9a10e",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "\n",
    "#### Co-localisation analysis\n",
    "ovSGraphKlList <- lapply(ovCellTrekList, function(ovCellTrek){\n",
    "    ovSGraphKl <- CellTrek::scoloc(ovCellTrek,\n",
    "                               col_cell='Final_Annotation',\n",
    "                               use_method='DT',\n",
    "                               boot_n=200)\n",
    "})\n",
    "samp1 <- ovCellTrekList[[1]]\n",
    "head(samp1$coord_x)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f88e3a8b",
   "metadata": {},
   "source": [
    "## 9. Visualization & output\n",
    "Plots cell maps overlaid on histology, generates combined figures, and saves results. Finally, writes CellTrek objects to disk for later use."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "700c6860",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "\n",
    "#### Write results\n",
    "saveRDS(ovCellTrekList, \"../data/cellTrek/ov_cellTrek_obj.rds\")\n",
    "sessionInfo()\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R (CellTrek Env)",
   "language": "R",
   "name": "celltrek_env"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.4.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
